<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>What's new in GAMBLR • GAMBLR.results</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="What's new in GAMBLR">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">GAMBLR.results</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Overview.html">What's new in GAMBLR</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>What’s new in GAMBLR</h1>
                        <h4 data-toc-skip class="author">Ryan Morin</h4>
            
      

      <div class="hidden name"><code>Overview.Rmd</code></div>

    </div>

    
    
<p>This document will demonstrate some commonly used features of
GAMBLR.results and the other packages. It’s intent is to familiarize
newcomers and current GAMBLR users to some recent improvements to the
GAMBLRverse.</p>
<p><strong>Requirements:</strong></p>
<ul>
<li><p>Have a GISTIC outputs named <em>all_lesions.DLBCL.grch37.txt</em>
and <em>scores.gistic</em> in your working directory.</p></li>
<li><p>Have the most recent version (1.3+) of GAMBLR.data, GAMBLR.open,
GAMBLR.utils, GAMBLR.viz and GAMBLR.results</p></li>
</ul>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">GAMBLR</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org" class="external-link">tidyverse</a></span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>We normally start by retrieving the metadata and sub-setting it to
the samples we desire. If working with mutations, you will probably want
to drop all rows with the “mrna” seq_type. It’s safest to remove those
from your table if you are working only with mutation data, although
most functions should just ignore seq_type rows that are not relevant.
This example also drops all FFPE samples and retains only DLBCL and FL
pathology.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">my_meta</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="fu"><a href="../reference/get_gambl_metadata.html">get_gambl_metadata</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span></span>
<span>    <span class="va">pathology</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"FL"</span>, <span class="st">"DLBCL"</span><span class="op">)</span>,</span>
<span>    <span class="va">ffpe_or_frozen</span> <span class="op">==</span> <span class="st">"frozen"</span></span>
<span>  <span class="op">)</span></span>
<span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">my_meta</span>, <span class="va">seq_type</span>, <span class="va">pathology</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/count.html" class="external-link">count</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 3</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># Groups:   seq_type, pathology [6]</span></span></span>
<span><span class="co">#&gt;   seq_type pathology     n</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> capture  DLBCL       690</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> capture  FL          103</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> genome   DLBCL       428</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> genome   FL          260</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> mrna     DLBCL       963</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> mrna     FL           79</span></span>
<span><span class="co"># drop mrna</span></span>
<span><span class="va">my_meta</span> <span class="op">&lt;-</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">my_meta</span>, <span class="va">seq_type</span> <span class="op">!=</span> <span class="st">"mrna"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">my_meta</span>, <span class="va">seq_type</span>, <span class="va">pathology</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/count.html" class="external-link">count</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 4 × 3</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># Groups:   seq_type, pathology [4]</span></span></span>
<span><span class="co">#&gt;   seq_type pathology     n</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> capture  DLBCL       690</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> capture  FL          103</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> genome   DLBCL       428</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> genome   FL          260</span></span></code></pre></div>
<p>If you are interested in an analysis of coding mutations
(i.e. mutations in coding exons), you can load them for a set of samples
that span exome/genome seq_type in a single step. Seasoned GAMBLRs will
recall that this used to require two separate calls to
<code>get_coding_ssm</code>, but this is no longer an issue thanks to
<code>get_all_coding_ssm</code>. The latter now requires that you
specify a metadata table via <code>these_samples_metadata</code> so it
can unambiguously determine which mutations to give you for each
seq_type for those samples that have both data types. A column
<code>maf_seq_type</code> is added to the output so you can always
determine which mutations came from <em>genome</em> vs <em>capture</em>
data.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dna_meta</span> <span class="op">&lt;-</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">my_meta</span>, <span class="va">seq_type</span> <span class="op">!=</span> <span class="st">"mrna"</span><span class="op">)</span></span>
<span><span class="va">all_coding</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="fu"><a href="../reference/get_all_coding_ssm.html">get_all_coding_ssm</a></span><span class="op">(</span></span>
<span>  these_samples_metadata <span class="op">=</span> <span class="va">dna_meta</span>,</span>
<span>  projection <span class="op">=</span> <span class="st">"hg38"</span>,</span>
<span>  include_silent <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">all_coding</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 366559</span></span>
<span><span class="co"># seq_type of each mutation is available via the maf_seq_type column</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">all_coding</span><span class="op">$</span><span class="va">maf_seq_type</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; capture  genome </span></span>
<span><span class="co">#&gt;  263848  102711</span></span></code></pre></div>
<p>If you look closely at the output of any function that returns a
MAF-like result (e.g. <code>get_all_coding_ssm</code>) you will notice
these are no longer just standard data frames. When you print the
contents, you will see “MAF Data Object” followed by the genome build.
This value is embedded directly in the data. The genome build can be
retrieved programmatically with the <code>get_genome_build</code>
function. This is one of a few special data types that will have a
built-in attribute <code>genome_build</code>. There are a bundle of S3
methods in <code>GAMBLR.utils</code> that ensure these attributes remain
<em>stuck</em> to these data types while you manipulate them with dplyr
verbs.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">all_coding</span>, <span class="fl">1</span><span class="op">)</span></span>
<span><span class="co">#&gt; genomic_data Object</span></span>
<span><span class="co">#&gt; Genome Build: hg38 </span></span>
<span><span class="co">#&gt; Showing first 10 rows:</span></span>
<span><span class="co">#&gt;   Hugo_Symbol Entrez_Gene_Id Center NCBI_Build Chromosome Start_Position</span></span>
<span><span class="co">#&gt; 1       OR4F5              0      .     GRCh38       chr1          69634</span></span>
<span><span class="co">#&gt;   End_Position Strand Variant_Classification Variant_Type Reference_Allele</span></span>
<span><span class="co">#&gt; 1        69634      +      Missense_Mutation          SNP                T</span></span>
<span><span class="co">#&gt;   Tumor_Seq_Allele1 Tumor_Seq_Allele2    dbSNP_RS dbSNP_Val_Status</span></span>
<span><span class="co">#&gt; 1                 T                 A rs745433879               NA</span></span>
<span><span class="co">#&gt;   Tumor_Sample_Barcode Matched_Norm_Sample_Barcode Match_Norm_Seq_Allele1</span></span>
<span><span class="co">#&gt; 1          DLBCL11262T           04-24937N-Schmitz                      T</span></span>
<span><span class="co">#&gt;   Match_Norm_Seq_Allele2 Tumor_Validation_Allele1 Tumor_Validation_Allele2</span></span>
<span><span class="co">#&gt; 1                      T                       NA                       NA</span></span>
<span><span class="co">#&gt;   Match_Norm_Validation_Allele1 Match_Norm_Validation_Allele2</span></span>
<span><span class="co">#&gt; 1                            NA                            NA</span></span>
<span><span class="co">#&gt;   Verification_Status Validation_Status Mutation_Status Sequencing_Phase</span></span>
<span><span class="co">#&gt; 1                  NA                NA              NA               NA</span></span>
<span><span class="co">#&gt;   Sequence_Source Validation_Method Score BAM_File Sequencer Tumor_Sample_UUID</span></span>
<span><span class="co">#&gt; 1              NA                NA    NA       NA        NA                NA</span></span>
<span><span class="co">#&gt;   Matched_Norm_Sample_UUID    HGVSc       HGVSp HGVSp_Short   Transcript_ID</span></span>
<span><span class="co">#&gt; 1                       NA c.544N&gt;A p.Cys182Ser     p.C182S ENST00000335137</span></span>
<span><span class="co">#&gt;   Exon_Number t_depth t_ref_count t_alt_count n_depth n_ref_count n_alt_count</span></span>
<span><span class="co">#&gt; 1         1/1      57           0          57      17          17           0</span></span>
<span><span class="co">#&gt;   maf_seq_type</span></span>
<span><span class="co">#&gt; 1      capture</span></span>
<span></span>
<span><span class="va">maf_selected</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">all_coding</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">7</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">maf_filtered</span> <span class="op">&lt;-</span> <span class="va">all_coding</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">Hugo_Symbol</span> <span class="op">==</span> <span class="st">"EZH2"</span><span class="op">)</span></span>
<span><span class="va">maf_mutated</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span><span class="va">all_coding</span>, new_column <span class="op">=</span> <span class="st">"justatest"</span><span class="op">)</span></span>
<span><span class="fu">get_genome_build</span><span class="op">(</span><span class="va">all_coding</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "hg38"</span></span>
<span><span class="fu">get_genome_build</span><span class="op">(</span><span class="va">maf_selected</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "hg38"</span></span>
<span><span class="fu">get_genome_build</span><span class="op">(</span><span class="va">maf_filtered</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "hg38"</span></span>
<span><span class="fu">get_genome_build</span><span class="op">(</span><span class="va">maf_mutated</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "hg38"</span></span></code></pre></div>
<p>This built-in feature now protects against accidental mixing of the
two coordinate systems. When we provide this data to another function,
the genome_build argument is no longer required because it’s inferred
from the data.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># This doesn't work</span></span>
<span><span class="va">coding_tabulated_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_coding_ssm_status.html">get_coding_ssm_status</a></span><span class="op">(</span></span>
<span>  maf_data <span class="op">=</span> <span class="va">all_coding</span>,</span>
<span>  these_samples_metadata <span class="op">=</span> <span class="va">dna_meta</span>,</span>
<span>  genome_build <span class="op">=</span> <span class="st">"grch37"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Error in `get_coding_ssm_status()`:</span></span>
<span><span class="co"># ! you have specified a genome_build that doesn't match the</span></span>
<span><span class="co"># genome_build attached to maf_data</span></span>
<span></span>
<span><span class="co"># This will also fail for a different reason. </span></span>
<span><span class="co"># The automatic annotation of hotspots is not available for hg38</span></span>
<span><span class="va">coding_tabulated_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_coding_ssm_status.html">get_coding_ssm_status</a></span><span class="op">(</span></span>
<span>  maf_data <span class="op">=</span> <span class="va">all_coding</span>,</span>
<span>  these_samples_metadata <span class="op">=</span> <span class="va">dna_meta</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Error in check_get_projection(list(this_maf = mutation_maf), "grch37",  : </span></span>
<span><span class="co"># This function currently only supports grch37</span></span></code></pre></div>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># The workaround until someone resolves this is to disable that feature</span></span>
<span><span class="va">coding_tabulated_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_coding_ssm_status.html">get_coding_ssm_status</a></span><span class="op">(</span></span>
<span>  maf_data <span class="op">=</span> <span class="va">all_coding</span>,</span>
<span>  these_samples_metadata <span class="op">=</span> <span class="va">dna_meta</span>,</span>
<span>  include_hotspots <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">coding_tabulated_df</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 10</span></span></span>
<span><span class="co">#&gt;   sample_id TNFRSF14  NOL9 KLHL21 CAMTA1 PIK3CD  MTOR  SPEN  E2F2   ID3</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> FL1011T1         0     0      0      0      0     0     0     0     0</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> FL3001T1         1     0      0      0      0     0     0     0     0</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> FL1015T1         0     0      0      0      0     0     0     0     0</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> FL1015T2         0     0      0      0      0     0     0     0     0</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> FL1002T1         1     0      0      0      0     0     0     0     0</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> FL1002T2         1     0      0      0      0     0     0     0     0</span></span>
<span><span class="va">coding_tabulated_df</span> <span class="op">=</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html" class="external-link">column_to_rownames</a></span><span class="op">(</span><span class="va">coding_tabulated_df</span>,<span class="st">"sample_id"</span><span class="op">)</span> </span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">colSums</a></span><span class="op">(</span><span class="va">coding_tabulated_df</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">20</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="co">#&gt; TNFRSF14     NOL9   KLHL21   CAMTA1   PIK3CD     MTOR     SPEN     E2F2 </span></span>
<span><span class="co">#&gt;      311       67       43       36       46       45      106        9 </span></span>
<span><span class="co">#&gt;      ID3   ARID1A   LAPTM5  COL16A1     AGO4   MAP7D1  ZC3H12A    RRAGC </span></span>
<span><span class="co">#&gt;       34      124       42       31        8       11       56       58 </span></span>
<span><span class="co">#&gt;   CDKN2C     JAK1   DNAJC6    FUBP1 </span></span>
<span><span class="co">#&gt;        6       25       15       25</span></span></code></pre></div>
<p>Testing some other visualizations</p>
<p><code>heatmap_mutation_frequency_bin</code></p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">my_meta</span> <span class="op">&lt;-</span> <span class="fu">GAMBLR.open</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/GAMBLR.open/man/get_gambl_metadata.html" class="external-link">get_gambl_metadata</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">cohort</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="st">"FL_Dreval"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Use the new approach to embed genome_build in your regions_bed</span></span>
<span><span class="co"># and create/reformat the name column</span></span>
<span><span class="va">my_regions</span> <span class="op">&lt;-</span> <span class="fu">create_bed_data</span><span class="op">(</span></span>
<span>  <span class="fu">GAMBLR.data</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/GAMBLR.data/man/hg38_ashm_regions.html" class="external-link">hg38_ashm_regions</a></span>,</span>
<span>  genome_build <span class="op">=</span> <span class="st">"hg38"</span>,</span>
<span>  fix_names <span class="op">=</span> <span class="st">"concat"</span>,</span>
<span>  concat_cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"gene"</span>, <span class="st">"region"</span><span class="op">)</span>, sep <span class="op">=</span> <span class="st">"-"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="va">meta_columns</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"pathology"</span>, <span class="st">"lymphgen"</span>, <span class="st">"COO_consensus"</span>, <span class="st">"DHITsig_consensus"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">all_wide</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calc_mutation_frequency_bin_regions.html">calc_mutation_frequency_bin_regions</a></span><span class="op">(</span></span>
<span>  regions_bed <span class="op">=</span> <span class="va">my_regions</span>,</span>
<span>  these_samples_metadata <span class="op">=</span> <span class="va">my_meta</span>,</span>
<span>  slide_by <span class="op">=</span> <span class="fl">100</span>,</span>
<span>  window_size <span class="op">=</span> <span class="fl">500</span>,</span>
<span>  return_format <span class="op">=</span> <span class="st">"wide"</span></span>
<span><span class="op">)</span></span>
<span><span class="va">all_wide</span> <span class="op">&lt;-</span> <span class="va">all_wide</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html" class="external-link">column_to_rownames</a></span><span class="op">(</span><span class="st">"sample_id"</span><span class="op">)</span></span>
<span><span class="co"># this can be used as mut_freq_matrix</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">all_wide</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="co">#&gt;  [1] "chr1_6600822" "chr1_6600922" "chr1_6601022" "chr1_6601122" "chr1_6601222"</span></span>
<span><span class="co">#&gt;  [6] "chr1_6601322" "chr1_6601422" "chr1_6601522" "chr1_6601822" "chr1_6601922"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rev.html" class="external-link">rev</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">colSums</a></span><span class="op">(</span><span class="va">all_wide</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; chr18_63318795 chr18_63318695 chr18_63318895 chr18_63318595 chr18_63318995 </span></span>
<span><span class="co">#&gt;           1336           1334           1291           1268           1128 </span></span>
<span><span class="co">#&gt; chr22_22888074 </span></span>
<span><span class="co">#&gt;           1102</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span></span>
<span>  <span class="fu">heatmap_mutation_frequency_bin</span><span class="op">(</span></span>
<span>    regions_bed <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">my_regions</span>, <span class="fl">25</span><span class="op">)</span>,</span>
<span>    these_samples_metadata <span class="op">=</span> <span class="va">my_meta</span>,</span>
<span>    metadataColumns <span class="op">=</span> <span class="va">meta_columns</span>,</span>
<span>    sortByColumns <span class="op">=</span> <span class="va">meta_columns</span>,</span>
<span>    backgroundColour <span class="op">=</span> <span class="st">"white"</span>,</span>
<span>    label_regions_rotate <span class="op">=</span> <span class="fl">90</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="Overview_files/figure-html/heatmap1-1.png" width="700"></p>
<p>This also works seamlessly for capture data.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">my_other_regions</span> <span class="op">&lt;-</span> <span class="fu">create_bed_data</span><span class="op">(</span></span>
<span>  <span class="fu">GAMBLR.data</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/GAMBLR.data/man/grch37_ashm_regions.html" class="external-link">grch37_ashm_regions</a></span>,</span>
<span>  fix_names <span class="op">=</span> <span class="st">"concat"</span>,</span>
<span>  concat_cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"gene"</span>, <span class="st">"region"</span><span class="op">)</span>, sep <span class="op">=</span> <span class="st">"-"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="va">cap_meta</span> <span class="op">&lt;-</span> <span class="fu">GAMBLR.results</span><span class="fu">::</span><span class="fu"><a href="../reference/get_gambl_metadata.html">get_gambl_metadata</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">cohort</span> <span class="op">==</span> <span class="st">"dlbcl_schmitz"</span>, <span class="va">seq_type</span> <span class="op">!=</span> <span class="st">"mrna"</span><span class="op">)</span></span>
<span><span class="va">meta_columns</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"pathology"</span>, <span class="st">"lymphgen"</span>, <span class="st">"COO_consensus"</span>, <span class="st">"seq_type"</span><span class="op">)</span></span>
<span><span class="fu">heatmap_mutation_frequency_bin</span><span class="op">(</span></span>
<span>  regions_bed <span class="op">=</span> <span class="va">my_other_regions</span>,</span>
<span>  these_samples_metadata <span class="op">=</span> <span class="va">cap_meta</span>,</span>
<span>  metadataColumns <span class="op">=</span> <span class="va">meta_columns</span>,</span>
<span>  sortByColumns <span class="op">=</span> <span class="va">meta_columns</span>,</span>
<span>  label_regions_rotate <span class="op">=</span> <span class="fl">90</span>,</span>
<span>  backgroundColour <span class="op">=</span> <span class="st">"white"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="Overview_files/figure-html/heatmap_2-1.png" width="700"></p>
<p>Note that both genome builds are supported and the correct coordinate
system is inferred from regions_bed.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># use the original my_regions (grch37)</span></span>
<span><span class="fu">heatmap_mutation_frequency_bin</span><span class="op">(</span></span>
<span>  regions_bed <span class="op">=</span> <span class="va">my_regions</span>,</span>
<span>  these_samples_metadata <span class="op">=</span> <span class="va">cap_meta</span>,</span>
<span>  metadataColumns <span class="op">=</span> <span class="va">meta_columns</span>,</span>
<span>  sortByColumns <span class="op">=</span> <span class="va">meta_columns</span>,</span>
<span>  label_regions_rotate <span class="op">=</span> <span class="fl">90</span>,</span>
<span>  backgroundColour <span class="op">=</span> <span class="st">"white"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="Overview_files/figure-html/heatmap_3-1.png" width="700"></p>
<p>Want to visualize the mutation pattern of a single gene? Sounds like
you want a <code>pretty_lollipop_plot</code></p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lolipop_result</span> <span class="op">&lt;-</span> <span class="fu">pretty_lollipop_plot</span><span class="op">(</span><span class="va">all_coding</span>, <span class="st">"MYC"</span><span class="op">)</span></span>
<span><span class="va">lolipop_result</span></span></code></pre></div>
<p><img src="Overview_files/figure-html/unnamed-chunk-3-1.png" width="700"></p>
<p>Planning to compare the mutation pattern of a set of genes across
many patients, possibly spanning multiple pathologies? This is
definitely a job for <code>prettyOncoplot</code></p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co">#| warning: false</span></span>
<span><span class="co">#| message: false</span></span>
<span><span class="va">my_meta</span> <span class="op">&lt;-</span> <span class="fu">GAMBLR.results</span><span class="fu">::</span><span class="fu"><a href="../reference/get_gambl_metadata.html">get_gambl_metadata</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span></span>
<span>    <span class="va">pathology</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"FL"</span>, <span class="st">"DLBCL"</span><span class="op">)</span>,</span>
<span>    <span class="va">ffpe_or_frozen</span> <span class="op">==</span> <span class="st">"frozen"</span>,</span>
<span>    <span class="va">seq_type</span> <span class="op">!=</span> <span class="st">"mrna"</span></span>
<span>  <span class="op">)</span></span>
<span><span class="co">#&gt; 3258 capture samples are missing a value for protocol. Assuming Exome.</span></span>
<span><span class="co">#&gt; 63 biopsies are missing from the biopsy metadata. This should be fixed!</span></span>
<span><span class="co">#&gt; affected cohorts:  DLBCL_LSARP_Trios,SMZL_Strefford,cHL_Maura</span></span>
<span><span class="co">#&gt; 110 biopsies with discrepancies in the pathology field. This should be fixed!</span></span>
<span><span class="co">#&gt; 10 biopsies with discrepancies in the time_point field. This should be fixed!</span></span>
<span><span class="fu">prettyOncoplot</span><span class="op">(</span><span class="va">all_coding</span>,</span>
<span>  genes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"EZH2"</span>, <span class="st">"CREBBP"</span>, <span class="st">"MYD88"</span>, <span class="st">"CCND3"</span>, <span class="st">"KMT2D"</span>, <span class="st">"IGLL5"</span><span class="op">)</span>,</span>
<span>  these_samples_metadata <span class="op">=</span> <span class="va">my_meta</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; There are no missing genes during onco matrix assembly.</span></span>
<span><span class="co">#&gt; You should name at least one metadata column to show as an annotation. Defaulting to pathology</span></span>
<span><span class="co">#&gt; All mutation types: Missense_Mutation, Multi_Hit, Splice_Site,</span></span>
<span><span class="co">#&gt; Frame_Shift_Del, In_Frame_Del, Nonsense_Mutation, Frame_Shift_Ins,</span></span>
<span><span class="co">#&gt; In_Frame_Ins, Translation_Start_Site, Nonstop_Mutation.</span></span>
<span><span class="co">#&gt; `alter_fun` is assumed vectorizable. If it does not generate correct</span></span>
<span><span class="co">#&gt; plot, please set `alter_fun_is_vectorized = FALSE` in `oncoPrint()`.</span></span>
<span><span class="co">#&gt; No legend element is put in the last 1 row under `nrow = 3`, maybe you</span></span>
<span><span class="co">#&gt; should set `by_row = FALSE`? Reset `nrow` to 2.</span></span>
<span><span class="co">#&gt; No legend element is put in the last 1 row under `nrow = 3`, maybe you</span></span>
<span><span class="co">#&gt; should set `by_row = FALSE`? Reset `nrow` to 2.</span></span>
<span><span class="co">#&gt; No legend element is put in the last 1 row under `nrow = 3`, maybe you</span></span>
<span><span class="co">#&gt; should set `by_row = FALSE`? Reset `nrow` to 2.</span></span></code></pre></div>
<p><img src="Overview_files/figure-html/pretty1-1.png" width="700"></p>
<p>Letting prettyOncoplot decide which genes to include based on
mutation frequency in the data.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="fu">prettyOncoplot</span><span class="op">(</span><span class="va">all_coding</span>,</span>
<span>  minMutationPercent <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  simplify_annotation <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  these_samples_metadata <span class="op">=</span> <span class="va">my_meta</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; creating oncomatrix with 182 genes</span></span>
<span><span class="co">#&gt; There are no missing genes during onco matrix assembly.</span></span>
<span><span class="co">#&gt; You should name at least one metadata column to show as an annotation. Defaulting to pathology</span></span>
<span><span class="co">#&gt; All mutation types: Missense, Truncating, Splice_Site.</span></span>
<span><span class="co">#&gt; `alter_fun` is assumed vectorizable. If it does not generate correct</span></span>
<span><span class="co">#&gt; plot, please set `alter_fun_is_vectorized = FALSE` in `oncoPrint()`.</span></span>
<span><span class="co">#&gt; No legend element is put in the last 1 row under `nrow = 3`, maybe you</span></span>
<span><span class="co">#&gt; should set `by_row = FALSE`? Reset `nrow` to 2.</span></span>
<span><span class="co">#&gt; No legend element is put in the last 1 row under `nrow = 3`, maybe you</span></span>
<span><span class="co">#&gt; should set `by_row = FALSE`? Reset `nrow` to 2.</span></span>
<span><span class="co">#&gt; No legend element is put in the last 1 row under `nrow = 3`, maybe you</span></span>
<span><span class="co">#&gt; should set `by_row = FALSE`? Reset `nrow` to 2.</span></span></code></pre></div>
<p><img src="Overview_files/figure-html/pretty2-1.png" width="700"></p>
<p>Attaching non-coding mutation data to an oncoplot involves first
creating an aSHM count matrix.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co">#| warning: false</span></span>
<span><span class="co">#| message: false</span></span>
<span><span class="va">my_meta</span> <span class="op">&lt;-</span> <span class="fu">GAMBLR.open</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/GAMBLR.open/man/get_gambl_metadata.html" class="external-link">get_gambl_metadata</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">cohort</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="st">"FL_Dreval"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Using the bundled metadata in GAMBLR.data...</span></span>
<span></span>
<span><span class="co"># Use the new approach to embed genome_build in your regions_bed</span></span>
<span><span class="co"># and create/reformat the name column</span></span>
<span><span class="va">my_regions</span> <span class="op">&lt;-</span> <span class="fu">create_bed_data</span><span class="op">(</span></span>
<span>  <span class="fu">GAMBLR.data</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/GAMBLR.data/man/hg38_ashm_regions.html" class="external-link">hg38_ashm_regions</a></span>,</span>
<span>  genome_build <span class="op">=</span> <span class="st">"hg38"</span>,</span>
<span>  fix_names <span class="op">=</span> <span class="st">"concat"</span>,</span>
<span>  concat_cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"gene"</span>, <span class="st">"region"</span><span class="op">)</span>, sep <span class="op">=</span> <span class="st">"-"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">my_regions</span><span class="op">)</span></span>
<span><span class="co">#&gt; BED Data Object</span></span>
<span><span class="co">#&gt; Genome Build: hg38 </span></span>
<span><span class="co">#&gt; Showing first 10 rows:</span></span>
<span><span class="co">#&gt;    chrom     start       end           name region regulatory_comment</span></span>
<span><span class="co">#&gt; 1   chr1   6601422   6602642     KLHL21-TSS    TSS               &lt;NA&gt;</span></span>
<span><span class="co">#&gt; 2   chr1  23559093  23559344        ID3-TSS    TSS               &lt;NA&gt;</span></span>
<span><span class="co">#&gt; 3   chr1  28506039  28509827       RCC1-TSS    TSS               &lt;NA&gt;</span></span>
<span><span class="co">#&gt; 4   chr1  30756165  30759164     LAPTM5-TSS    TSS               &lt;NA&gt;</span></span>
<span><span class="co">#&gt; 5   chr1 150578338 150579659    MCL1-intron intron               &lt;NA&gt;</span></span>
<span><span class="co">#&gt; 6   chr2  88605321  88609578 EIF2Ak3-intron intron               &lt;NA&gt;</span></span>
<span><span class="co">#&gt; 7   chr2  88625938  88628063    EIF2Ak3-TSS    TSS               &lt;NA&gt;</span></span>
<span><span class="co">#&gt; 8   chr2 231707930 231709587       PTMA-TSS    TSS               &lt;NA&gt;</span></span>
<span><span class="co">#&gt; 9   chr2 156812978 156814787      FCRL3-TSS    TSS               &lt;NA&gt;</span></span>
<span><span class="co">#&gt; 10  chr1 203305570 203306650    BTG2-intron intron    active_promoter</span></span>
<span></span>
<span><span class="va">ashm_mat</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_ashm_count_matrix.html">get_ashm_count_matrix</a></span><span class="op">(</span></span>
<span>  regions_bed <span class="op">=</span> <span class="va">my_regions</span>,</span>
<span>  these_samples_metadata <span class="op">=</span> <span class="va">my_meta</span>,</span>
<span>  this_seq_type <span class="op">=</span> <span class="st">"genome"</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Streamlined is set to TRUE, this function will disregard anything specified with basic_columns</span></span>
<span><span class="co">#&gt; To return a MAF with standard 45 columns, set streamlioned = FALSE and basic_columns = TRUE</span></span>
<span><span class="co">#&gt; To return a maf with all (116) columns, set streamlined = FALSE and basic_columns = FALSE</span></span>
<span><span class="co">#&gt; Joining with `by = join_by(sample_id, region_name)`</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">ashm_mat</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 443 130</span></span></code></pre></div>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co">#| warning: false</span></span>
<span><span class="co">#| message: false</span></span>
<span><span class="co"># get coding mutations (hg38)</span></span>
<span><span class="va">my_coding</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_all_coding_ssm.html">get_all_coding_ssm</a></span><span class="op">(</span></span>
<span>  these_samples_metadata <span class="op">=</span> <span class="va">my_meta</span>,</span>
<span>  include_silent <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  projection <span class="op">=</span> <span class="st">"hg38"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><code>get_ssm_by_samples</code></p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># get a few exomes and a few genomes</span></span>
<span></span>
<span><span class="va">some_exome</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_gambl_metadata.html">get_gambl_metadata</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span></span>
<span>  <span class="va">seq_type</span> <span class="op">==</span> <span class="st">"capture"</span>,</span>
<span>  <span class="va">sample_id</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>    <span class="st">"00-22011_tumorB"</span>,</span>
<span>    <span class="st">"PA064"</span>,</span>
<span>    <span class="st">"01-1961"</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">some_genome</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_gambl_metadata.html">get_gambl_metadata</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span></span>
<span>  <span class="va">seq_type</span> <span class="op">==</span> <span class="st">"genome"</span>,</span>
<span>  <span class="va">sample_id</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>    <span class="st">"HTMCP-01-06-00485-01A-01D"</span>,</span>
<span>    <span class="st">"DOHH-2"</span>,</span>
<span>    <span class="st">"SU-DHL-10"</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">some_meta</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/bind_rows.html" class="external-link">bind_rows</a></span><span class="op">(</span><span class="va">some_exome</span>, <span class="va">some_genome</span><span class="op">)</span></span>
<span></span>
<span><span class="va">some_from_samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_ssm_by_samples.html">get_ssm_by_samples</a></span><span class="op">(</span></span>
<span>  these_samples_metadata <span class="op">=</span> <span class="va">some_meta</span>,</span>
<span>  projection <span class="op">=</span> <span class="st">"hg38"</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "The value of this_seq_type was ignored and all samples in these_samples_metadata and these_sample_ids (if provided) were included"</span></span>
<span></span>
<span></span>
<span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">some_from_samples</span>, <span class="va">Tumor_Sample_Barcode</span>, <span class="va">maf_seq_type</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/count.html" class="external-link">tally</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 3</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># Groups:   Tumor_Sample_Barcode [6]</span></span></span>
<span><span class="co">#&gt;   Tumor_Sample_Barcode      maf_seq_type     n</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> 00-22011_tumorB           capture       <span style="text-decoration: underline;">1</span>002</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> 01-1961                   capture        547</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> DOHH-2                    genome       <span style="text-decoration: underline;">21</span>930</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> HTMCP-01-06-00485-01A-01D genome        <span style="text-decoration: underline;">2</span>165</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> PA064                     capture       <span style="text-decoration: underline;">1</span>056</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> SU-DHL-10                 genome       <span style="text-decoration: underline;">26</span>707</span></span></code></pre></div>
<p>We have to link our metadata to the numeric data to include it in the
prettyOncoplot output.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co">#| warning: false</span></span>
<span><span class="co">#| message: false</span></span>
<span><span class="co"># join the aSHM matrix to the metadata</span></span>
<span></span>
<span><span class="va">meta_with_ashm</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html" class="external-link">rownames_to_column</a></span><span class="op">(</span><span class="va">ashm_mat</span>,</span>
<span>                                     <span class="st">"sample_id"</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html" class="external-link">left_join</a></span><span class="op">(</span><span class="va">my_meta</span>, <span class="va">.</span><span class="op">)</span></span>
<span><span class="co">#&gt; Joining with `by = join_by(sample_id)`</span></span>
<span></span>
<span><span class="fu">prettyOncoplot</span><span class="op">(</span></span>
<span>  maf_df <span class="op">=</span> <span class="va">my_coding</span>,</span>
<span>  these_samples_metadata <span class="op">=</span> <span class="va">meta_with_ashm</span>,</span>
<span>  metadataColumns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"pathology"</span>, <span class="st">"cohort"</span><span class="op">)</span>,</span>
<span>  numericMetadataColumns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">ashm_mat</span><span class="op">)</span>,</span>
<span>  stacked <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  minMutationPercent <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  sortByColumns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"pathology"</span>, <span class="st">"cohort"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; stacked mode is only compatible in combination with simplify = TRUE. Setting this for you.</span></span>
<span><span class="co">#&gt; creating oncomatrix with 80 genes</span></span>
<span><span class="co">#&gt; There are no missing genes during onco matrix assembly.</span></span>
<span><span class="co">#&gt; The column(s) specified both in metadata and numeric metadata will be plotted as numeric values...</span></span>
<span><span class="co">#&gt; dropping 22 genes due to sparseness</span></span>
<span><span class="co">#&gt; HNRNPU-TSS,EIF2Ak3-TSS,LPP-intron-1,CD44-TSS,FOXP1-TSS-2,FOXP1-intron-1a,FANK1-TSS,NFKBIZ-TSS-1,SEPTIN9-intron-1,FOXP1-TSS-1,MEF2C-TSS-1,WEE1-intron,MCL1-intron,SEPTIN9-intron-3,P2RX5-TSS,ID3-TSS,NFKBIZ-TSS-2,RNF144B-TSS,RHEX-TSS,FOXP1-intron-1b,KLHL21-TSS,FCRL3-TSS</span></span>
<span><span class="co">#&gt; All mutation types: Missense, Truncating, Splice_Site.</span></span>
<span><span class="co">#&gt; `alter_fun` is assumed vectorizable. If it does not generate correct</span></span>
<span><span class="co">#&gt; plot, please set `alter_fun_is_vectorized = FALSE` in `oncoPrint()`.</span></span></code></pre></div>
<p><img src="Overview_files/figure-html/join_to_meta-1.png" width="700"></p>
<p>Comparing two groups with <code>prettyCoOncoplot</code></p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="fu">prettyCoOncoplot</span><span class="op">(</span><span class="va">all_coding</span>,</span>
<span>  <span class="va">my_meta</span>,</span>
<span>  genes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>    <span class="st">"EZH2"</span>, <span class="st">"CREBBP"</span>, <span class="st">"MYD88"</span>, <span class="st">"CCND3"</span>,</span>
<span>    <span class="st">"KMT2D"</span>, <span class="st">"IGLL5"</span>, <span class="st">"CARD11"</span>, <span class="st">"BTG1"</span>, <span class="st">"ATP6V1B2"</span></span>
<span>  <span class="op">)</span>,</span>
<span>  <span class="st">"pathology"</span>,</span>
<span>  label1 <span class="op">=</span> <span class="st">"FL"</span>,</span>
<span>  label2 <span class="op">=</span> <span class="st">"DLBCL"</span>,</span>
<span>  comparison_values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"FL"</span>, <span class="st">"DLBCL"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Want to know which mutations are more prevalent in one group of
patients vs another? You need <code>prettyForestPlot</code> in your
life!</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">out_list</span> <span class="op">=</span> <span class="fu">prettyForestPlot</span><span class="op">(</span><span class="va">all_coding</span>,</span>
<span>  metadata <span class="op">=</span> <span class="va">my_meta</span>,</span>
<span>  genes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"EZH2"</span>, <span class="st">"CREBBP"</span>, <span class="st">"MYD88"</span>, <span class="st">"CCND3"</span>, <span class="st">"KMT2D"</span>,</span>
<span>            <span class="st">"IGLL5"</span>, <span class="st">"CARD11"</span>, <span class="st">"BTG1"</span>, <span class="st">"ATP6V1B2"</span><span class="op">)</span>,</span>
<span>  comparison_column <span class="op">=</span> <span class="st">"pathology"</span>,</span>
<span>  comparison_values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"FL"</span>, <span class="st">"DLBCL"</span><span class="op">)</span>,</span>
<span>  separate_hotspots <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  comparison_name <span class="op">=</span> <span class="st">"FL vs DLBCL"</span></span>
<span><span class="op">)</span></span>
<span><span class="va">out_list</span><span class="op">$</span><span class="va">arranged</span></span></code></pre></div>
<p><img src="Overview_files/figure-html/forestplot-1.png" width="700"></p>
<p>There are other data types that benefit from the same approach.
Another major player is segmented copy number results (seg_data), which
you can obtain from get_cn_segments in <code>GAMBLR.results</code> or
<code>GAMBLR.data</code>.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co">#| warning: false</span></span>
<span><span class="co">#| message: false</span></span>
<span><span class="va">all_segments</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_cn_segments.html">get_cn_segments</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; no metadata provided</span></span>
<span><span class="co">#&gt; will get segments for all available genome and capture samples</span></span>
<span><span class="co">#&gt; dummy segments are not annotated in the inputs</span></span>
<span><span class="co">#&gt; fill_missing_with parameter will be ignored</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">all_segments</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; SEG Data Object</span></span>
<span><span class="co">#&gt; Genome Build: grch37 </span></span>
<span><span class="co">#&gt; Showing first 10 rows:</span></span>
<span><span class="co">#&gt;                ID chrom    start      end LOH_flag log.ratio seg_seq_type</span></span>
<span><span class="co">#&gt; 1 00-22011_tumorB     1    10001    69372       NA    0.0000      capture</span></span>
<span><span class="co">#&gt; 2 00-22011_tumorB     1    69373 10395622       NA    1.0000      capture</span></span>
<span><span class="co">#&gt; 3 00-22011_tumorB     1 10395622 12783034       NA    0.0000      capture</span></span>
<span><span class="co">#&gt; 4 00-22011_tumorB     1 12783034 13292951       NA   -2.0000      capture</span></span>
<span><span class="co">#&gt; 5 00-22011_tumorB     1 13292951 33768448       NA    0.0000      capture</span></span>
<span><span class="co">#&gt; 6 00-22011_tumorB     1 33768448 33775298       NA    2.5453      capture</span></span>
<span><span class="co">#&gt;         CN</span></span>
<span><span class="co">#&gt; 1  2.00000</span></span>
<span><span class="co">#&gt; 2  4.00000</span></span>
<span><span class="co">#&gt; 3  2.00000</span></span>
<span><span class="co">#&gt; 4  0.50000</span></span>
<span><span class="co">#&gt; 5  2.00000</span></span>
<span><span class="co">#&gt; 6 11.67459</span></span>
<span></span>
<span><span class="fu">get_genome_build</span><span class="op">(</span><span class="va">all_segments</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "grch37"</span></span></code></pre></div>
<p>Some new functionality for copy number data is implemented in this
function:</p>
<p><code><a href="https://rdrr.io/pkg/GAMBLR.utils/man/segmented_data_to_cn_matrix.html" class="external-link">GAMBLR.utils::segmented_data_to_cn_matrix</a></code></p>
<p>This function uses one of four approaches to convert segmented data
into a matrix of CN values. The most flexible approach is to let it
automatically split the genome into equal-sized bins for you. It then
uses the data in seg_data to infer the copy number value for each bin
using all overlapping segments. The interface for this function has
changed to add flexibility and transparency.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co">#| warning: false</span></span>
<span><span class="co">#| message: false</span></span>
<span><span class="va">start_time</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Sys.time.html" class="external-link">Sys.time</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">cn_mat</span> <span class="op">&lt;-</span> <span class="fu">segmented_data_to_cn_matrix</span><span class="op">(</span></span>
<span>  seg_data <span class="op">=</span> <span class="va">all_segments</span>,</span>
<span>  n_bins_split <span class="op">=</span> <span class="fl">500</span>,</span>
<span>  strategy <span class="op">=</span> <span class="st">"auto_split"</span>, <span class="co"># this is the default</span></span>
<span>  these_samples_metadata <span class="op">=</span> <span class="va">dna_meta</span>,</span>
<span>  fill_missing_with <span class="op">=</span> <span class="st">"diploid"</span>,</span>
<span>  adjust_for_ploidy <span class="op">=</span> <span class="cn">TRUE</span></span>
<span>  <span class="co"># This centers all samples using their genome-wide average CN</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "0 rows with NA values"</span></span>
<span><span class="va">complete_time</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Sys.time.html" class="external-link">Sys.time</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co"># this process is slow if you use a lot of bins</span></span>
<span><span class="va">complete_time</span> <span class="op">-</span> <span class="va">start_time</span></span>
<span><span class="co">#&gt; Time difference of 27.71177 secs</span></span>
<span></span>
<span><span class="fu">pretty_CN_heatmap</span><span class="op">(</span><span class="va">cn_mat</span>,</span>
<span>  scale_by_sample <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  these_samples_metadata <span class="op">=</span> <span class="va">dna_meta</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: The input is a data frame-like object, convert it to a matrix.</span></span></code></pre></div>
<p><img src="Overview_files/figure-html/unnamed-chunk-5-1.png" width="700"></p>
<p>For the GISTIC strategy, it internally relies on
<code>gistic_to_cn_state_matrix</code>.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co">#| warning: false</span></span>
<span><span class="co">#| message: false</span></span>
<span></span>
<span></span>
<span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/jokergoo/ComplexHeatmap" class="external-link">ComplexHeatmap</a></span><span class="op">)</span></span>
<span><span class="va">cn_mat</span> <span class="op">&lt;-</span> <span class="fu">segmented_data_to_cn_matrix</span><span class="op">(</span></span>
<span>  seg_data <span class="op">=</span> <span class="va">all_segments</span>,</span>
<span>  strategy <span class="op">=</span> <span class="st">"GISTIC"</span>,</span>
<span>  these_samples_metadata <span class="op">=</span> <span class="va">dna_meta</span>,</span>
<span>  adjust_for_ploidy <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  gistic_lesions_file <span class="op">=</span> <span class="st">"all_lesions.DLBCL.grch37.txt"</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "0 rows with NA values"</span></span>
<span><span class="co">#&gt; [1] "0 rows with NA values"</span></span>
<span></span>
<span></span>
<span><span class="fu">pretty_CN_heatmap</span><span class="op">(</span>cn_state_matrix <span class="op">=</span> <span class="va">cn_mat</span>,</span>
<span>  scale_by_sample <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  these_samples_metadata <span class="op">=</span> <span class="va">dna_meta</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: The input is a data frame-like object, convert it to a matrix.</span></span></code></pre></div>
<p><img src="Overview_files/figure-html/unnamed-chunk-6-1.png" width="700"></p>
<p>This can also be used directly if you want to sanity check the
results.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co">#| warning: false</span></span>
<span><span class="co">#| message: false</span></span>
<span></span>
<span></span>
<span><span class="va">gistic_wide</span> <span class="op">&lt;-</span> <span class="fu">gistic_to_cn_state_matrix</span><span class="op">(</span></span>
<span>  seg_data <span class="op">=</span> <span class="va">all_segments</span>,</span>
<span>  gistic_lesions_file <span class="op">=</span> <span class="st">"all_lesions.DLBCL.grch37.txt"</span>,</span>
<span>  wide_peaks <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  these_samples_metadata <span class="op">=</span> <span class="va">dna_meta</span>,</span>
<span>  scale_by_sample <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  fill_missing_with <span class="op">=</span> <span class="st">"diploid"</span> <span class="co"># Note: this needs to be set to diploid for now</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: The input is a data frame-like object, convert it to a matrix.</span></span>
<span><span class="co">#&gt; [1] "0 rows with NA values"</span></span>
<span><span class="co">#&gt; Warning: The input is a data frame-like object, convert it to a matrix.</span></span>
<span></span>
<span><span class="va">h_saved</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Heatmap.html" class="external-link">Heatmap</a></span><span class="op">(</span><span class="va">cn_mat</span>,</span>
<span>  show_row_names <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  column_names_gp <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/gpar.html" class="external-link">gpar</a></span><span class="op">(</span>fontsize <span class="op">=</span> <span class="fl">5</span><span class="op">)</span>,</span>
<span>  clustering_distance_columns <span class="op">=</span> <span class="st">"binary"</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: The input is a data frame-like object, convert it to a matrix.</span></span>
<span><span class="va">h_saved</span></span></code></pre></div>
<p><img src="Overview_files/figure-html/unnamed-chunk-7-1.png" width="700"></p>
<p>Another visualization for GISTIC outputs:
<code>prettyChromoplot</code></p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co">#| warning: false</span></span>
<span><span class="co">#| message: false</span></span>
<span></span>
<span><span class="va">genes</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"MYC"</span>,<span class="st">"FCGR2B"</span>,<span class="st">"TNFRSF14"</span>,<span class="st">"FAS"</span>,<span class="st">"PTEN"</span>,<span class="st">"B2M"</span>,<span class="st">"RB1"</span>,<span class="st">"TCL1A"</span>,<span class="st">"CD70"</span>,</span>
<span>  <span class="st">"BCL2"</span>,<span class="st">"KLHL14"</span>,<span class="st">"TCF4"</span>,<span class="st">"REL"</span>,<span class="st">"BCL6"</span>,<span class="st">"HIST1H1C"</span>,<span class="st">"SMARCA4"</span>,<span class="st">"CDKN2A"</span>,<span class="st">"RHOA"</span>,</span>
<span>  <span class="st">"TNFAIP3"</span>,<span class="st">"TP53"</span>,<span class="st">"CDK14"</span>,<span class="st">"RELN"</span>,<span class="st">"ETS1"</span>,<span class="st">"MDM1"</span>,<span class="st">"MIR17HG"</span>,<span class="st">"CD58"</span>,<span class="st">"HNRNPD"</span>,</span>
<span>  <span class="st">"TOX"</span>,<span class="st">"PRAME"</span>,<span class="st">"CD38"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">gene_bed</span> <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">grch37_gene_coordinates</span>,<span class="op">-</span><span class="fl">1</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="co">#remove ensembl ID column</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">hugo_symbol</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">genes</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="co">#keep genes of interest</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>length <span class="op">=</span> <span class="va">end</span> <span class="op">-</span> <span class="va">start</span>,mid <span class="op">=</span> <span class="va">start</span> <span class="op">+</span> <span class="va">length</span><span class="op">/</span><span class="fl">2</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>start <span class="op">=</span> <span class="va">mid</span>,end<span class="op">=</span><span class="va">start</span><span class="op">+</span><span class="fl">1</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">create_bed_data</span><span class="op">(</span>genome_build <span class="op">=</span> <span class="st">"grch37"</span><span class="op">)</span> <span class="co">#convert to bed_data format</span></span>
<span></span>
<span><span class="fu">prettyChromoplot</span><span class="op">(</span>scores_path <span class="op">=</span> <span class="st">"scores.gistic"</span>,</span>
<span>                 labels_bed <span class="op">=</span> <span class="va">gene_bed</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="font-weight: bold;">Rows: </span><span style="color: #0000BB;">31533</span> <span style="font-weight: bold;">Columns: </span><span style="color: #0000BB;">8</span></span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">──</span> <span style="font-weight: bold;">Column specification</span> <span style="color: #00BBBB;">────────────────────────────────────────────────────────</span></span></span>
<span><span class="co">#&gt; <span style="font-weight: bold;">Delimiter:</span> "\t"</span></span>
<span><span class="co">#&gt; <span style="color: #BB0000;">chr</span> (1): Type</span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">dbl</span> (7): Chromosome, Start, End, -log10(q-value), G-score, average amplitude...</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Use `spec()` to retrieve the full column specification for this data.</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Specify the column types or set `show_col_types = FALSE` to quiet this message.</span></span>
<span><span class="co">#&gt; Running in default mode of any...</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 31,533 × 15</span></span></span>
<span><span class="co">#&gt;    Chromosome  Start    End `-log10(q-value)` `G-score` `average amplitude`</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>             <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>               <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> 1           <span style="text-decoration: underline;">10</span>001 <span style="text-decoration: underline;">768</span>252                 0    0.065<span style="text-decoration: underline;">0</span>               0.296</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> 1          <span style="text-decoration: underline;">768</span>253 <span style="text-decoration: underline;">768</span>447                 0    0.068<span style="text-decoration: underline;">5</span>               0.330</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> 1          <span style="text-decoration: underline;">768</span>448 <span style="text-decoration: underline;">769</span>962                 0    0.068<span style="text-decoration: underline;">0</span>               0.332</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> 1          <span style="text-decoration: underline;">769</span>963 <span style="text-decoration: underline;">770</span>180                 0    0.073<span style="text-decoration: underline;">8</span>               0.344</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> 1          <span style="text-decoration: underline;">770</span>181 <span style="text-decoration: underline;">770</span>734                 0    0.074<span style="text-decoration: underline;">7</span>               0.349</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> 1          <span style="text-decoration: underline;">770</span>735 <span style="text-decoration: underline;">771</span>656                 0    0.075<span style="text-decoration: underline;">1</span>               0.347</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> 1          <span style="text-decoration: underline;">771</span>657 <span style="text-decoration: underline;">773</span>105                 0    0.079<span style="text-decoration: underline;">3</span>               0.357</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> 1          <span style="text-decoration: underline;">773</span>106 <span style="text-decoration: underline;">776</span>545                 0    0.090<span style="text-decoration: underline;">4</span>               0.396</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> 1          <span style="text-decoration: underline;">776</span>546 <span style="text-decoration: underline;">783</span>070                 0    0.086<span style="text-decoration: underline;">4</span>               0.480</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> 1          <span style="text-decoration: underline;">783</span>071 <span style="text-decoration: underline;">785</span>521                 0    0.087<span style="text-decoration: underline;">4</span>               0.469</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 31,523 more rows</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 9 more variables: frequency &lt;dbl&gt;, Type &lt;chr&gt;, fill &lt;chr&gt;, start &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   end &lt;dbl&gt;, name &lt;chr&gt;, hugo_symbol &lt;chr&gt;, length &lt;dbl&gt;, mid &lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt;    chrom     start       end     name hugo_symbol length       mid</span></span>
<span><span class="co">#&gt; 1     13  92003453  92003454  MIR17HG     MIR17HG   6759  92003454</span></span>
<span><span class="co">#&gt; 2      6 138196400 138196401  TNFAIP3     TNFAIP3  16098 138196400</span></span>
<span><span class="co">#&gt; 3     14  96178418  96178419    TCL1A       TCL1A   4229  96178418</span></span>
<span><span class="co">#&gt; 4     10  90762978  90762979      FAS         FAS  25128  90762978</span></span>
<span><span class="co">#&gt; 5     18  53110790  53110791     TCF4        TCF4 442456  53110790</span></span>
<span><span class="co">#&gt; 6     12  68696192  68696193     MDM1        MDM1  59938  68696192</span></span>
<span><span class="co">#&gt; 7      9  21981525  21981526   CDKN2A      CDKN2A  27549  21981526</span></span>
<span><span class="co">#&gt; 8      6  26056333  26056334 HIST1H1C    HIST1H1C    731  26056334</span></span>
<span><span class="co">#&gt; 9     10  89677278  89677279     PTEN        PTEN 108817  89677278</span></span>
<span><span class="co">#&gt; 10    18  30302829  30302830   KLHL14      KLHL14 100391  30302830</span></span>
<span><span class="co">#&gt; 11    13  48967004  48967005      RB1         RB1 178235  48967004</span></span>
<span><span class="co">#&gt; 12    18  60888970  60888971     BCL2        BCL2 196782  60888970</span></span>
<span><span class="co">#&gt; 13    22  22895945  22895946    PRAME       PRAME  11645  22895946</span></span>
<span><span class="co">#&gt; 14     1   2491949   2491950 TNFRSF14    TNFRSF14   9743   2491950</span></span>
<span><span class="co">#&gt; 15    15  45007375  45007376      B2M         B2M   7400  45007375</span></span>
<span><span class="co">#&gt; 16    19  11123834  11123835  SMARCA4     SMARCA4 104473  11123834</span></span>
<span><span class="co">#&gt; 17     4  83284653  83284654   HNRNPD      HNRNPD  22005  83284654</span></span>
<span><span class="co">#&gt; 18     3 187451340 187451341     BCL6        BCL6  24350 187451340</span></span>
<span><span class="co">#&gt; 19     8  59874872  59874873      TOX         TOX 313790  59874872</span></span>
<span><span class="co">#&gt; 20     7  90467821  90467822    CDK14       CDK14 744167  90467822</span></span>
<span><span class="co">#&gt; 21     4  15817375  15817376     CD38        CD38  74955  15817376</span></span>
<span><span class="co">#&gt; 22    11 128393054 128393055     ETS1        ETS1 128797 128393054</span></span>
<span><span class="co">#&gt; 23     8 128750677 128750678      MYC         MYC   5994 128750677</span></span>
<span><span class="co">#&gt; 24     7 103371097 103371098     RELN        RELN 517732 103371097</span></span>
<span><span class="co">#&gt; 25     1 117085409 117085410     CD58        CD58  56504 117085409</span></span>
<span><span class="co">#&gt; 26    19   6593654   6593655     CD70        CD70  20920   6593654</span></span>
<span><span class="co">#&gt; 27     3  49423504  49423505     RHOA        RHOA  53853  49423504</span></span>
<span><span class="co">#&gt; 28    17   7577976   7577977     TP53        TP53  25759   7577976</span></span>
<span><span class="co">#&gt; 29     1 161599772 161599773   FCGR2B      FCGR2B  97343 161599772</span></span>
<span><span class="co">#&gt; 30     2  61133700  61133701      REL         REL  50089  61133700</span></span></code></pre></div>
<p><img src="Overview_files/figure-html/unnamed-chunk-8-1.png" width="700"></p>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Ryan Morin, Kostiantyn Dreval, Vladimir Souza.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

      </footer>
</div>






  </body>
</html>
