<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Annotate mutations target motif — annotate_ssm_motif_context • GAMBLR.results</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Annotate mutations target motif — annotate_ssm_motif_context"><meta name="description" content="Checks for the presence of mutations at a given motif"><meta property="og:description" content="Checks for the presence of mutations at a given motif"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">GAMBLR.results</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.3</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/GAMBLR-results.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Developers</h6></li>
    <li><a class="dropdown-item" href="../articles/testing.html">Testing GAMBLR packages</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Annotate mutations target motif</h1>

      <div class="d-none name"><code>annotate_ssm_motif_context.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Checks for the presence of mutations at a given motif</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">annotate_ssm_motif_context</span><span class="op">(</span></span>
<span>  <span class="va">maf</span>,</span>
<span>  motif <span class="op">=</span> <span class="st">"WRCY"</span>,</span>
<span>  index <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  <span class="va">genome_build</span>,</span>
<span>  <span class="va">fastaPath</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-maf">maf<a class="anchor" aria-label="anchor" href="#arg-maf"></a></dt>
<dd><p>MAF data frame (required columns: Reference_Allele, Chromosome, Start_Position, End_Position)</p></dd>


<dt id="arg-motif">motif<a class="anchor" aria-label="anchor" href="#arg-motif"></a></dt>
<dd><p>The motif sequence (default is WRCY)</p></dd>


<dt id="arg-index">index<a class="anchor" aria-label="anchor" href="#arg-index"></a></dt>
<dd><p>Position of the mutated allele in the motif</p></dd>


<dt id="arg-genome-build">genome_build<a class="anchor" aria-label="anchor" href="#arg-genome-build"></a></dt>
<dd><p>The genome build for the variants you are working with (default is grch37)</p></dd>


<dt id="arg-fastapath">fastaPath<a class="anchor" aria-label="anchor" href="#arg-fastapath"></a></dt>
<dd><p>Can be a path to a FASTA file</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A data frame with two extra columns (seq and motif).</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>In positions that reference allele has been mutated, it will capture (motif length - 1) before and (motif length + 1) alleles after the mutated position.
Then, it looks for the presence of motif in the captured sequence and check if the mutation has occurred in the indexed position, it will return
SITE and if the the motif is present, but the mutation is not in the indexed position, it will return MOTIF. In other cases, it will return FALSE.
NA will be shown if the mutation is an indel mutation.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">my_maf</span> <span class="op">&lt;-</span> <span class="fu"><a href="get_coding_ssm.html">get_coding_ssm</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="pipe.html">%&gt;%</a></span> </span></span>
<span class="r-in"><span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">Hugo_Symbol</span><span class="op">==</span><span class="st">"BCL2"</span><span class="op">)</span> <span class="op"><a href="pipe.html">%&gt;%</a></span></span></span>
<span class="r-in"><span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="va">Chromosome</span>,<span class="va">Start_Position</span>,<span class="va">Tumor_Sample_Barcode</span><span class="op">)</span> <span class="op"><a href="pipe.html">%&gt;%</a></span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>One or more parsing issues, call `problems()` on your data frame for details,</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> e.g.:</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   dat &lt;- vroom(...)</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   problems(dat)</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">annotated</span> <span class="op">=</span> <span class="fu">annotate_ssm_motif_context</span><span class="op">(</span>maf <span class="op">=</span> <span class="va">my_maf</span>,</span></span>
<span class="r-in"><span>                                       motif <span class="op">=</span> <span class="st">"WRCY"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">annotated</span>,<span class="fl">1</span>,<span class="fl">5</span>,<span class="fl">6</span>,<span class="fl">11</span>,<span class="fl">13</span>,<span class="fl">16</span>,<span class="va">seq</span>,<span class="va">WRCY</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> genomic_data Object</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Genome Build: grch37 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Showing first 10 rows:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   Hugo_Symbol Chromosome Start_Position Reference_Allele Tumor_Seq_Allele2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1        BCL2         18       60795859                C                 T</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2        BCL2         18       60795894                A                 T</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3        BCL2         18       60795911                A                 G</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4        BCL2         18       60795911                A                 G</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5        BCL2         18       60795947                C                 G</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6        BCL2         18       60985305              GTG               AAT</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   Tumor_Sample_Barcode       seq  WRCY</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1      CAR_126_PreCART   CTTCACT FALSE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2            08-13706T   GCAAGCT MOTIF</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3    00-15336_CLC01491   CCAAACT MOTIF</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4    00-15336_CLC02290   CCAAACT MOTIF</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5             FL1010T2   AATCAAA FALSE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6             SP192988 CAAGTGCAC FALSE</span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Ryan Morin, Kostiantyn Dreval, Vladimir Souza.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.9000.</p>
</div>

    </footer></div>





  </body></html>

