<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Annotate Hotspots. — annotate_hotspots • GAMBLR.results</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Annotate Hotspots. — annotate_hotspots"><meta name="description" content="Annotate MAF-like data frome with a hot_spot column indicating recurrent mutations."><meta property="og:description" content="Annotate MAF-like data frome with a hot_spot column indicating recurrent mutations."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">GAMBLR.results</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.3</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/GAMBLR-results.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Developers</h6></li>
    <li><a class="dropdown-item" href="../articles/testing.html">Testing GAMBLR packages</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Annotate Hotspots.</h1>

      <div class="d-none name"><code>annotate_hotspots.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Annotate MAF-like data frome with a hot_spot column indicating recurrent mutations.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">annotate_hotspots</span><span class="op">(</span></span>
<span>  <span class="va">mutation_maf</span>,</span>
<span>  recurrence_min <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  analysis_base <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"FL--DLBCL"</span>, <span class="st">"BL--DLBCL"</span><span class="op">)</span>,</span>
<span>  p_thresh <span class="op">=</span> <span class="fl">0.05</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-mutation-maf">mutation_maf<a class="anchor" aria-label="anchor" href="#arg-mutation-maf"></a></dt>
<dd><p>A data frame in MAF format.</p></dd>


<dt id="arg-recurrence-min">recurrence_min<a class="anchor" aria-label="anchor" href="#arg-recurrence-min"></a></dt>
<dd><p>minimum number of recurrences for mutation to be included, default is 5.</p></dd>


<dt id="arg-analysis-base">analysis_base<a class="anchor" aria-label="anchor" href="#arg-analysis-base"></a></dt>
<dd><p>Base name for hot spot output directory.</p></dd>


<dt id="arg-p-thresh">p_thresh<a class="anchor" aria-label="anchor" href="#arg-p-thresh"></a></dt>
<dd><p>P value threshold, default is 0.05.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>The same data frame with one additional column "hot_spot".</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>This function takes an already loaded MAF data frame with the <code>mutation_maf</code> parameter.
The user can then control the minimum number of recurrences for mutations to be included with <code>recurrance_min</code>,
The default is 5. <code>analysis_base</code> controls the base name go hotspot output directory.
Lastly, <code>p_thresh</code> sets the p value threshold, default is 0.05.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">my_metadata</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="fu"><a href="get_gambl_metadata.html">get_gambl_metadata</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># get a few SSMs to annotate</span></span></span>
<span class="r-in"><span><span class="va">some_coding_ssm</span> <span class="op">=</span> <span class="fu"><a href="get_coding_ssm.html">get_coding_ssm</a></span><span class="op">(</span>these_samples_metadata <span class="op">=</span> <span class="va">my_metadata</span>,</span></span>
<span class="r-in"><span>                                projection <span class="op">=</span> <span class="st">"grch37"</span>,</span></span>
<span class="r-in"><span>                                this_seq_type <span class="op">=</span> <span class="st">"genome"</span><span class="op">)</span> <span class="op"><a href="pipe.html">%&gt;%</a></span> </span></span>
<span class="r-in"><span>                  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">Hugo_Symbol</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"EZH2"</span>,<span class="st">"MEF2B"</span>,<span class="st">"MYD88"</span>,<span class="st">"KMT2D"</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="pipe.html">%&gt;%</a></span></span></span>
<span class="r-in"><span>                  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="va">Hugo_Symbol</span><span class="op">)</span></span></span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>One or more parsing issues, call `problems()` on your data frame for details,</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> e.g.:</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   dat &lt;- vroom(...)</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   problems(dat)</span>
<span class="r-in"><span><span class="co"># peek at the data</span></span></span>
<span class="r-in"><span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">some_coding_ssm</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">10</span>,<span class="fl">37</span><span class="op">)</span> <span class="op"><a href="pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> genomic_data Object</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Genome Build: grch37 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Showing first 10 rows:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   Hugo_Symbol Entrez_Gene_Id Center NCBI_Build Chromosome Start_Position</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1        EZH2              0      .     GRCh37          7      148504773</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2        EZH2              0      .     GRCh37          7      148504791</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3        EZH2              0      .     GRCh37          7      148504802</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4        EZH2              0      .     GRCh37          7      148506215</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5        EZH2              0      .     GRCh37          7      148506437</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6        EZH2              0      .     GRCh37          7      148506437</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   End_Position Strand Variant_Classification Variant_Type HGVSp_Short</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1    148504773      +      Missense_Mutation          SNP     p.Y741H</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2    148504791      +      Missense_Mutation          SNP     p.Q735K</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3    148504802      +          Splice_Region          SNP        &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4    148506215      +      Missense_Mutation          SNP     p.I715F</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5    148506437      +      Missense_Mutation          SNP     p.A692V</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6    148506437      +      Missense_Mutation          SNP     p.A692V</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">hot_ssms</span> <span class="op">=</span> <span class="fu">annotate_hotspots</span><span class="op">(</span><span class="va">some_coding_ssm</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Adding missing grouping variables: `SYMBOL`</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Adding missing grouping variables: `SYMBOL`</span>
<span class="r-in"><span><span class="va">hot_ssms</span> <span class="op"><a href="pipe.html">%&gt;%</a></span> </span></span>
<span class="r-in"><span>   <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">hot_spot</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="pipe.html">%&gt;%</a></span> </span></span>
<span class="r-in"><span>   <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span>,<span class="fl">37</span>,<span class="va">hot_spot</span><span class="op">)</span> </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> genomic_data Object</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Genome Build: grch37 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Showing first 10 rows:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    Hugo_Symbol Entrez_Gene_Id Center NCBI_Build Chromosome Start_Position</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1         EZH2              0      .     GRCh37          7      148508727</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2         EZH2              0      .     GRCh37          7      148508727</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3         EZH2              0      .     GRCh37          7      148508727</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4         EZH2              0      .     GRCh37          7      148508727</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5         EZH2              0      .     GRCh37          7      148508727</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6         EZH2              0      .     GRCh37          7      148508727</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 7         EZH2              0      .     GRCh37          7      148508727</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 8         EZH2              0      .     GRCh37          7      148508727</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 9         EZH2              0      .     GRCh37          7      148508727</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 10        EZH2              0      .     GRCh37          7      148508727</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    End_Position Strand Variant_Classification Variant_Type HGVSp_Short hot_spot</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1     148508727      +      Missense_Mutation          SNP     p.Y646F     TRUE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2     148508727      +      Missense_Mutation          SNP     p.Y646F     TRUE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3     148508727      +      Missense_Mutation          SNP     p.Y646F     TRUE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4     148508727      +      Missense_Mutation          SNP     p.Y646F     TRUE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5     148508727      +      Missense_Mutation          SNP     p.Y646F     TRUE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6     148508727      +      Missense_Mutation          SNP     p.Y646F     TRUE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 7     148508727      +      Missense_Mutation          SNP     p.Y646F     TRUE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 8     148508727      +      Missense_Mutation          SNP     p.Y646F     TRUE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 9     148508727      +      Missense_Mutation          SNP     p.Y646F     TRUE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 10    148508727      +      Missense_Mutation          SNP     p.Y646F     TRUE</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span> <span class="co"># \dontrun{</span></span></span>
<span class="r-in"><span><span class="co">#This example will raise an error due to the user supplying an unsupported genome build:</span></span></span>
<span class="r-in"><span><span class="va">more_coding_ssm</span> <span class="op">=</span> <span class="fu"><a href="get_coding_ssm.html">get_coding_ssm</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>                                these_samples_metadata <span class="op">=</span> <span class="va">my_metadata</span>,</span></span>
<span class="r-in"><span>                                projection <span class="op">=</span> <span class="st">"hg38"</span>,</span></span>
<span class="r-in"><span>                                this_seq_type <span class="op">=</span> <span class="st">"capture"</span><span class="op">)</span> <span class="op"><a href="pipe.html">%&gt;%</a></span> </span></span>
<span class="r-in"><span>                  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">Hugo_Symbol</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"EZH2"</span>,<span class="st">"MEF2B"</span>,<span class="st">"MYD88"</span>,<span class="st">"KMT2D"</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="pipe.html">%&gt;%</a></span></span></span>
<span class="r-in"><span>                  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="va">Hugo_Symbol</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># peek at the data</span></span></span>
<span class="r-in"><span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">more_coding_ssm</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">10</span>,<span class="fl">37</span><span class="op">)</span> <span class="op"><a href="pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">more_hot_ssms</span> <span class="op">=</span> <span class="fu">annotate_hotspots</span><span class="op">(</span><span class="va">more_coding_ssm</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">more_hot_ssms</span> <span class="op"><a href="pipe.html">%&gt;%</a></span> </span></span>
<span class="r-in"><span>   <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">hot_spot</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="pipe.html">%&gt;%</a></span> </span></span>
<span class="r-in"><span>   <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span>,<span class="fl">37</span>,<span class="va">hot_spot</span><span class="op">)</span> </span></span>
<span class="r-in"><span><span class="op">}</span> <span class="co"># }</span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Ryan Morin, Kostiantyn Dreval, Vladimir Souza.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.9000.</p>
</div>

    </footer></div>





  </body></html>

