<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Annotate MAF with triplet context — annotate_maf_triplet • GAMBLR.results</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Annotate MAF with triplet context — annotate_maf_triplet"><meta name="description" content="Give triple sequence of mutated base with its adjacent bases (-1
and +1)"><meta property="og:description" content="Give triple sequence of mutated base with its adjacent bases (-1
and +1)"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">GAMBLR.results</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.3</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/GAMBLR-results.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Developers</h6></li>
    <li><a class="dropdown-item" href="../articles/testing.html">Testing GAMBLR packages</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Annotate MAF with triplet context</h1>

      <div class="d-none name"><code>annotate_maf_triplet.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Give triple sequence of mutated base with its adjacent bases (-1
and +1)</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">annotate_maf_triplet</span><span class="op">(</span></span>
<span>  <span class="va">maf</span>,</span>
<span>  all_SNVs <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  <span class="va">ref</span>,</span>
<span>  <span class="va">alt</span>,</span>
<span>  <span class="va">genome_build</span>,</span>
<span>  <span class="va">fastaPath</span>,</span>
<span>  pyrimidine_collapse <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-maf">maf<a class="anchor" aria-label="anchor" href="#arg-maf"></a></dt>
<dd><p>MAF file (required columns: Reference_Allele, Tumor_Seq_Allele2)</p></dd>


<dt id="arg-all-snvs">all_SNVs<a class="anchor" aria-label="anchor" href="#arg-all-snvs"></a></dt>
<dd><p>To give us all the triplet sequences of SNVs and not
specifying any specific ref and alt alleles (default is TRUE)</p></dd>


<dt id="arg-ref">ref<a class="anchor" aria-label="anchor" href="#arg-ref"></a></dt>
<dd><p>Reference allele</p></dd>


<dt id="arg-alt">alt<a class="anchor" aria-label="anchor" href="#arg-alt"></a></dt>
<dd><p>Alternative allele</p></dd>


<dt id="arg-genome-build">genome_build<a class="anchor" aria-label="anchor" href="#arg-genome-build"></a></dt>
<dd><p>The genome build for the variants you are
working with (default is to infer it from the MAF)</p></dd>


<dt id="arg-fastapath">fastaPath<a class="anchor" aria-label="anchor" href="#arg-fastapath"></a></dt>
<dd><p>Can be a path to a FASTA file on a disk. When on GSC,
this is first attempted to be inferred from the gambl reference through
path specified in config. Local files are also accepted as value here.</p></dd>


<dt id="arg-pyrimidine-collapse">pyrimidine_collapse<a class="anchor" aria-label="anchor" href="#arg-pyrimidine-collapse"></a></dt>
<dd><p>Estimate mutation_strand and</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A data frame with two to three extra columns, in case
pyrimidine_collapse = FALSE, it will add triple sequence (seq) and the
strand (mutation_strand). When pyrimidine_collapse = T, another column
also will be added to the those extra columns which shows the mutation
(mutation)</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>It gives the reference and alternative alleles and looks for the
rows of the data frame based on these values for + strand genes and their
complement alleles rows for - strand genes, then it can look for the adjacent
bases in that mutation position. Also, it can look for all the SNVs in
the MAF data frame and provide triple sequences for them (reverse complement
sequence for the - strand).</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">maf</span> <span class="op">&lt;-</span> <span class="fu"><a href="get_coding_ssm.html">get_coding_ssm</a></span><span class="op">(</span>projection <span class="op">=</span> <span class="st">"grch37"</span><span class="op">)</span> <span class="op"><a href="pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">500</span><span class="op">)</span></span></span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>One or more parsing issues, call `problems()` on your data frame for details,</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> e.g.:</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   dat &lt;- vroom(...)</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   problems(dat)</span>
<span class="r-in"><span><span class="co"># peek at the data</span></span></span>
<span class="r-in"><span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">maf</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">12</span><span class="op">)</span> <span class="op"><a href="pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> genomic_data Object</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Genome Build: grch37 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Showing first 10 rows:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     Hugo_Symbol Entrez_Gene_Id Center NCBI_Build Chromosome Start_Position</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1    AL627309.1              0      .     GRCh37          1         138626</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2    AL627309.1              0      .     GRCh37          1         138972</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3 RP11-206L10.9              0      .     GRCh37          1         730845</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4        FAM87B              0      .     GRCh37          1         753589</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5        SAMD11              0      .     GRCh37          1         871158</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6        SAMD11              0      .     GRCh37          1         871192</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   End_Position Strand Variant_Classification Variant_Type Reference_Allele</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1       138626      +                 Silent          SNP                T</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2       138973      +        Frame_Shift_Ins          INS                -</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3       730845      +          Splice_Region          SNP                G</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4       753589      +          Splice_Region          SNP                A</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5       871158      +                 Silent          SNP                C</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6       871192      +      Missense_Mutation          SNP                C</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   Tumor_Seq_Allele1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1                 T</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2                 -</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3                 G</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4                 A</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5                 C</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6                 C</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">maf_anno</span> <span class="op">&lt;-</span> <span class="fu">annotate_maf_triplet</span><span class="op">(</span><span class="va">maf</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">maf_anno</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">12</span>, <span class="va">seq</span><span class="op">)</span> <span class="op"><a href="pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> genomic_data Object</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Genome Build: grch37 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Showing first 10 rows:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     Hugo_Symbol Entrez_Gene_Id Center NCBI_Build Chromosome Start_Position</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1    AL627309.1              0      .     GRCh37          1         138626</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2    AL627309.1              0      .     GRCh37          1         138972</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3 RP11-206L10.9              0      .     GRCh37          1         730845</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4        FAM87B              0      .     GRCh37          1         753589</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5        SAMD11              0      .     GRCh37          1         871158</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6        SAMD11              0      .     GRCh37          1         871192</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   End_Position Strand Variant_Classification Variant_Type Reference_Allele</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1       138626      +                 Silent          SNP                T</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2       138973      +        Frame_Shift_Ins          INS                -</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3       730845      +          Splice_Region          SNP                G</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4       753589      +          Splice_Region          SNP                A</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5       871158      +                 Silent          SNP                C</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6       871192      +      Missense_Mutation          SNP                C</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   Tumor_Seq_Allele1 seq</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1                 T ATG</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2                 -  NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3                 G CGT</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4                 A AAA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5                 C ACG</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6                 C CCG</span>
<span class="r-in"><span><span class="co"># Each mutation is now associated with it's sequence context in the</span></span></span>
<span class="r-in"><span><span class="co"># reference genome in a column named seq</span></span></span>
<span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span> <span class="co"># \dontrun{</span></span></span>
<span class="r-in"><span><span class="fu">annotate_maf_triplet</span><span class="op">(</span><span class="va">maf</span>, all_SNVs <span class="op">=</span> <span class="cn">FALSE</span>, <span class="st">"C"</span>, <span class="st">"T"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">annotate_maf_triplet</span><span class="op">(</span><span class="va">maf</span>, ref <span class="op">=</span> <span class="st">"C"</span>, alt <span class="op">=</span> <span class="st">"T"</span>, pyrimidine_collapse <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span> <span class="co"># }</span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Ryan Morin, Kostiantyn Dreval, Vladimir Souza.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.9000.</p>
</div>

    </footer></div>





  </body></html>

