<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Annotate mutations target motif — annotate_ssm_motif_context • GAMBLR.results</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Annotate mutations target motif — annotate_ssm_motif_context"><meta property="og:description" content="Checks for the presence of mutations at a given motif"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">GAMBLR.results</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/Overview.html">What's new in GAMBLR</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Annotate mutations target motif</h1>

    <div class="hidden name"><code>annotate_ssm_motif_context.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Checks for the presence of mutations at a given motif</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">annotate_ssm_motif_context</span><span class="op">(</span></span>
<span>  <span class="va">maf</span>,</span>
<span>  motif <span class="op">=</span> <span class="st">"WRCY"</span>,</span>
<span>  index <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  <span class="va">genome_build</span>,</span>
<span>  <span class="va">fastaPath</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>


<dl><dt id="arg-maf">maf<a class="anchor" aria-label="anchor" href="#arg-maf"></a></dt>
<dd><p>MAF data frame (required columns: Reference_Allele, Chromosome, Start_Position, End_Position)</p></dd>


<dt id="arg-motif">motif<a class="anchor" aria-label="anchor" href="#arg-motif"></a></dt>
<dd><p>The motif sequence (default is WRCY)</p></dd>


<dt id="arg-index">index<a class="anchor" aria-label="anchor" href="#arg-index"></a></dt>
<dd><p>Position of the mutated allele in the motif</p></dd>


<dt id="arg-genome-build">genome_build<a class="anchor" aria-label="anchor" href="#arg-genome-build"></a></dt>
<dd><p>The genome build for the variants you are working with (default is grch37)</p></dd>


<dt id="arg-fastapath">fastaPath<a class="anchor" aria-label="anchor" href="#arg-fastapath"></a></dt>
<dd><p>Can be a path to a FASTA file</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>A data frame with two extra columns (seq and motif).</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>In positions that reference allele has been mutated, it will capture (motif length - 1) before and (motif length + 1) alleles after the mutated position.
Then, it looks for the presence of motif in the captured sequence and check if the mutation has occurred in the indexed position, it will return
SITE and if the the motif is present, but the mutation is not in the indexed position, it will return MOTIF. In other cases, it will return FALSE.
NA will be shown if the mutation is an indel mutation.</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">my_maf</span> <span class="op">&lt;-</span> <span class="fu"><a href="get_coding_ssm.html">get_coding_ssm</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>One or more parsing issues, call `problems()` on your data frame for details,</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> e.g.:</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   dat &lt;- vroom(...)</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   problems(dat)</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">annotated</span> <span class="op">=</span> <span class="fu">annotate_ssm_motif_context</span><span class="op">(</span>maf <span class="op">=</span> <span class="va">my_maf</span>, motif <span class="op">=</span> <span class="st">"WRCY"</span>, index <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">annotated</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> genomic_data Object</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Genome Build: grch37 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Showing first 10 rows:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     Hugo_Symbol Entrez_Gene_Id Center NCBI_Build Chromosome Start_Position</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1    AL627309.1              0      .     GRCh37          1         138626</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2    AL627309.1              0      .     GRCh37          1         138972</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3 RP11-206L10.9              0      .     GRCh37          1         730845</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4        FAM87B              0      .     GRCh37          1         753589</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5        SAMD11              0      .     GRCh37          1         871158</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6        SAMD11              0      .     GRCh37          1         871192</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   End_Position Strand Variant_Classification Variant_Type Reference_Allele</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1       138626      +                 Silent          SNP                T</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2       138973      +        Frame_Shift_Ins          INS                -</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3       730845      +          Splice_Region          SNP                G</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4       753589      +          Splice_Region          SNP                A</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5       871158      +                 Silent          SNP                C</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6       871192      +      Missense_Mutation          SNP                C</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   Tumor_Seq_Allele1 Tumor_Seq_Allele2    dbSNP_RS dbSNP_Val_Status</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1                 T                 C       novel               NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2                 -                GC rs760414759               NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3                 G                 A rs368628623               NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4                 A                 C       novel               NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5                 C                 T rs770660212               NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6                 C                 T rs548621473               NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>        Tumor_Sample_Barcode Matched_Norm_Sample_Barcode Match_Norm_Seq_Allele1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1                 09-25670T                   09-25670N                      T</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2 BLGSP-71-30-00673-01A-01E   BLGSP-71-06-00286-99A-01D                      -</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3                  SP116668                    SP116667                      G</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4                      EB-3   BLGSP-71-06-00286-99A-01D                      A</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5                 05-33344T                   14-11247N                      C</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6 BLGSP-71-30-00662-01A-01E   BLGSP-71-06-00286-99A-01D                      C</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   Match_Norm_Seq_Allele2 Tumor_Validation_Allele1 Tumor_Validation_Allele2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1                      T                       NA                       NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2                      -                       NA                       NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3                      G                       NA                       NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4                      A                       NA                       NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5                      C                       NA                       NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6                      C                       NA                       NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   Match_Norm_Validation_Allele1 Match_Norm_Validation_Allele2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1                            NA                            NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2                            NA                            NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3                            NA                            NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4                            NA                            NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5                            NA                            NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6                            NA                            NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   Verification_Status Validation_Status Mutation_Status Sequencing_Phase</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1                  NA                NA              NA               NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2                  NA                NA              NA               NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3                  NA                NA              NA               NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4                  NA                NA              NA               NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5                  NA                NA              NA               NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6                  NA                NA              NA               NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   Sequence_Source Validation_Method Score BAM_File Sequencer Tumor_Sample_UUID</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1              NA                NA    NA       NA        NA                NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2              NA                NA    NA       NA        NA                NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3              NA                NA    NA       NA        NA                NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4              NA                NA    NA       NA        NA                NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5              NA                NA    NA       NA        NA                NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6              NA                NA    NA       NA        NA                NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   Matched_Norm_Sample_UUID          HGVSc             HGVSp HGVSp_Short</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1                       NA       c.684A&gt;G               p.=     p.S228=</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2                       NA c.337_338insGC p.Pro113ArgfsTer3 p.P113Rfs*3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3                       NA     n.259-5G&gt;A              &lt;NA&gt;        &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4                       NA     n.832+7N&gt;C              &lt;NA&gt;        &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5                       NA       c.312C&gt;T               p.=     p.D104=</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6                       NA       c.346N&gt;T       p.Arg116Cys     p.R116C</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     Transcript_ID Exon_Number t_depth t_ref_count t_alt_count n_depth</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1 ENST00000423372         1/2      69          48          21      35</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2 ENST00000423372         1/2      52          31          21      36</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3 ENST00000586288        &lt;NA&gt;      31          21          10      23</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4 ENST00000326734        &lt;NA&gt;      97          47          50      35</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5 ENST00000342066        5/14      93          42          51      16</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6 ENST00000342066        5/14      20          16           4      51</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   n_ref_count n_alt_count maf_seq_type     seq  WRCY</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1          35           0       genome CCATGAG FALSE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2          36           0       genome      NA    NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3          23           0       genome TTCGTTA FALSE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4          35           0       genome AGAAATT FALSE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5          16           0       genome GGACGGC FALSE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6          51           0       genome CACCGCA FALSE</span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Ryan Morin, Kostiantyn Dreval, Vladimir Souza.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

      </footer></div>






  </body></html>

