<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Estimate Purity. — estimate_purity • GAMBLR.results</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Estimate Purity. — estimate_purity"><meta name="description" content="Annotate a MAF with segmented absolute copy number data and added additional columns (VAF, Ploidy and Final_purity)."><meta property="og:description" content="Annotate a MAF with segmented absolute copy number data and added additional columns (VAF, Ploidy and Final_purity)."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">GAMBLR.results</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.3</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/GAMBLR-results.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Developers</h6></li>
    <li><a class="dropdown-item" href="../articles/testing.html">Testing GAMBLR packages</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Estimate Purity.</h1>

      <div class="d-none name"><code>estimate_purity.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Annotate a MAF with segmented absolute copy number data and added additional columns (VAF, Ploidy and Final_purity).</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">estimate_purity</span><span class="op">(</span></span>
<span>  <span class="va">these_samples_metadata</span>,</span>
<span>  <span class="va">maf_data</span>,</span>
<span>  <span class="va">seg_data</span>,</span>
<span>  show_plots <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  assume_diploid <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  coding_only <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  <span class="va">projection</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-these-samples-metadata">these_samples_metadata<a class="anchor" aria-label="anchor" href="#arg-these-samples-metadata"></a></dt>
<dd><p>Metadata for one sample</p></dd>


<dt id="arg-maf-data">maf_data<a class="anchor" aria-label="anchor" href="#arg-maf-data"></a></dt>
<dd><p>Optional. Instead of using the path to a maf file, use a local dataframe as the maf file.</p></dd>


<dt id="arg-seg-data">seg_data<a class="anchor" aria-label="anchor" href="#arg-seg-data"></a></dt>
<dd><p>Data frame or seg_data object for the sample of interest.
Can contain data from other samples, which will be ignored.</p></dd>


<dt id="arg-show-plots">show_plots<a class="anchor" aria-label="anchor" href="#arg-show-plots"></a></dt>
<dd><p>Optional. Show two faceted plots that display the VAF and purity distributions for each copy number state in the sample. Default is FALSE.</p></dd>


<dt id="arg-assume-diploid">assume_diploid<a class="anchor" aria-label="anchor" href="#arg-assume-diploid"></a></dt>
<dd><p>Optional. If no local seg file is provided, instead of defaulting to a GAMBL sample, this parameter annotates every mutation as copy neutral. Default is FALSE.</p></dd>


<dt id="arg-coding-only">coding_only<a class="anchor" aria-label="anchor" href="#arg-coding-only"></a></dt>
<dd><p>Optional. set to TRUE to restrict to only coding variants. Default is FALSE.</p></dd>


<dt id="arg-in-maf">in_maf<a class="anchor" aria-label="anchor" href="#arg-in-maf"></a></dt>
<dd><p>Path to a local maf file.</p></dd>


<dt id="arg-in-seg">in_seg<a class="anchor" aria-label="anchor" href="#arg-in-seg"></a></dt>
<dd><p>Path to a local corresponding seg file for the same sample ID as the input maf.</p></dd>


<dt id="arg-this-seq-type">this_seq_type<a class="anchor" aria-label="anchor" href="#arg-this-seq-type"></a></dt>
<dd><p>Seq type for returned CN segments. One of "genome" (default) or "capture".</p></dd>


<dt id="arg-verbose-">verbose.<a class="anchor" aria-label="anchor" href="#arg-verbose-"></a></dt>
<dd><p>Set to TRUE for more feedback.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A list containing a data frame (MAF-like format) with the segmented absolute copy number data and three extra columns:
VAF is the variant allele frequency calculated from the t_ref_count and t_alt_count
Ploidy is the number of copies of an allele in the tumour cell
Final_purity is the finalized purity estimation per mutation after considering different copy number states and LOH events.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>This function takes a single row of metadata that defines a
sample you wish to estimate the purity of.
The user can also use an already loaded maf file with <code>maf_df</code>. In addition, a path to the maf/seq file of interest can also be passed to this function with
<code>in_maf</code> and <code>in_seg</code>. To visualize VAF and purity distributions, set the <code>show_plots</code> to TRUE (default is FALSE).
For more information on how to run this function with the parameters at hand, refer to the parameter descriptions and function examples.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># get metadata for one sample </span></span></span>
<span class="r-in"><span><span class="va">my_meta</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="fu"><a href="get_gambl_metadata.html">get_gambl_metadata</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="pipe.html">%&gt;%</a></span> </span></span>
<span class="r-in"><span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">sample_id</span> <span class="op">==</span> <span class="st">"HTMCP-01-06-00422-01A-01D"</span>,</span></span>
<span class="r-in"><span>  <span class="va">seq_type</span> <span class="op">==</span> <span class="st">"genome"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">#estimate purity, allowing the data to be retrieved for you</span></span></span>
<span class="r-in"><span><span class="va">outputs</span> <span class="op">=</span> <span class="fu">estimate_purity</span><span class="op">(</span>these <span class="op">=</span> <span class="va">my_meta</span>,</span></span>
<span class="r-in"><span>                show_plots <span class="op">=</span> <span class="cn">TRUE</span>,</span></span>
<span class="r-in"><span>                projection  <span class="op">=</span> <span class="st">"grch37"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> dummy segments are not annotated in the inputs</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> fill_missing_with parameter will be ignored</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Running in default mode of any...</span>
<span class="r-in"><span><span class="va">outputs</span><span class="op">$</span><span class="va">sample_purity_estimation</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 1</span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Ryan Morin, Kostiantyn Dreval, Vladimir Souza.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.9000.</p>
</div>

    </footer></div>





  </body></html>

