<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Get Raw Expression Counts — get_raw_expression_counts • GAMBLR.results</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Get Raw Expression Counts — get_raw_expression_counts"><meta name="description" content="Get the raw read counts from RNA-seq for one or more genes for all GAMBL samples."><meta property="og:description" content="Get the raw read counts from RNA-seq for one or more genes for all GAMBL samples."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">GAMBLR.results</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.3</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/GAMBLR-results.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Developers</h6></li>
    <li><a class="dropdown-item" href="../articles/testing.html">Testing GAMBLR packages</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Get Raw Expression Counts</h1>

      <div class="d-none name"><code>get_raw_expression_counts.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Get the raw read counts from RNA-seq for one or more genes for all GAMBL samples.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">get_raw_expression_counts</span><span class="op">(</span></span>
<span>  <span class="va">these_samples_metadata</span>,</span>
<span>  <span class="va">existing_sample_id_column</span>,</span>
<span>  <span class="va">new_sample_id_column</span>,</span>
<span>  all_samples <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  check <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  from_flatfile <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  map_to_symbol <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  TPM <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-these-samples-metadata">these_samples_metadata<a class="anchor" aria-label="anchor" href="#arg-these-samples-metadata"></a></dt>
<dd><p>The data frame with sample metadata. Usually output of the get_gambl_metadata().</p></dd>


<dt id="arg-existing-sample-id-column">existing_sample_id_column<a class="anchor" aria-label="anchor" href="#arg-existing-sample-id-column"></a></dt>
<dd><p>Specify which column contains the sample_id you want replaced with the contents of new_sample_id_column</p></dd>


<dt id="arg-new-sample-id-column">new_sample_id_column<a class="anchor" aria-label="anchor" href="#arg-new-sample-id-column"></a></dt>
<dd><p>Specify which column in your metadata contains the sample_id you want used instead of the existing_sample_id</p></dd>


<dt id="arg-all-samples">all_samples<a class="anchor" aria-label="anchor" href="#arg-all-samples"></a></dt>
<dd><p>Set to TRUE to force the function to return all available data (should rarely be necessary)</p></dd>


<dt id="arg-check">check<a class="anchor" aria-label="anchor" href="#arg-check"></a></dt>
<dd><p>For basic debugging. Set to TRUE to obtain basic information about the number of samples in your metadata with expression data available</p></dd>


<dt id="arg-from-flatfile">from_flatfile<a class="anchor" aria-label="anchor" href="#arg-from-flatfile"></a></dt>
<dd><p>Set to FALSE to use the database instead of reading from flatfiles</p></dd>


<dt id="arg-map-to-symbol">map_to_symbol<a class="anchor" aria-label="anchor" href="#arg-map-to-symbol"></a></dt>
<dd><p>Set to TRUE to obtain the mappings between the rows in the count matrix and HGNC gene symbol/alias</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>Set to TRUE for more print statements and such</p></dd>


<dt id="arg-tpm">TPM<a class="anchor" aria-label="anchor" href="#arg-tpm"></a></dt>
<dd><p>Set to TRUE to get TPM estimates instead of counts</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A list containing a counts matrix and the associated metadata for DESeq2</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>Efficiently retrieve raw gene expression values (read counts) for one, multiple or all genes for all GAMBL samples.
For examples and more info, refer to the parameter descriptions as well as vignette examples.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span> <span class="co"># \dontrun{</span></span></span>
<span class="r-in"><span><span class="va">schmitz_meta</span> <span class="op">=</span> <span class="fu"><a href="get_gambl_metadata.html">get_gambl_metadata</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="pipe.html">%&gt;%</a></span> </span></span>
<span class="r-in"><span>    <span class="fu"><a href="https://rdrr.io/r/stats/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">seq_type</span><span class="op">==</span><span class="st">"mrna"</span>,<span class="va">cohort</span><span class="op">==</span><span class="st">"dlbcl_schmitz"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">exp_out</span> <span class="op">=</span> <span class="fu">get_raw_expression_counts</span><span class="op">(</span>these_samples_metadata <span class="op">=</span> <span class="va">schmitz_meta</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Create DESeq data set directly from the two named objects in the output</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">dds</span> <span class="op">&lt;-</span> <span class="fu">DESeqDataSetFromMatrix</span><span class="op">(</span>countData <span class="op">=</span> <span class="va">exp_out</span><span class="op">$</span><span class="va">counts</span>,</span></span>
<span class="r-in"><span>    colData <span class="op">=</span> <span class="va">exp_out</span><span class="op">$</span><span class="va">metadata</span>,</span></span>
<span class="r-in"><span>    design <span class="op">=</span> <span class="op">~</span> <span class="va">COO_consensus</span> <span class="op">+</span> <span class="va">sex</span><span class="op">)</span></span></span>
<span class="r-in"><span>    </span></span>
<span class="r-in"><span><span class="co"># Run a basic DESeq analysis</span></span></span>
<span class="r-in"><span><span class="va">dds</span> <span class="op">&lt;-</span> <span class="fu">DESeq</span><span class="op">(</span><span class="va">dds</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu">results</span><span class="op">(</span><span class="va">dds</span>, </span></span>
<span class="r-in"><span>    name<span class="op">=</span><span class="st">"COO_consensus_GCB_vs_ABC"</span>,</span></span>
<span class="r-in"><span>    lfcThreshold<span class="op">=</span><span class="fl">2</span>,alpha<span class="op">=</span><span class="fl">0.1</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># Filter outputs using padj, logFC and baseMean (more highly expressed overall)     </span></span></span>
<span class="r-in"><span><span class="va">res_df</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span> <span class="op"><a href="pipe.html">%&gt;%</a></span> </span></span>
<span class="r-in"><span>    <span class="fu"><a href="https://rdrr.io/r/stats/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">padj</span><span class="op">&lt;</span><span class="fl">0.1</span>,<span class="va">baseMean</span><span class="op">&gt;</span><span class="fl">500</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">show_genes</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">res_df</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">vsd</span> <span class="op">&lt;-</span> <span class="fu">vst</span><span class="op">(</span><span class="va">dds</span>, blind<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">#Visualize the results with a heatmap</span></span></span>
<span class="r-in"><span><span class="va">column_ha</span> <span class="op">=</span> <span class="fu">HeatmapAnnotation</span><span class="op">(</span>df<span class="op">=</span><span class="fu">select</span><span class="op">(</span><span class="va">exp_out</span><span class="op">$</span><span class="va">metadata</span>,<span class="va">COO_consensus</span>,<span class="va">sex</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">Heatmap</span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">vsd</span><span class="op">)</span><span class="op">[</span><span class="va">show_genes</span>,<span class="op">]</span>,</span></span>
<span class="r-in"><span>    row_names_gp <span class="op">=</span> <span class="fu">gpar</span><span class="op">(</span>fontsize<span class="op">=</span><span class="fl">5</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    bottom_annotation <span class="op">=</span> <span class="va">column_ha</span>,</span></span>
<span class="r-in"><span>    show_column_names <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span> <span class="co"># }</span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Ryan Morin, Kostiantyn Dreval, Vladimir Souza.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.9000.</p>
</div>

    </footer></div>





  </body></html>

