<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Get ASHM Count Matrix. — get_ashm_count_matrix • GAMBLR.results</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Get ASHM Count Matrix. — get_ashm_count_matrix"><meta name="description" content="Prepare a matrix with one row per sample and one column per
region using a set of hypermutated regions."><meta property="og:description" content="Prepare a matrix with one row per sample and one column per
region using a set of hypermutated regions."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">GAMBLR.results</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.3</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/GAMBLR-results.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Developers</h6></li>
    <li><a class="dropdown-item" href="../articles/testing.html">Testing GAMBLR packages</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Get ASHM Count Matrix.</h1>

      <div class="d-none name"><code>get_ashm_count_matrix.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Prepare a matrix with one row per sample and one column per
region using a set of hypermutated regions.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">get_ashm_count_matrix</span><span class="op">(</span></span>
<span>  <span class="va">regions_bed</span>,</span>
<span>  <span class="va">maf_data</span>,</span>
<span>  <span class="va">these_samples_metadata</span>,</span>
<span>  this_seq_type <span class="op">=</span> <span class="st">"genome"</span>,</span>
<span>  <span class="va">projection</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-regions-bed">regions_bed<a class="anchor" aria-label="anchor" href="#arg-regions-bed"></a></dt>
<dd><p>A bed file with one row for each region.
The first three columns in this file MUST contain the Chromosome, Start and End position</p></dd>


<dt id="arg-maf-data">maf_data<a class="anchor" aria-label="anchor" href="#arg-maf-data"></a></dt>
<dd><p>Optionally provide a data frame in the MAF format, otherwise
either GAMBLR.data or GAMBLR.results will be used.</p></dd>


<dt id="arg-these-samples-metadata">these_samples_metadata<a class="anchor" aria-label="anchor" href="#arg-these-samples-metadata"></a></dt>
<dd><p>This is used to complete your matrix. All GAMBL
samples of the specified seq_type will be used by default. Provide a data frame with at least
sample_id for all samples if you are using non-GAMBL data.</p></dd>


<dt id="arg-this-seq-type">this_seq_type<a class="anchor" aria-label="anchor" href="#arg-this-seq-type"></a></dt>
<dd><p>The seq_type to return results for. Must be a single value. Only used if no
metadata is provided with these_samples_metadata.</p></dd>


<dt id="arg-projection">projection<a class="anchor" aria-label="anchor" href="#arg-projection"></a></dt>
<dd><p>The genome build we are working with</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A data frame with a row for every sample in these_samples_metadata and a column for every region in regions_bed</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>Values are the number of mutations in that patient in the region.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span>  <span class="va">DLBCL_genome_meta</span> <span class="op">=</span> <span class="fu"><a href="get_gambl_metadata.html">get_gambl_metadata</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="pipe.html">%&gt;%</a></span> </span></span>
<span class="r-in"><span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">pathology</span><span class="op">==</span><span class="st">"DLBCL"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 3273 capture samples are missing a value for protocol. Assuming Exome.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 138 biopsies are missing from the biopsy metadata. This should be fixed!</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> affected cohorts:  DLBCL_LSARP_Trios,Ennishi_tapestri,SMZL_Strefford,cHL_Maura,MCL_Barcelona</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 110 biopsies with discrepancies in the pathology field. This should be fixed!</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 10 biopsies with discrepancies in the time_point field. This should be fixed!</span>
<span class="r-in"><span><span class="co">#get ashm regions</span></span></span>
<span class="r-in"><span><span class="va">some_regions</span> <span class="op">=</span> <span class="fu">GAMBLR.utils</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/GAMBLR.utils/man/create_bed_data.html" class="external-link">create_bed_data</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>                              <span class="fu">GAMBLR.data</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/GAMBLR.data/man/grch37_ashm_regions.html" class="external-link">grch37_ashm_regions</a></span>,</span></span>
<span class="r-in"><span>                              fix_names <span class="op">=</span> <span class="st">"concat"</span>,</span></span>
<span class="r-in"><span>                              concat_cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"gene"</span>,<span class="st">"region"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                              sep<span class="op">=</span><span class="st">"-"</span><span class="op">)</span> <span class="op"><a href="pipe.html">%&gt;%</a></span></span></span>
<span class="r-in"><span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grepl</a></span><span class="op">(</span><span class="st">"PAX5"</span>,<span class="va">name</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">pax5_matrix</span> <span class="op">&lt;-</span> <span class="fu">get_ashm_count_matrix</span><span class="op">(</span></span></span>
<span class="r-in"><span>     regions_bed <span class="op">=</span> <span class="va">some_regions</span>,</span></span>
<span class="r-in"><span>     this_seq_type <span class="op">=</span> <span class="st">"genome"</span>,</span></span>
<span class="r-in"><span>     these_samples_metadata <span class="op">=</span> <span class="va">DLBCL_genome_meta</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Streamlined is set to TRUE, this function will disregard anything specified with basic_columns</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> To return a MAF with standard 45 columns, set streamlioned = FALSE and basic_columns = TRUE</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> To return a maf with all (116) columns, set streamlined = FALSE and basic_columns = FALSE</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Joining with `by = join_by(sample_id, region_name)`</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">pax5_matrix</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                   PAX5-TSS-1 PAX5-distal-enhancer-1 PAX5-distal-enhancer-3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 00-14595_tumorC            2                      4                      3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 00-15201_tumorA            0                      0                      3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 00-15201_tumorB            0                      0                      0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 00-17960_CLC01670          0                      0                      0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> FL1015T2                   0                      1                      0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 00-23442_tumorB            0                      0                      2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                   PAX5-intron-1 PAX5-distal-enhancer-2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 00-14595_tumorC              12                      0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 00-15201_tumorA               7                      0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 00-15201_tumorB               1                      0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 00-17960_CLC01670            11                      2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> FL1015T2                      0                      0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 00-23442_tumorB               0                      0</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">colMeans</a></span><span class="op">(</span><span class="va">pax5_matrix</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>             PAX5-TSS-1 PAX5-distal-enhancer-1 PAX5-distal-enhancer-3 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>              0.4402619              0.5319149              1.2487725 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          PAX5-intron-1 PAX5-distal-enhancer-2 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>              1.9247136              0.6579378 </span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Ryan Morin, Kostiantyn Dreval, Vladimir Souza.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.9000.</p>
</div>

    </footer></div>





  </body></html>

