<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Get Coding SSM. — get_coding_ssm • GAMBLR.results</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Get Coding SSM. — get_coding_ssm"><meta property="og:description" content="Retrieve all coding SSMs from one seq_type in GAMBL
in MAF-like format."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">GAMBLR.results</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Get Coding SSM.</h1>
    
    <div class="hidden name"><code>get_coding_ssm.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Retrieve all coding SSMs from one seq_type in GAMBL
in MAF-like format.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">get_coding_ssm</span><span class="op">(</span></span>
<span>  these_samples_metadata <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  <span class="va">force_unmatched_samples</span>,</span>
<span>  projection <span class="op">=</span> <span class="st">"grch37"</span>,</span>
<span>  this_seq_type <span class="op">=</span> <span class="st">"genome"</span>,</span>
<span>  basic_columns <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  maf_cols <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  augmented <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  min_read_support <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  groups <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"gambl"</span>, <span class="st">"icgc_dart"</span><span class="op">)</span>,</span>
<span>  include_silent <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  <span class="va">engine</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  <span class="va">limit_cohort</span>,</span>
<span>  <span class="va">exclude_cohort</span>,</span>
<span>  <span class="va">limit_pathology</span>,</span>
<span>  <span class="va">limit_samples</span>,</span>
<span>  <span class="va">from_flatfile</span>,</span>
<span>  <span class="va">these_sample_ids</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>these_samples_metadata</dt>
<dd><p>Optional (but highly recommended) metadata
table to tell the function how to subset the data. Only sample_id in
this table with the matching seq_type will be in the output. Not all
samples may be in the output, though, if they are missing SSM results
or had no mutations detected.</p></dd>


<dt>force_unmatched_samples</dt>
<dd><p>Optional argument for forcing unmatched
samples, using <a href="get_ssm_by_samples.html">get_ssm_by_samples</a>.</p></dd>


<dt>projection</dt>
<dd><p>Reference genome build for the coordinates in the MAF file.
The default is grch37.</p></dd>


<dt>this_seq_type</dt>
<dd><p>The seq_type you want SSMs from, default is genome.</p></dd>


<dt>basic_columns</dt>
<dd><p>Basic columns refers to the first 45 standard MAF
columns. Set this to FALSE if you want all the available columns instead.</p></dd>


<dt>maf_cols</dt>
<dd><p>if basic_columns is set to FALSE, the user can specify
which columns to be returned within the MAF. This parameter can either
be a vector of indexes (integer) or a vector of characters (matching
columns in MAF).</p></dd>


<dt>augmented</dt>
<dd><p>Set to FALSE if you instead want the original MAF from
each sample for multi-sample patients instead of the augmented MAF</p></dd>


<dt>min_read_support</dt>
<dd><p>Only returns variants with at least this many
reads in t_alt_count (for cleaning up augmented MAFs)</p></dd>


<dt>groups</dt>
<dd><p>Deprecated. Use these_samples_metadata instead.</p></dd>


<dt>include_silent</dt>
<dd><p>Logical indicating whether to include
silent mutations in coding regions (i.e. synonymous). Default is TRUE.</p></dd>


<dt>engine</dt>
<dd><p>Deprecated. Ignored</p></dd>


<dt>verbose</dt>
<dd><p>Controls the "verboseness" of this function (and
internally called helpers).</p></dd>


<dt>limit_cohort</dt>
<dd><p>Deprecated. Use these_samples_metadata instead.</p></dd>


<dt>exclude_cohort</dt>
<dd><p>Deprecated. Use these_samples_metadata instead.</p></dd>


<dt>limit_pathology</dt>
<dd><p>Deprecated. Use these_samples_metadata instead.</p></dd>


<dt>limit_samples</dt>
<dd><p>Deprecated. Use these_samples_metadata instead.</p></dd>


<dt>from_flatfile</dt>
<dd><p>Deprecated. Ignored</p></dd>


<dt>these_sample_ids</dt>
<dd><p>Deprecated. Use these_samples_metadata instead.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>A data frame containing all the MAF data columns
(one row per mutation).</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>Effectively retrieve simple somatic mutations (SSM) results for
either capture or genome seq_type (but not both at once).
The resulting data frame will be a maf_data object, which tracks the
genome build (projection) for the variants and will have a
maf_seq_type column that tracks the origin seq_type each variant.
In most cases, users should be using the related function that
is able to obtain SSMs across both genome and capture seq_type:
<a href="get_all_coding_ssm.html">get_all_coding_ssm</a>
Is this function not what you are looking for? Try one of:
<a href="get_coding_ssm_status.html">get_coding_ssm_status</a>,
get_ssm_by_patients,
<a href="get_ssm_by_sample.html">get_ssm_by_sample</a>,
<a href="get_ssm_by_samples.html">get_ssm_by_samples</a>,
<a href="get_ssm_by_region.html">get_ssm_by_region</a>,
<a href="get_ssm_by_regions.html">get_ssm_by_regions</a></p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span></span></span>
<span class="r-in"><span>  <span class="co">#basic usage (defaults to genome seq_type)</span></span></span>
<span class="r-in"><span>  <span class="va">maf_genome</span> <span class="op">=</span> <span class="fu">get_coding_ssm</span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>One or more parsing issues, call `problems()` on your data frame for details,</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> e.g.:</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   dat &lt;- vroom(...)</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   problems(dat)</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">maf_genome</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 248754</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">maf_genome</span>,<span class="fl">1</span>,<span class="fl">4</span>,<span class="fl">5</span>,<span class="fl">6</span>,<span class="fl">9</span>,<span class="va">maf_seq_type</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> genomic_data Object</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Genome Build: grch37 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Showing first 10 rows:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      Hugo_Symbol NCBI_Build Chromosome Start_Position Variant_Classification</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1     AL627309.1     GRCh37          1         138626                 Silent</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2     AL627309.1     GRCh37          1         138972        Frame_Shift_Ins</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3  RP11-206L10.9     GRCh37          1         730845          Splice_Region</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4         FAM87B     GRCh37          1         753589          Splice_Region</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5         SAMD11     GRCh37          1         871158                 Silent</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6         SAMD11     GRCh37          1         871192      Missense_Mutation</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 7         SAMD11     GRCh37          1         874416          Splice_Region</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 8         SAMD11     GRCh37          1         874467      Missense_Mutation</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 9         SAMD11     GRCh37          1         874648          Splice_Region</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 10        SAMD11     GRCh37          1         874763      Missense_Mutation</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    maf_seq_type</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1        genome</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2        genome</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3        genome</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4        genome</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5        genome</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6        genome</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 7        genome</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 8        genome</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 9        genome</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 10       genome</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span>  <span class="va">maf_exome_hg38</span> <span class="op">=</span> <span class="fu">get_coding_ssm</span><span class="op">(</span>this_seq_type <span class="op">=</span> <span class="st">"capture"</span>,</span></span>
<span class="r-in"><span>                                  projection<span class="op">=</span><span class="st">"hg38"</span><span class="op">)</span> </span></span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>The following named parsers don't match the column names: GENE_PHENO, FILTER, flanking_bps, vcf_id, vcf_qual, gnomAD_AF, gnomAD_AFR_AF, gnomAD_AMR_AF, gnomAD_SAS_AF, vcf_pos, gnomADg_AF, blacklist_count</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>One or more parsing issues, call `problems()` on your data frame for details,</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> e.g.:</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   dat &lt;- vroom(...)</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   problems(dat)</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">maf_exome_hg38</span>,<span class="fl">1</span>,<span class="fl">4</span>,<span class="fl">5</span>,<span class="fl">6</span>,<span class="fl">9</span>,<span class="va">maf_seq_type</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> genomic_data Object</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Genome Build: hg38 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Showing first 10 rows:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    Hugo_Symbol NCBI_Build Chromosome Start_Position Variant_Classification</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1        OR4F5     GRCh38       chr1          69634      Missense_Mutation</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2        OR4F5     GRCh38       chr1          69644      Missense_Mutation</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3   FO538757.2     GRCh38       chr1         183189      Missense_Mutation</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4   FO538757.2     GRCh38       chr1         183937      Missense_Mutation</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5   FO538757.1     GRCh38       chr1         186356      Nonsense_Mutation</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6   FO538757.1     GRCh38       chr1         186385      Missense_Mutation</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 7   FO538757.1     GRCh38       chr1         186404      Missense_Mutation</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 8   FO538757.1     GRCh38       chr1         186440      Missense_Mutation</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 9   FO538757.1     GRCh38       chr1         186440      Missense_Mutation</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 10  FO538757.1     GRCh38       chr1         186475          Splice_Region</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    maf_seq_type</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1       capture</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2       capture</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3       capture</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4       capture</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5       capture</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6       capture</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 7       capture</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 8       capture</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 9       capture</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 10      capture</span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Ryan Morin, Kostiantyn Dreval, Vladimir Souza.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

