<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Get Manta SVs — get_manta_sv • GAMBLR.results</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Get Manta SVs — get_manta_sv"><meta property="og:description" content="Retrieve Manta SVs for one or many samples"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">GAMBLR.results</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Get Manta SVs</h1>
    
    <div class="hidden name"><code>get_manta_sv.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Retrieve Manta SVs for one or many samples</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">get_manta_sv</span><span class="op">(</span></span>
<span>  these_samples_metadata <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  projection <span class="op">=</span> <span class="st">"grch37"</span>,</span>
<span>  <span class="va">region</span>,</span>
<span>  min_vaf <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>  min_score <span class="op">=</span> <span class="fl">40</span>,</span>
<span>  pass_filters <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  from_cache <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  write_to_file <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  <span class="va">chromosome</span>,</span>
<span>  <span class="va">qstart</span>,</span>
<span>  <span class="va">qend</span>,</span>
<span>  these_sample_ids <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  <span class="va">pairing_status</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>these_samples_metadata</dt>
<dd><p>A metadata data frame to limit the
result to sample_ids within it</p></dd>


<dt>projection</dt>
<dd><p>The projection genome build. Default is grch37.</p></dd>


<dt>region</dt>
<dd><p>Specify a single region to fetch SVs anchored within
using the format "chrom:start-end"</p></dd>


<dt>min_vaf</dt>
<dd><p>The minimum tumour VAF for a SV to be returned.
Default is 0.1.</p></dd>


<dt>min_score</dt>
<dd><p>The lowest Manta somatic score for a SV to be returned.
Default is 40.</p></dd>


<dt>pass_filters</dt>
<dd><p>If TRUE (default) only return SVs that are annotated with
PASS in the FILTER column. Set to FALSE to keep all variants,
regardless if they PASS the filters.</p></dd>


<dt>verbose</dt>
<dd><p>Set to FALSE to minimize the output to console.
Default is TRUE. This parameter also dictates the verbose-ness of
any helper function internally called inside the main function.</p></dd>


<dt>from_cache</dt>
<dd><p>Boolean variable for using cached results, default is TRUE.
If <code>write_to_file = TRUE</code>, this parameter auto-defaults to FALSE.</p></dd>


<dt>write_to_file</dt>
<dd><p>Boolean statement that outputs bedpe file if TRUE,
default is FALSE.
Setting this to TRUE forces <code>from_cache = FALSE</code>.</p></dd>


<dt>chromosome</dt>
<dd><p>DEPRECATED. Use <code>region</code> instead.</p></dd>


<dt>qstart</dt>
<dd><p>DEPRECATED. Use <code>region</code> instead.</p></dd>


<dt>qend</dt>
<dd><p>DEPRECATED. Use <code>region</code> instead.</p></dd>


<dt>these_sample_ids</dt>
<dd><p>DEPRECATED. Use <code>these_samples_metadata</code> instead.</p></dd>


<dt>pairing_status</dt>
<dd><p>DEPRECATED.
Subset your metadata and supply these_samples_metadata instead.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>A data frame in a bedpe-like format with additional
columns that allow filtering of high-confidence SVs.</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>Retrieve Manta SVs with additional VCF information to allow for
filtering of high-confidence variants.
To get SV calls for multiple samples, supply a metadata table via
<code>these_samples_metadata</code> that has been subset to only those samples.
The results will be restricted to the sample_ids within that data frame.
This function relies on a set of specific internal functions
<a href="get_manta_sv_by_samples.html">get_manta_sv_by_samples</a> (if <code>from_cache = FALSE</code>).
This function can also restrict the returned breakpoints within a genomic
region specified via <code>region</code> (in chr:start-end format).
Useful filtering parameters are also available, use <code>min_vaf</code> to set the
minimum tumour VAF for a SV to be returned and <code>min_score</code>
to set the lowest Manta somatic score for a SV to be returned.
In addition, the user can chose to return all variants, even
the ones not passing the filter criteria. To do so,
set <code>pass_filters = FALSE</code> (defaults to TRUE).</p><div class="section">
<h3 id="advanced-settings-probably-not-for-you-"><strong>Advanced settings (probably not for you)</strong><a class="anchor" aria-label="anchor" href="#advanced-settings-probably-not-for-you-"></a></h3>


<p>Is it advised to leave the default <code>from_cache</code> setting to TRUE.
To ensure manta results arre pulled from a pre-generated merge
(i.e. the cached result).
If set to FALSE in combination with <code>write_to_file = TRUE</code>,
the function will (re)generate new merged manta calls, if the user has
the required file permissions.
Note, that if <code>write_to_file</code> is set to TRUE, the function defaults
<code>from_cache = FALSE</code> to avoid nonsense parameter combinations.
Is this function not what you are looking for? You may want:
<a href="get_combined_sv.html">get_combined_sv</a>
After running this or <a href="get_combined_sv.html">get_combined_sv</a>, you most likely want to
annotate the result using <a href="https://rdrr.io/pkg/GAMBLR.utils/man/annotate_sv.html" class="external-link">GAMBLR.utils::annotate_sv</a></p>
</div>

    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># lazily get every SV in the table with default quality filters</span></span></span>
<span class="r-in"><span><span class="va">all_sv</span> <span class="op">&lt;-</span> <span class="fu">get_manta_sv</span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "no metadata provided, fetching all samples..."</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "dropping capture samples because manta results\n      are only available for genome seq_type"</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The cached results were last updated: 2025-02-24 16:06:43.114603</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Reading cached results...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "No Manta SVs found for 327 samples and 13 cohorts"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [1] "DLBCL_LSARP_Trios"   "tFL_LSARP_Trios"     "pFL_LSARP_Trios"    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [4] "FL_FOLL_BR"          "DLBCL_TFRI_DarkZone" "DLBCL_Pasqualucci"  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [7] "DLBCL_montreal"      "DLBCL_Jain"          "DLBCL_cell_lines"   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [10] "MCL_CellLines"       "cHL_Maura"           "MM_mmsanger"        </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [13] "SMZL_Strefford"     </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following VCF filters are applied;</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   Minimum VAF: 0.1</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   Minimum Score: 40</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   Only keep variants passing the quality filter: TRUE</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Returning 789098 variants from 1664 sample(s)</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Done!</span>
<span class="r-in"><span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">all_sv</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">14</span><span class="op">)</span> <span class="op"><a href="pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> genomic_data Object</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Genome Build: grch37 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Showing first 10 rows:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   CHROM_A START_A END_A CHROM_B   START_B     END_B</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1       1   10286 10286       8 146301391 146301391</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2       1   10309 10837      12     95038     95505</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3       1   10347 10630      15 102520227 102520676</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4       1   10438 10438       8 146301391 146301391</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5       1   10438 10438       8 146301391 146301391</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6       1   10457 10839      12     94873     95291</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                     manta_name SCORE STRAND_A STRAND_B tumour_sample_id</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1   MantaBND:5:1923:1927:0:0:0    46        +        +        09-41114T</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2   MantaBND:1:6049:6050:1:0:0    52        +        +     4687-03-01BD</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3  MantaBND:11:3940:4135:0:0:0    58        -        -        12-34927T</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4   MantaBND:2:7221:7224:0:1:0    84        +        +      102-01-01TD</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5   MantaBND:2:1723:1728:0:0:0    81        +        +    102-0202-1DVT</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6 MantaBND:3:26317:26320:0:0:0    56        +        +     4690-03-01BD</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   normal_sample_id VAF_tumour  DP</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1        14-11247N      0.118 110</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2   14-11247Normal      0.250  52</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3        14-11247N      0.135 104</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4   14-11247Normal      0.520  25</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5   14-11247Normal      0.630  27</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6   14-11247Normal      0.333  18</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># get all SVs for just one cohort</span></span></span>
<span class="r-in"><span><span class="va">cohort_meta</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="fu"><a href="get_gambl_metadata.html">get_gambl_metadata</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="pipe.html">%&gt;%</a></span> </span></span>
<span class="r-in"><span>              <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">cohort</span> <span class="op">==</span> <span class="st">"DLBCL_cell_lines"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">some_sv</span> <span class="op">&lt;-</span> <span class="fu">get_manta_sv</span><span class="op">(</span>these_samples_metadata <span class="op">=</span> <span class="va">cohort_meta</span>, verbose<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">some_sv</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">14</span><span class="op">)</span> <span class="op"><a href="pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> genomic_data Object</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Genome Build: grch37 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Showing first 10 rows:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   CHROM_A START_A   END_A CHROM_B START_B   END_B               manta_name</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1       1  963851  963870       1  964461  964461 MantaDEL:14848:0:0:0:0:0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2       1 1142719 1142719       1 1143140 1143140 MantaDEL:14306:0:0:0:0:0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3       1 1142719 1142719       1 1143140 1143140 MantaDEL:14173:0:0:0:0:0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4       1 1142719 1142719       1 1143140 1143140 MantaDEL:11910:0:0:0:0:0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5       1 1161716 1161716       1 1161780 1161780 MantaDEL:15361:0:0:0:0:0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6       1 1161716 1161716       1 1161780 1161780 MantaDEL:11880:0:0:0:0:0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   SCORE STRAND_A STRAND_B tumour_sample_id normal_sample_id VAF_tumour  DP</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1   144        +        -           Toledo        14-11247N      0.923  26</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2    94        +        -            HBL-1        14-11247N      0.300 100</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3    81        +        -         SU-DHL-4        14-11247N      0.256  78</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4    55        +        -         SU-DHL-9        14-11247N      0.183  60</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5    48        +        -               HT        14-11247N      0.273  44</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6    58        +        -            MD903        14-11247N      0.471  34</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">some_sv</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 21216</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># get the SVs in a region around MYC</span></span></span>
<span class="r-in"><span><span class="co"># WARNING: This is not the best way to find MYC SVs.</span></span></span>
<span class="r-in"><span><span class="co"># Use annotate_sv on the full SV set instead.</span></span></span>
<span class="r-in"><span><span class="va">myc_region_hg38</span> <span class="op">=</span> <span class="st">"chr8:127710883-127761821"</span></span></span>
<span class="r-in"><span><span class="va">myc_region_grch37</span> <span class="op">=</span> <span class="st">"8:128723128-128774067"</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">hg38_myc_locus_sv</span> <span class="op">&lt;-</span> <span class="fu">get_manta_sv</span><span class="op">(</span>region <span class="op">=</span> <span class="va">myc_region_hg38</span>,</span></span>
<span class="r-in"><span>                                projection <span class="op">=</span> <span class="st">"hg38"</span>,</span></span>
<span class="r-in"><span>                                verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">hg38_myc_locus_sv</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">14</span><span class="op">)</span> <span class="op"><a href="pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> genomic_data Object</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Genome Build: hg38 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Showing first 10 rows:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   CHROM_A  START_A    END_A CHROM_B   START_B     END_B</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1    chr2  9700440  9700440    chr8 127726024 127726024</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2    chr2 28983233 28983240    chr8 127711264 127711271</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3    chr2 88858802 88858802    chr8 127744262 127744262</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4    chr2 88860304 88860306    chr8 127751936 127751938</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5    chr2 88860417 88860417    chr8 127751955 127751955</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6    chr2 88861500 88861500    chr8 127748752 127748752</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                     manta_name SCORE STRAND_A STRAND_B</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1     MantaBND:80035:1:8:0:0:0   103        +        +</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2 MantaBND:3:52907:52908:0:3:0    43        -        -</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3    MantaBND:279432:0:1:0:0:0   148        +        +</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4  MantaBND:194837:0:1:0:0:0:0   102        +        +</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5  MantaBND:194837:0:1:0:0:0:0    73        -        -</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6   MantaBND:1102030:0:1:0:0:0    89        +        +</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>            tumour_sample_id          normal_sample_id VAF_tumour  DP</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1 BLGSP-71-06-00252-01A-01D BLGSP-71-06-00252-10A-01D      0.194 252</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2           02-14764_tumorB           02-14764_normal      0.109  55</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3                   SP59344                   SP59342      0.386  88</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4 BLGSP-71-27-00414-01A-01E BLGSP-71-27-00414-10A-01D      0.171 280</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5 BLGSP-71-27-00414-01A-01E BLGSP-71-27-00414-10A-01D      0.117 230</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6 BLGSP-71-30-00647-01A-01E BLGSP-71-06-00286-99A-01D      0.283  46</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">hg38_myc_locus_sv</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 458</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">incorrect_myc_locus_sv</span> <span class="op">&lt;-</span> <span class="fu">get_manta_sv</span><span class="op">(</span>region <span class="op">=</span> <span class="va">myc_region_grch37</span>,</span></span>
<span class="r-in"><span>                                projection <span class="op">=</span> <span class="st">"hg38"</span>,</span></span>
<span class="r-in"><span>                                verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">incorrect_myc_locus_sv</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">14</span><span class="op">)</span> <span class="op"><a href="pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> genomic_data Object</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Genome Build: hg38 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Showing first 10 rows:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   CHROM_A   START_A     END_A CHROM_B   START_B     END_B</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1    chr4  77227094  77227100    chr8 128767241 128767247</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2    chr8   1287381   1287381    chr8   1287384   1287384</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3    chr8 128726344 128727379   chr11  93629113  93629647</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4    chr8 128726820 128726820    chr8 128726825 128726825</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5    chr8 128726820 128726820    chr8 128726825 128726825</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6    chr8 128738979 128738983    chr8 128752584 128752588</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                   manta_name SCORE STRAND_A STRAND_B tumour_sample_id</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1  MantaBND:658884:1:2:0:0:0    42        -        -  14-33798_tumorB</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2 MantaINS:1063533:0:0:0:4:0    51        +        -  97-28459_tumorB</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3   MantaBND:28037:1:9:0:0:0    66        -        +        01-20774T</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4  MantaINS:242009:0:0:0:3:0    76        +        -         PD26403a</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5  MantaINS:226876:7:7:1:3:0    84        +        -         PD26403c</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6 MantaDEL:1407936:0:1:0:0:0   118        +        -  04-14093_tumorA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   normal_sample_id VAF_tumour  DP</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1  14-33798_normal      0.136  44</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2          FL3006N      0.308  26</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3        14-11247N      0.280  25</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4         PD26403b      0.400 105</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5         PD26403b      0.407 113</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6  04-14093_normal      0.442  43</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">incorrect_myc_locus_sv</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 28</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Despite potentially being incomplete, we can nonetheless</span></span></span>
<span class="r-in"><span><span class="co"># annotate these directly for more details</span></span></span>
<span class="r-in"><span><span class="va">annotated_myc_hg38</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>         <span class="fu">GAMBLR.utils</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/GAMBLR.utils/man/annotate_sv.html" class="external-link">annotate_sv</a></span><span class="op">(</span><span class="va">hg38_myc_locus_sv</span>, genome_build <span class="op">=</span> <span class="st">"hg38"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">annotated_myc_hg38</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   chrom1    start1      end1 chrom2    start2      end2 name score strand1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1      2  28983233  28983240      8 127711264 127711271    .    43       -</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2      4   1746419   1746421      8 127723483 127723485    .    77       -</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3      8 127226860 127226862      8 127759782 127759784    .    56       +</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4      8 127226860 127226860      8 127759821 127759821    .    51       -</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5      8 127301019 127301020      8 127742838 127742839    .    71       -</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6      8 127301020 127301022      8 127742838 127742840    .    65       +</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   strand2 tumour_sample_id  gene partner   fusion</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1       -  02-14764_tumorB   ALK    &lt;NA&gt;   NA-ALK</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2       -        09-41114T WHSC1    &lt;NA&gt; NA-WHSC1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3       +          SP13307   MYC    &lt;NA&gt;   NA-MYC</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4       -          SP13307   MYC    &lt;NA&gt;   NA-MYC</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5       -      365-16-01TD   MYC    &lt;NA&gt;   NA-MYC</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6       +      365-16-01TD   MYC    &lt;NA&gt;   NA-MYC</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">annotated_myc_hg38</span><span class="op">$</span><span class="va">partner</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  BCL6 CCNL1   DMD   IGH   IGK   IGL  LRMP  PAX5 RFTN1 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     3     1     2   293     5     6     1     5     1 </span>
<span class="r-in"><span><span class="co"># The usual MYC partners are seen here</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">annotated_myc_incorrect</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>         <span class="fu">GAMBLR.utils</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/GAMBLR.utils/man/annotate_sv.html" class="external-link">annotate_sv</a></span><span class="op">(</span><span class="va">incorrect_myc_locus_sv</span>, genome_build <span class="op">=</span> <span class="st">"hg38"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">annotated_myc_incorrect</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   chrom1    start1      end1 chrom2    start2      end2 name score strand1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1      8 128726344 128727379     11  93629113  93629647    .    66       -</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2      8 128726820 128726820      8 128726825 128726825    .    76       +</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3      8 128726820 128726820      8 128726825 128726825    .    84       +</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4      8 128738979 128738983      8 128752584 128752588    .   118       +</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5      8 128738979 128738983      8 128752584 128752588    .   127       +</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6      8 128738981 128738981      8 128752584 128752584    .   126       +</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   strand2 tumour_sample_id gene partner fusion</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1       +        01-20774T  MYC    &lt;NA&gt; NA-MYC</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2       -         PD26403a  MYC    &lt;NA&gt; NA-MYC</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3       -         PD26403c  MYC    &lt;NA&gt; NA-MYC</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4       -  04-14093_tumorA  MYC    &lt;NA&gt; NA-MYC</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5       -  04-14093_tumorB  MYC    &lt;NA&gt; NA-MYC</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6       -        05-24065T  MYC    &lt;NA&gt; NA-MYC</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">annotated_myc_incorrect</span><span class="op">$</span><span class="va">partner</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> &lt; table of extent 0 &gt;</span>
<span class="r-in"><span><span class="co"># The effect of specifying the wrong coordinate is evident</span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Ryan Morin, Kostiantyn Dreval, Vladimir Souza.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

