<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Get SSM By Sample. — get_ssm_by_sample • GAMBLR.results</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Get SSM By Sample. — get_ssm_by_sample"><meta name="description" content="Get the SSMs (i.e. load MAF) for a single sample."><meta property="og:description" content="Get the SSMs (i.e. load MAF) for a single sample."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">GAMBLR.results</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.3</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/GAMBLR-results.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Developers</h6></li>
    <li><a class="dropdown-item" href="../articles/testing.html">Testing GAMBLR packages</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Get SSM By Sample.</h1>

      <div class="d-none name"><code>get_ssm_by_sample.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Get the SSMs (i.e. load MAF) for a single sample.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">get_ssm_by_sample</span><span class="op">(</span></span>
<span>  <span class="va">these_samples_metadata</span>,</span>
<span>  tool_name <span class="op">=</span> <span class="st">"slms-3"</span>,</span>
<span>  projection <span class="op">=</span> <span class="st">"grch37"</span>,</span>
<span>  augmented <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  flavour <span class="op">=</span> <span class="st">"clustered"</span>,</span>
<span>  min_read_support <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  basic_columns <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  maf_cols <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  <span class="va">this_sample_id</span>,</span>
<span>  <span class="va">this_seq_type</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-these-samples-metadata">these_samples_metadata<a class="anchor" aria-label="anchor" href="#arg-these-samples-metadata"></a></dt>
<dd><p>Required if not specifying both this_sample_id and this_seq_type a single row or entire metadata table containing your sample_id.</p></dd>


<dt id="arg-tool-name">tool_name<a class="anchor" aria-label="anchor" href="#arg-tool-name"></a></dt>
<dd><p>The name of the variant calling pipeline (currently only slms-3 is supported).</p></dd>


<dt id="arg-projection">projection<a class="anchor" aria-label="anchor" href="#arg-projection"></a></dt>
<dd><p>The projection genome build. Supports hg38 and grch37.</p></dd>


<dt id="arg-augmented">augmented<a class="anchor" aria-label="anchor" href="#arg-augmented"></a></dt>
<dd><p>default: TRUE. Set to FALSE if you instead want the original MAF from each sample for multi-sample patients instead of the augmented MAF.</p></dd>


<dt id="arg-flavour">flavour<a class="anchor" aria-label="anchor" href="#arg-flavour"></a></dt>
<dd><p>Currently this function only supports one flavour option but this feature is meant for eventual compatibility with additional variant calling parameters/versions.</p></dd>


<dt id="arg-min-read-support">min_read_support<a class="anchor" aria-label="anchor" href="#arg-min-read-support"></a></dt>
<dd><p>Only returns variants with at least this many reads in t_alt_count (for cleaning up augmented MAFs).</p></dd>


<dt id="arg-basic-columns">basic_columns<a class="anchor" aria-label="anchor" href="#arg-basic-columns"></a></dt>
<dd><p>Return first 43 columns of MAF rather than full details. Default is TRUE.</p></dd>


<dt id="arg-maf-cols">maf_cols<a class="anchor" aria-label="anchor" href="#arg-maf-cols"></a></dt>
<dd><p>if basic_columns is set to FALSE, the user can specify what columns to be returned within the MAF. This parameter can either be a vector of indexes (integer) or a vector of characters.</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>Enable for debugging/noisier output.</p></dd>


<dt id="arg-this-sample-id">this_sample_id<a class="anchor" aria-label="anchor" href="#arg-this-sample-id"></a></dt>
<dd><p>Deprecated. Inferred from these_samples_metadata</p></dd>


<dt id="arg-this-seq-type">this_seq_type<a class="anchor" aria-label="anchor" href="#arg-this-seq-type"></a></dt>
<dd><p>Deprecated. Inferred from these_samples_metadata</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>data frame in MAF format.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>This was implemented to allow flexibility because there are some samples that we may want to use a different set of variants than those in the main GAMBL merge.
The current use case is to allow a force_unmatched output to be used to replace the SSMs from the merge for samples with known contamination in the normal.
This will also be useful to apply a blacklist to individual MAFs when coupled with <a href="annotate_ssm_blacklist.html">annotate_ssm_blacklist</a>.
Is this function not what you are looking for? Try one of the following, similar, functions; <a href="get_coding_ssm.html">get_coding_ssm</a>, <a href="get_coding_ssm_status.html">get_coding_ssm_status</a>,
get_ssm_by_patients, <a href="get_ssm_by_samples.html">get_ssm_by_samples</a>, <a href="get_ssm_by_region.html">get_ssm_by_region</a>, <a href="get_ssm_by_regions.html">get_ssm_by_regions</a></p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">maf_samp</span> <span class="op">=</span> <span class="fu">GAMBLR.results</span><span class="fu">:::</span><span class="fu">get_ssm_by_sample</span><span class="op">(</span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="get_gambl_metadata.html">get_gambl_metadata</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="pipe.html">%&gt;%</a></span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">sample_id</span><span class="op">==</span><span class="st">"13-27975_tumorA"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  augmented <span class="op">=</span> <span class="cn">FALSE</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 3273 capture samples are missing a value for protocol. Assuming Exome.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 138 biopsies are missing from the biopsy metadata. This should be fixed!</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> affected cohorts:  DLBCL_LSARP_Trios,Ennishi_tapestri,SMZL_Strefford,cHL_Maura,MCL_Barcelona</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 110 biopsies with discrepancies in the pathology field. This should be fixed!</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 10 biopsies with discrepancies in the time_point field. This should be fixed!</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">maf_samp</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 4705</span>
<span class="r-in"><span><span class="va">maf_samp_aug</span> <span class="op">=</span> <span class="fu">GAMBLR.results</span><span class="fu">:::</span><span class="fu">get_ssm_by_sample</span><span class="op">(</span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="get_gambl_metadata.html">get_gambl_metadata</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="pipe.html">%&gt;%</a></span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">sample_id</span><span class="op">==</span><span class="st">"13-27975_tumorA"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  augmented <span class="op">=</span> <span class="cn">TRUE</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 3273 capture samples are missing a value for protocol. Assuming Exome.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 138 biopsies are missing from the biopsy metadata. This should be fixed!</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> affected cohorts:  DLBCL_LSARP_Trios,Ennishi_tapestri,SMZL_Strefford,cHL_Maura,MCL_Barcelona</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 110 biopsies with discrepancies in the pathology field. This should be fixed!</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 10 biopsies with discrepancies in the time_point field. This should be fixed!</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">maf_samp_aug</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 6118</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span> <span class="va">some_maf</span> <span class="op">=</span> <span class="fu">GAMBLR.results</span><span class="fu">:::</span><span class="fu">get_ssm_by_sample</span><span class="op">(</span></span></span>
<span class="r-in"><span>                          these_samples_metadata <span class="op">=</span> <span class="fu"><a href="get_gambl_metadata.html">get_gambl_metadata</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="pipe.html">%&gt;%</a></span></span></span>
<span class="r-in"><span>                            <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">sample_id</span> <span class="op">==</span> <span class="st">"HTMCP-01-06-00485-01A-01D"</span>,</span></span>
<span class="r-in"><span>                                     <span class="va">seq_type</span> <span class="op">==</span> <span class="st">"genome"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                         projection <span class="op">=</span> <span class="st">"hg38"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 3273 capture samples are missing a value for protocol. Assuming Exome.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 138 biopsies are missing from the biopsy metadata. This should be fixed!</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> affected cohorts:  DLBCL_LSARP_Trios,Ennishi_tapestri,SMZL_Strefford,cHL_Maura,MCL_Barcelona</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 110 biopsies with discrepancies in the pathology field. This should be fixed!</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 10 biopsies with discrepancies in the time_point field. This should be fixed!</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>The following named parsers don't match the column names: GENE_PHENO, FILTER, flanking_bps, vcf_id, vcf_qual, gnomAD_AF, gnomAD_AFR_AF, gnomAD_AMR_AF, gnomAD_SAS_AF, vcf_pos, gnomADg_AF, blacklist_count</span>
<span class="r-in"><span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">some_maf</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> genomic_data Object</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Genome Build: hg38 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Showing first 10 rows:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>        Hugo_Symbol Entrez_Gene_Id Center NCBI_Build Chromosome Start_Position</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1          Unknown              0      .     GRCh38       chr1        5446963</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2           CAMTA1              0      .     GRCh38       chr1        7244319</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3             PER3              0      .     GRCh38       chr1        7828459</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4          TNFRSF9              0      .     GRCh38       chr1        7911008</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5  ENSR00000000893              0      .     GRCh38       chr1        8181883</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6           PIK3CD              0      .     GRCh38       chr1        9726972</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 7            CASZ1              0      .     GRCh38       chr1       10720942</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 8           CELA2A              0      .     GRCh38       chr1       15454068</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 9          PLA2G2A              0      .     GRCh38       chr1       19975558</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 10         PLA2G2C              0      .     GRCh38       chr1       20172387</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    End_Position Strand Variant_Classification Variant_Type</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1       5446963      +                    IGR          SNP</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2       7244319      +                 Intron          SNP</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3       7828459      +                 Intron          SNP</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4       7911008      +                3'Flank          SNP</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5       8181883      +                    IGR          SNP</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6       9726972      +      Missense_Mutation          SNP</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 7      10720942      +                 Intron          SNP</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 8      15454068      +                5'Flank          SNP</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 9      19975559      +                  3'UTR          DEL</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 10     20172387      +                 Intron          SNP</span>
<span class="r-in"><span> </span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Ryan Morin, Kostiantyn Dreval, Vladimir Souza.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.9000.</p>
</div>

    </footer></div>





  </body></html>

