<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Get ASHM Count Matrix. — get_ashm_count_matrix • GAMBLR.results</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Get ASHM Count Matrix. — get_ashm_count_matrix"><meta property="og:description" content="Prepare a matrix with one row per sample and one column per
region using a set of hypermutated regions."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">GAMBLR.results</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/Overview.html">What's new in GAMBLR</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Get ASHM Count Matrix.</h1>

    <div class="hidden name"><code>get_ashm_count_matrix.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Prepare a matrix with one row per sample and one column per
region using a set of hypermutated regions.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">get_ashm_count_matrix</span><span class="op">(</span></span>
<span>  <span class="va">regions_bed</span>,</span>
<span>  <span class="va">maf_data</span>,</span>
<span>  <span class="va">these_samples_metadata</span>,</span>
<span>  this_seq_type <span class="op">=</span> <span class="st">"genome"</span>,</span>
<span>  <span class="va">projection</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>


<dl><dt id="arg-regions-bed">regions_bed<a class="anchor" aria-label="anchor" href="#arg-regions-bed"></a></dt>
<dd><p>A bed file with one row for each region.
The first three columns in this file MUST contain the Chromosome, Start and End position</p></dd>


<dt id="arg-maf-data">maf_data<a class="anchor" aria-label="anchor" href="#arg-maf-data"></a></dt>
<dd><p>Optionally provide a data frame in the MAF format, otherwise
either GAMBLR.data or GAMBLR.results will be used.</p></dd>


<dt id="arg-these-samples-metadata">these_samples_metadata<a class="anchor" aria-label="anchor" href="#arg-these-samples-metadata"></a></dt>
<dd><p>This is used to complete your matrix. All GAMBL
samples of the specified seq_type will be used by default. Provide a data frame with at least
sample_id for all samples if you are using non-GAMBL data.</p></dd>


<dt id="arg-this-seq-type">this_seq_type<a class="anchor" aria-label="anchor" href="#arg-this-seq-type"></a></dt>
<dd><p>The seq_type to return results for. Must be a single value. Only used if no
metadata is provided with these_samples_metadata.</p></dd>


<dt id="arg-projection">projection<a class="anchor" aria-label="anchor" href="#arg-projection"></a></dt>
<dd><p>The genome build we are working with</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>A data frame with a row for every sample in these_samples_metadata and a column for every region in regions_bed</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>Values are the number of mutations in that patient in the region.</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span>  <span class="va">DLBCL_genome_meta</span> <span class="op">=</span> <span class="fu"><a href="get_gambl_metadata.html">get_gambl_metadata</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="pipe.html">%&gt;%</a></span> </span></span>
<span class="r-in"><span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">pathology</span><span class="op">==</span><span class="st">"DLBCL"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 3258 capture samples are missing a value for protocol. Assuming Exome.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 63 biopsies are missing from the biopsy metadata. This should be fixed!</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> affected cohorts:  DLBCL_LSARP_Trios,SMZL_Strefford,cHL_Maura</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 110 biopsies with discrepancies in the pathology field. This should be fixed!</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 10 biopsies with discrepancies in the time_point field. This should be fixed!</span>
<span class="r-in"><span><span class="co">#get ashm regions</span></span></span>
<span class="r-in"><span><span class="va">some_regions</span> <span class="op">=</span> <span class="fu">GAMBLR.utils</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/GAMBLR.utils/man/create_bed_data.html" class="external-link">create_bed_data</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>                              <span class="fu">GAMBLR.data</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/GAMBLR.data/man/grch37_ashm_regions.html" class="external-link">grch37_ashm_regions</a></span>,</span></span>
<span class="r-in"><span>                              fix_names <span class="op">=</span> <span class="st">"concat"</span>,</span></span>
<span class="r-in"><span>                              concat_cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"gene"</span>,<span class="st">"region"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                              sep<span class="op">=</span><span class="st">"-"</span><span class="op">)</span> <span class="op"><a href="pipe.html">%&gt;%</a></span></span></span>
<span class="r-in"><span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grepl</a></span><span class="op">(</span><span class="st">"PAX5"</span>,<span class="va">name</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">pax5_matrix</span> <span class="op">&lt;-</span> <span class="fu">get_ashm_count_matrix</span><span class="op">(</span></span></span>
<span class="r-in"><span>     regions_bed <span class="op">=</span> <span class="va">some_regions</span>,</span></span>
<span class="r-in"><span>     this_seq_type <span class="op">=</span> <span class="st">"genome"</span>,</span></span>
<span class="r-in"><span>     these_samples_metadata <span class="op">=</span> <span class="va">DLBCL_genome_meta</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Streamlined is set to TRUE, this function will disregard anything specified with basic_columns</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> To return a MAF with standard 45 columns, set streamlioned = FALSE and basic_columns = TRUE</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> To return a maf with all (116) columns, set streamlined = FALSE and basic_columns = FALSE</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Joining with `by = join_by(sample_id, region_name)`</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">pax5_matrix</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                   PAX5-TSS-1 PAX5-distal-enhancer-1 PAX5-distal-enhancer-3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 00-14595_tumorC            2                      4                      3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 00-15201_tumorA            0                      0                      3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 00-15201_tumorB            0                      0                      0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 00-17960_CLC01670          0                      0                      0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> FL1015T2                   0                      1                      0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 00-23442_tumorB            0                      0                      2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                   PAX5-intron-1 PAX5-distal-enhancer-2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 00-14595_tumorC              12                      0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 00-15201_tumorA               7                      0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 00-15201_tumorB               1                      0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 00-17960_CLC01670            11                      2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> FL1015T2                      0                      0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 00-23442_tumorB               0                      0</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">colMeans</a></span><span class="op">(</span><span class="va">pax5_matrix</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>             PAX5-TSS-1 PAX5-distal-enhancer-1 PAX5-distal-enhancer-3 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>              0.4402619              0.5319149              1.2487725 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          PAX5-intron-1 PAX5-distal-enhancer-2 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>              1.9247136              0.6579378 </span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Ryan Morin, Kostiantyn Dreval, Vladimir Souza.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

      </footer></div>






  </body></html>

