<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Get GAMBL metadata. — get_gambl_metadata • GAMBLR.results</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Get GAMBL metadata. — get_gambl_metadata"><meta property="og:description" content="Return metadata for a selection of samples."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">GAMBLR.results</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Get GAMBL metadata.</h1>
    
    <div class="hidden name"><code>get_gambl_metadata.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Return metadata for a selection of samples.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">get_gambl_metadata</span><span class="op">(</span></span>
<span>  dna_seq_type_priority <span class="op">=</span> <span class="st">"genome"</span>,</span>
<span>  capture_protocol_priority <span class="op">=</span> <span class="st">"Exome"</span>,</span>
<span>  dna_preservation_priority <span class="op">=</span> <span class="st">"frozen"</span>,</span>
<span>  exome_capture_space_priority <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"agilent-sureselect-human-all-exon-v7"</span>,</span>
<span>    <span class="st">"agilent-sureselect-V5-plus-utr"</span>, <span class="st">"idt-xgen-v2-grch37"</span>, <span class="st">"exome-utr-grch37"</span>,</span>
<span>    <span class="st">"exome-utr-grch38"</span>, <span class="st">"none"</span><span class="op">)</span>,</span>
<span>  mrna_collapse_redundancy <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  also_normals <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  everything <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  invert <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>dna_seq_type_priority</dt>
<dd><p>The default is "genome" and the only other option is "capture". For duplicate biopsy_id/patient combinations with different seq_type available, prioritize this seq_type and drop the others.</p></dd>


<dt>capture_protocol_priority</dt>
<dd><p>For duplicate biopsy_id/patient combinations with different seq_type available, prioritize this seq_type and drop the others.</p></dd>


<dt>dna_preservation_priority</dt>
<dd><p>Which to prioritize between FFPE and frozen samples from the same biopsy (default: "frozen")</p></dd>


<dt>exome_capture_space_priority</dt>
<dd><p>A vector specifying how to prioritize exome capture space #TODO: implement and test once examples are available</p></dd>


<dt>mrna_collapse_redundancy</dt>
<dd><p>Default: TRUE. Set to FALSE to obtain all rows for the mrna seq_type including those that would otherwise be collapsed.</p></dd>


<dt>also_normals</dt>
<dd><p>Set to TRUE to force the return of rows where tissue_status is normal (default is to restrict to tumour)</p></dd>


<dt>everything</dt>
<dd><p>Set to TRUE to include samples with <code>bam_available == FALSE</code>. Default: FALSE - only samples with <code>bam_available = TRUE</code> are retained.</p></dd>


<dt>verbose</dt>
<dd><p>Set to TRUE for a chatty output (mostly for debugging)</p></dd>


<dt>invert</dt>
<dd><p>Set to TRUE to force the function to return only the rows that are lost in all the prioritization steps (mostly for debugging)</p></dd>


<dt>...</dt>
<dd><p>Additional arguments</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>A data frame with metadata for each biopsy in GAMBL</p>


<dl><dt>compression</dt>
<dd><p>Format of the original data used as input for our analysis pipelines (cram, bam or fastq)</p></dd>

<dt>bam_available</dt>
<dd><p>Whether or not this file was available when last checked.</p></dd>

<dt>patient_id</dt>
<dd><p>The anonymized unique identifier for this patient. For BC samples, this will be Res ID.</p></dd>

<dt>sample_id</dt>
<dd><p>A unique identifier for the sample analyzed.</p></dd>

<dt>seq_type</dt>
<dd><p>The assay type used to produce this data (one of "genome","capture, "mrna", "promethION")</p></dd>

<dt>capture_space</dt>
<dd><p>Unique ID for the capture space, where applicable</p></dd>

<dt>genome_build</dt>
<dd><p>The name of the genome reference the data were aligned to.</p></dd>

<dt>tissue_status</dt>
<dd><p>Whether the sample was atumour or normal.</p></dd>

<dt>cohort</dt>
<dd><p>Name for a group of samples that were added together (usually from a single study), often in the format pathology_cohort_descriptor.</p></dd>

<dt>library_id</dt>
<dd><p>The unique identifier for the sequencing library.</p></dd>

<dt>pathology</dt>
<dd><p>The diagnosis or pathology for the sample</p></dd>

<dt>time_point</dt>
<dd><p>Timing of biopsy in increasing alphabetical order (A = diagnosis, B = first relapse etc)</p></dd>

<dt>protocol</dt>
<dd><p>General protocol for library construction. e.g. "Ribodepletion", "PolyA", or "Genome"</p></dd>

<dt>ffpe_or_frozen</dt>
<dd><p>Whether the nucleic acids were extracted from a frozen or FFPE sample</p></dd>

<dt>read_length</dt>
<dd><p>The length of reads (required for RNA-seq libraries)</p></dd>

<dt>strandedness</dt>
<dd><p>Whether the RNA-seq librayr construction was strand-specific and, if so, which strand. Required for RNAseq; "positive", "negative", or "unstranded")</p></dd>

<dt>seq_source_type</dt>
<dd><p>Required for RNAseq. Usually the same value as ffpe_or_frozen but sometimes immunotube or sorted cells</p></dd>

<dt>data_path</dt>
<dd><p>Symbolic link to the bam or cram file (not usually relevant for GAMBLR)</p></dd>

<dt>link_name</dt>
<dd><p>Standardized naming for symbolic link (not usually relevant for GAMBLR)</p></dd>

<dt>fastq_data_path</dt>
<dd><p>Symbolic link to the fastq file (not usually relevant for GAMBLR)</p></dd>

<dt>fastq_link_name</dt>
<dd><p>Standardized naming for symbolic link for FASTQ file, if used (not usually relevant for GAMBLR)</p></dd>

<dt>unix_group</dt>
<dd><p>Whether a source is external and restricted by data access agreements (icgc_dart) or internal (gambl)</p></dd>

<dt>COO_consensus</dt>
<dd><p>TODO</p></dd>

<dt>DHITsig_consensus</dt>
<dd><p>TODO</p></dd>

<dt>COO_PRPS_class</dt>
<dd><p>TODO</p></dd>

<dt>DHITsig_PRPS_class</dt>
<dd><p>TODO</p></dd>

<dt>DLBCL90_dlbcl_call</dt>
<dd><p>TODO</p></dd>

<dt>DLBCL90_dhitsig_call</dt>
<dd><p>TODO</p></dd>

<dt>res_id</dt>
<dd><p>duplicate of sample_id for local samples and NA otherwise</p></dd>

<dt>DLBCL90_code_set</dt>
<dd><p>Code set used for DLBCL90 call. One of DLBCL90 DLBCL90v2 DLBCL90v3</p></dd>

<dt>DLBCL90_dlbcl_score</dt>
<dd><p>TODO</p></dd>

<dt>DLBCL90_pmbl_score</dt>
<dd><p>TODO</p></dd>

<dt>DLBCL90_pmbl_call</dt>
<dd><p>TODO</p></dd>

<dt>DLBCL90_dhitsig_score</dt>
<dd><p>TODO</p></dd>

<dt>myc_ba</dt>
<dd><p>Result from breakapart FISH for MYC locus</p></dd>

<dt>myc_cn</dt>
<dd><p>Result from copy number FISH for MYC locus</p></dd>

<dt>bcl2_ba</dt>
<dd><p>Result from breakapart FISH for BCL2 locus</p></dd>

<dt>bcl2_cn</dt>
<dd><p>Result from copy number FISH for BCL2 locus</p></dd>

<dt>bcl6_ba</dt>
<dd><p>Result from breakapart FISH for BCL6 locus</p></dd>

<dt>bcl6_cn</dt>
<dd><p>Result from copy number FISH for BCL6 locus</p></dd>

<dt>time_since_diagnosis_years</dt>
<dd><p>TODO</p></dd>

<dt>relapse_timing</dt>
<dd><p>TODO</p></dd>

<dt>dtbx</dt>
<dd><p>TODO. OR REMOVE?</p></dd>

<dt>dtdx</dt>
<dd><p>TODO. OR REMOVE?</p></dd>

<dt>lymphgen_no_cnv</dt>
<dd><p>TODO</p></dd>

<dt>lymphgen_with_cnv</dt>
<dd><p>TODO</p></dd>

<dt>lymphgen_cnv_noA53</dt>
<dd><p>TODO</p></dd>

<dt>lymphgen_wright</dt>
<dd><p>The LymphGen call for this sample from Wright et all (if applicable)</p></dd>

<dt>fl_grade</dt>
<dd><p>TODO</p></dd>

<dt>transformation</dt>
<dd><p>TODO</p></dd>

<dt>relapse</dt>
<dd><p>TODO</p></dd>

<dt>ighv_mutation_original</dt>
<dd><p>TODO</p></dd>


</dl></div>
    <div id="details">
    <h2>Details</h2>
    <p>This function returns metadata for GAMBL samples.
This replaces the functionality of the original version, which is still available under the new name <a href="og_get_gambl_metadata.html">og_get_gambl_metadata</a>.
The purpose of this function is to provide the metadata for the non-redundant set of samples from GAMBL, dealing with all
types of redundancy caused by samples or biopsies that have data from &gt;1 seq_type (genome or capture), different capture protocols (exome or targeted capture) etc.</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Ryan Morin, Kostiantyn Dreval, Vladimir Souza.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

