<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Get GAMBL metadata. — get_gambl_metadata • GAMBLR.results</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Get GAMBL metadata. — get_gambl_metadata"><meta name="description" content="Return metadata for a selection of samples."><meta property="og:description" content="Return metadata for a selection of samples."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">GAMBLR.results</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.3</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/GAMBLR-results.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Developers</h6></li>
    <li><a class="dropdown-item" href="../articles/testing.html">Testing GAMBLR packages</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Get GAMBL metadata.</h1>

      <div class="d-none name"><code>get_gambl_metadata.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Return metadata for a selection of samples.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">get_gambl_metadata</span><span class="op">(</span></span>
<span>  dna_seq_type_priority <span class="op">=</span> <span class="st">"genome"</span>,</span>
<span>  capture_protocol_priority <span class="op">=</span> <span class="st">"Exome"</span>,</span>
<span>  dna_preservation_priority <span class="op">=</span> <span class="st">"frozen"</span>,</span>
<span>  exome_capture_space_priority <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"agilent-sureselect-human-all-exon-v7"</span>,</span>
<span>    <span class="st">"agilent-sureselect-V5-plus-utr"</span>, <span class="st">"idt-xgen-v2-grch37"</span>, <span class="st">"exome-utr-grch37"</span>,</span>
<span>    <span class="st">"exome-utr-grch38"</span>, <span class="st">"none"</span><span class="op">)</span>,</span>
<span>  mrna_collapse_redundancy <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  also_normals <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  everything <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  invert <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  exclude <span class="op">=</span> <span class="st">"promethION"</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-dna-seq-type-priority">dna_seq_type_priority<a class="anchor" aria-label="anchor" href="#arg-dna-seq-type-priority"></a></dt>
<dd><p>The default is "genome" and the only other option is "capture". For duplicate biopsy_id/patient combinations with different seq_type available, prioritize this seq_type and drop the others.</p></dd>


<dt id="arg-capture-protocol-priority">capture_protocol_priority<a class="anchor" aria-label="anchor" href="#arg-capture-protocol-priority"></a></dt>
<dd><p>For duplicate biopsy_id/patient combinations with different seq_type available, prioritize this seq_type and drop the others.</p></dd>


<dt id="arg-dna-preservation-priority">dna_preservation_priority<a class="anchor" aria-label="anchor" href="#arg-dna-preservation-priority"></a></dt>
<dd><p>Which to prioritize between FFPE and frozen samples from the same biopsy (default: "frozen")</p></dd>


<dt id="arg-exome-capture-space-priority">exome_capture_space_priority<a class="anchor" aria-label="anchor" href="#arg-exome-capture-space-priority"></a></dt>
<dd><p>A vector specifying how to prioritize exome capture space #TODO: implement and test once examples are available</p></dd>


<dt id="arg-mrna-collapse-redundancy">mrna_collapse_redundancy<a class="anchor" aria-label="anchor" href="#arg-mrna-collapse-redundancy"></a></dt>
<dd><p>Default: TRUE. Set to FALSE to obtain all rows for the mrna seq_type including those that would otherwise be collapsed.</p></dd>


<dt id="arg-also-normals">also_normals<a class="anchor" aria-label="anchor" href="#arg-also-normals"></a></dt>
<dd><p>Set to TRUE to force the return of rows where tissue_status is normal (default is to restrict to tumour)</p></dd>


<dt id="arg-everything">everything<a class="anchor" aria-label="anchor" href="#arg-everything"></a></dt>
<dd><p>Set to TRUE to include samples with <code>bam_available == FALSE</code>. Default: FALSE - only samples with <code>bam_available = TRUE</code> are retained.</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>Set to TRUE for a chatty output (mostly for debugging)</p></dd>


<dt id="arg-invert">invert<a class="anchor" aria-label="anchor" href="#arg-invert"></a></dt>
<dd><p>Set to TRUE to force the function to return only the rows that are lost in all the prioritization steps (mostly for debugging)</p></dd>


<dt id="arg-exclude">exclude<a class="anchor" aria-label="anchor" href="#arg-exclude"></a></dt>
<dd><p>Specify one or more seq_type to drop from the output.
This prevents metadata from containing anythong other than the three standard
seq_type (genome, capture, mrna). Default setting will exclude "promethION".</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>Additional arguments</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A data frame with metadata for each biopsy in GAMBL</p>
<dl><dt>compression</dt>
<dd><p>Format of the original data used as input for our analysis pipelines (cram, bam or fastq)</p></dd>

<dt>bam_available</dt>
<dd><p>Whether or not this file was available when last checked.</p></dd>

<dt>patient_id</dt>
<dd><p>The anonymized unique identifier for this patient. For BC samples, this will be Res ID.</p></dd>

<dt>sample_id</dt>
<dd><p>A unique identifier for the sample analyzed.</p></dd>

<dt>seq_type</dt>
<dd><p>The assay type used to produce this data (one of "genome","capture, "mrna", "promethION")</p></dd>

<dt>capture_space</dt>
<dd><p>Unique ID for the capture space, where applicable</p></dd>

<dt>genome_build</dt>
<dd><p>The name of the genome reference the data were aligned to.</p></dd>

<dt>tissue_status</dt>
<dd><p>Whether the sample was atumour or normal.</p></dd>

<dt>cohort</dt>
<dd><p>Name for a group of samples that were added together (usually from a single study), often in the format pathology_cohort_descriptor.</p></dd>

<dt>library_id</dt>
<dd><p>The unique identifier for the sequencing library.</p></dd>

<dt>pathology</dt>
<dd><p>The diagnosis or pathology for the sample</p></dd>

<dt>time_point</dt>
<dd><p>Timing of biopsy in increasing alphabetical order (A = diagnosis, B = first relapse etc)</p></dd>

<dt>protocol</dt>
<dd><p>General protocol for library construction. e.g. "Ribodepletion", "PolyA", or "Genome"</p></dd>

<dt>ffpe_or_frozen</dt>
<dd><p>Whether the nucleic acids were extracted from a frozen or FFPE sample</p></dd>

<dt>read_length</dt>
<dd><p>The length of reads (required for RNA-seq libraries)</p></dd>

<dt>strandedness</dt>
<dd><p>Whether the RNA-seq librayr construction was strand-specific and, if so, which strand. Required for RNAseq; "positive", "negative", or "unstranded")</p></dd>

<dt>seq_source_type</dt>
<dd><p>Required for RNAseq. Usually the same value as ffpe_or_frozen but sometimes immunotube or sorted cells</p></dd>

<dt>data_path</dt>
<dd><p>Symbolic link to the bam or cram file (not usually relevant for GAMBLR)</p></dd>

<dt>link_name</dt>
<dd><p>Standardized naming for symbolic link (not usually relevant for GAMBLR)</p></dd>

<dt>fastq_data_path</dt>
<dd><p>Symbolic link to the fastq file (not usually relevant for GAMBLR)</p></dd>

<dt>fastq_link_name</dt>
<dd><p>Standardized naming for symbolic link for FASTQ file, if used (not usually relevant for GAMBLR)</p></dd>

<dt>unix_group</dt>
<dd><p>Whether a source is external and restricted by data access agreements (icgc_dart) or internal (gambl)</p></dd>

<dt>COO_consensus</dt>
<dd><p>TODO</p></dd>

<dt>DHITsig_consensus</dt>
<dd><p>TODO</p></dd>

<dt>COO_PRPS_class</dt>
<dd><p>TODO</p></dd>

<dt>DHITsig_PRPS_class</dt>
<dd><p>TODO</p></dd>

<dt>DLBCL90_dlbcl_call</dt>
<dd><p>TODO</p></dd>

<dt>DLBCL90_dhitsig_call</dt>
<dd><p>TODO</p></dd>

<dt>res_id</dt>
<dd><p>duplicate of sample_id for local samples and NA otherwise</p></dd>

<dt>DLBCL90_code_set</dt>
<dd><p>Code set used for DLBCL90 call. One of DLBCL90 DLBCL90v2 DLBCL90v3</p></dd>

<dt>DLBCL90_dlbcl_score</dt>
<dd><p>TODO</p></dd>

<dt>DLBCL90_pmbl_score</dt>
<dd><p>TODO</p></dd>

<dt>DLBCL90_pmbl_call</dt>
<dd><p>TODO</p></dd>

<dt>DLBCL90_dhitsig_score</dt>
<dd><p>TODO</p></dd>

<dt>myc_ba</dt>
<dd><p>Result from breakapart FISH for MYC locus</p></dd>

<dt>myc_cn</dt>
<dd><p>Result from copy number FISH for MYC locus</p></dd>

<dt>bcl2_ba</dt>
<dd><p>Result from breakapart FISH for BCL2 locus</p></dd>

<dt>bcl2_cn</dt>
<dd><p>Result from copy number FISH for BCL2 locus</p></dd>

<dt>bcl6_ba</dt>
<dd><p>Result from breakapart FISH for BCL6 locus</p></dd>

<dt>bcl6_cn</dt>
<dd><p>Result from copy number FISH for BCL6 locus</p></dd>

<dt>time_since_diagnosis_years</dt>
<dd><p>TODO</p></dd>

<dt>relapse_timing</dt>
<dd><p>TODO</p></dd>

<dt>dtbx</dt>
<dd><p>TODO. OR REMOVE?</p></dd>

<dt>dtdx</dt>
<dd><p>TODO. OR REMOVE?</p></dd>

<dt>lymphgen_no_cnv</dt>
<dd><p>TODO</p></dd>

<dt>lymphgen_with_cnv</dt>
<dd><p>TODO</p></dd>

<dt>lymphgen_cnv_noA53</dt>
<dd><p>TODO</p></dd>

<dt>lymphgen_wright</dt>
<dd><p>The LymphGen call for this sample from Wright et all (if applicable)</p></dd>

<dt>fl_grade</dt>
<dd><p>TODO</p></dd>

<dt>transformation</dt>
<dd><p>TODO</p></dd>

<dt>relapse</dt>
<dd><p>TODO</p></dd>

<dt>ighv_mutation_original</dt>
<dd><p>TODO</p></dd>


</dl></div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>This function returns metadata for GAMBL samples.
This replaces the functionality of the original version, which is still available under the new name <a href="og_get_gambl_metadata.html">og_get_gambl_metadata</a>.
The purpose of this function is to provide the metadata for the non-redundant set of samples from GAMBL, dealing with all
types of redundancy caused by samples or biopsies that have data from &gt;1 seq_type (genome or capture), different capture protocols (exome or targeted capture) etc.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>

    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Ryan Morin, Kostiantyn Dreval, Vladimir Souza.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.9000.</p>
</div>

    </footer></div>





  </body></html>

