<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Get Coding SSM Status. — get_coding_ssm_status • GAMBLR.results</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Get Coding SSM Status. — get_coding_ssm_status"><meta property="og:description" content="Tabulate mutation status (SSM) for a set of genes."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">GAMBLR.results</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/Overview.html">What's new in GAMBLR</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Get Coding SSM Status.</h1>

    <div class="hidden name"><code>get_coding_ssm_status.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Tabulate mutation status (SSM) for a set of genes.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">get_coding_ssm_status</span><span class="op">(</span></span>
<span>  <span class="va">gene_symbols</span>,</span>
<span>  <span class="va">these_samples_metadata</span>,</span>
<span>  augmented <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  min_read_support <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  maf_path <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  <span class="va">maf_data</span>,</span>
<span>  include_hotspots <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  keep_multihit_hotspot <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  recurrence_min <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  review_hotspots <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  genes_of_interest <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"FOXO1"</span>, <span class="st">"MYD88"</span>, <span class="st">"CREBBP"</span><span class="op">)</span>,</span>
<span>  <span class="va">genome_build</span>,</span>
<span>  include_silent <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  <span class="va">include_silent_genes</span>,</span>
<span>  <span class="va">suffix</span>,</span>
<span>  <span class="va">this_seq_type</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>


<dl><dt id="arg-gene-symbols">gene_symbols<a class="anchor" aria-label="anchor" href="#arg-gene-symbols"></a></dt>
<dd><p>A vector of gene symbols for which the mutation status
will be tabulated. If not provided, lymphoma genes will be returned by
default.</p></dd>


<dt id="arg-these-samples-metadata">these_samples_metadata<a class="anchor" aria-label="anchor" href="#arg-these-samples-metadata"></a></dt>
<dd><p>The metadata for samples of interest to be
included in the returned matrix. Only the column "sample_id" is
required. If not provided, the matrix is tabulated for all available
samples as default.</p></dd>


<dt id="arg-augmented">augmented<a class="anchor" aria-label="anchor" href="#arg-augmented"></a></dt>
<dd><p>default: TRUE. Set to FALSE if you instead want the original
MAF from each sample for multi-sample patients instead of the augmented
MAF.</p></dd>


<dt id="arg-min-read-support">min_read_support<a class="anchor" aria-label="anchor" href="#arg-min-read-support"></a></dt>
<dd><p>Only returns variants with at least this many reads
in t_alt_count (for cleaning up augmented MAFs).</p></dd>


<dt id="arg-maf-path">maf_path<a class="anchor" aria-label="anchor" href="#arg-maf-path"></a></dt>
<dd><p>If the status of coding SSM should be tabulated from a custom
maf file, provide path to the maf in this argument. The default is set
to NULL.</p></dd>


<dt id="arg-maf-data">maf_data<a class="anchor" aria-label="anchor" href="#arg-maf-data"></a></dt>
<dd><p>Either a maf loaded from disk or from the database using a
get_ssm function.</p></dd>


<dt id="arg-include-hotspots">include_hotspots<a class="anchor" aria-label="anchor" href="#arg-include-hotspots"></a></dt>
<dd><p>Logical parameter indicating whether hotspots object
should also be tabulated. Default is TRUE.</p></dd>


<dt id="arg-keep-multihit-hotspot">keep_multihit_hotspot<a class="anchor" aria-label="anchor" href="#arg-keep-multihit-hotspot"></a></dt>
<dd><p>Logical parameter indicating whether to keep the
gene annotation as mutated when the gene has both hot spot and
non-hotspot mutation. Default is FALSE. If set to TRUE, will report the
number of non-hotspot mutations instead of tabulating for just mutation
presence.</p></dd>


<dt id="arg-recurrence-min">recurrence_min<a class="anchor" aria-label="anchor" href="#arg-recurrence-min"></a></dt>
<dd><p>Integer value indicating minimal recurrence level.</p></dd>


<dt id="arg-review-hotspots">review_hotspots<a class="anchor" aria-label="anchor" href="#arg-review-hotspots"></a></dt>
<dd><p>Logical parameter indicating whether hotspots object
should be reviewed to include functionally relevant mutations or rare
lymphoma-related genes. Default is TRUE.</p></dd>


<dt id="arg-genes-of-interest">genes_of_interest<a class="anchor" aria-label="anchor" href="#arg-genes-of-interest"></a></dt>
<dd><p>A vector of genes for hotspot review. Currently only
FOXO1, MYD88, and CREBBP are supported.</p></dd>


<dt id="arg-genome-build">genome_build<a class="anchor" aria-label="anchor" href="#arg-genome-build"></a></dt>
<dd><p>Reference genome build for the coordinates in the MAF
file. The default is hg19 genome build.</p></dd>


<dt id="arg-include-silent">include_silent<a class="anchor" aria-label="anchor" href="#arg-include-silent"></a></dt>
<dd><p>Logical parameter indicating whether to include silent
mutations into coding mutations. Default is FALSE.</p></dd>


<dt id="arg-include-silent-genes">include_silent_genes<a class="anchor" aria-label="anchor" href="#arg-include-silent-genes"></a></dt>
<dd><p>Optionally, provide a list of genes for which the
Silent variants to be considered. If provided, the Silent variants for
these genes will be included regardless of the include_silent argument.</p></dd>


<dt id="arg-suffix">suffix<a class="anchor" aria-label="anchor" href="#arg-suffix"></a></dt>
<dd><p>Optionally provide a character that will be appended to the end of each name</p></dd>


<dt id="arg-this-seq-type">this_seq_type<a class="anchor" aria-label="anchor" href="#arg-this-seq-type"></a></dt>
<dd><p>Deprecated. This is now determined from the metadata provided.</p></dd>


<dt id="arg-projection">projection<a class="anchor" aria-label="anchor" href="#arg-projection"></a></dt>
<dd><p>Specify projection (grch37 or hg38) of mutations. Default
is grch37.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>A data frame with tabulated mutation status.</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>This function takes a data frame (in MAF-like format) and converts
it to a binary one-hot encoded matrix of mutation status for either a set of
user-specified genes (via gene_symbols) or, if no genes are provided, default
to all lymphoma genes. The default behaviour is to assign each gene/sample_id
combination as mutated only if there is a protein coding mutation for that
sample in the MAF but this can be configured to use synonymous variants in
some (via include_silent_genes) or all (via include_silent) genes.
This function also has other filtering and convenience parameters giving
the user full control of the return. For more information, refer to the
parameter descriptions and examples.
Is this function not what you are looking for? Try one of the following,
similar, functions; <a href="get_coding_ssm.html">get_coding_ssm</a>,
get_ssm_by_patients, <a href="get_ssm_by_sample.html">get_ssm_by_sample</a>,
<a href="get_ssm_by_samples.html">get_ssm_by_samples</a>, <a href="get_ssm_by_region.html">get_ssm_by_region</a>,
<a href="get_ssm_by_regions.html">get_ssm_by_regions</a>.</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># FL Tier 1 and 2 genes</span></span></span>
<span class="r-in"><span><span class="va">genes</span> <span class="op">=</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="fu">GAMBLR.data</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/GAMBLR.data/man/lymphoma_genes_lymphoma_genes_v0.0.html" class="external-link">lymphoma_genes</a></span>,</span></span>
<span class="r-in"><span>                      <span class="va">FL</span><span class="op">==</span><span class="cn">TRUE</span><span class="op">)</span> <span class="op"><a href="pipe.html">%&gt;%</a></span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/pull.html" class="external-link">pull</a></span><span class="op">(</span><span class="va">Gene</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Metadata for FL genomes and exomes</span></span></span>
<span class="r-in"><span><span class="va">fl_meta</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="fu"><a href="get_gambl_metadata.html">get_gambl_metadata</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="pipe.html">%&gt;%</a></span> </span></span>
<span class="r-in"><span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">pathology</span><span class="op">==</span><span class="st">"FL"</span>,</span></span>
<span class="r-in"><span>                  <span class="va">cohort</span> <span class="op">!=</span> <span class="st">"FL_Crouch"</span>,</span></span>
<span class="r-in"><span>                  <span class="va">seq_type</span> <span class="op">!=</span> <span class="st">"mrna"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">fl_meta</span><span class="op">$</span><span class="va">seq_type</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> capture  genome </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     321     463 </span>
<span class="r-in"><span><span class="co"># Here, we let the function load the data for us</span></span></span>
<span class="r-in"><span> <span class="va">coding_tabulated_df</span> <span class="op">=</span> <span class="fu">get_coding_ssm_status</span><span class="op">(</span></span></span>
<span class="r-in"><span>  gene_symbols<span class="op">=</span><span class="va">genes</span>,</span></span>
<span class="r-in"><span>  include_hotspots<span class="op">=</span><span class="cn">FALSE</span>,</span></span>
<span class="r-in"><span>  genome_build <span class="op">=</span> <span class="st">"hg38"</span>,</span></span>
<span class="r-in"><span>  these_samples_metadata <span class="op">=</span> <span class="va">fl_meta</span></span></span>
<span class="r-in"><span> <span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Joining with `by = join_by(sample_id)`</span>
<span class="r-in"><span> <span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">coding_tabulated_df</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 784 116</span>
<span class="r-in"><span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">coding_tabulated_df</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "sample_id" "TNFRSF14"  "ARID1A"    "LAPTM5"    "MAP7D1"    "ZC3H12A"  </span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Alternatively, we can provide the MAF data directly</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Load the MAF data (let's use the other genome build this time)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">maf_data</span> <span class="op">=</span> <span class="fu"><a href="get_all_coding_ssm.html">get_all_coding_ssm</a></span><span class="op">(</span>these_samples_metadata <span class="op">=</span> <span class="va">fl_meta</span>,</span></span>
<span class="r-in"><span>                              projection <span class="op">=</span> <span class="st">"grch37"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">coding_tabulated2</span> <span class="op">=</span> <span class="fu">get_coding_ssm_status</span><span class="op">(</span>gene_symbols<span class="op">=</span><span class="va">genes</span>,</span></span>
<span class="r-in"><span>                                          these_samples_metadata <span class="op">=</span> <span class="va">fl_meta</span>,</span></span>
<span class="r-in"><span>                                          maf_data <span class="op">=</span> <span class="va">maf_data</span>,</span></span>
<span class="r-in"><span>                                          include_hotspots<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Joining with `by = join_by(sample_id)`</span>
<span class="r-in"><span> <span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">coding_tabulated2</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 784 118</span>
<span class="r-in"><span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">coding_tabulated2</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "sample_id" "TNFRSF14"  "ARID1A"    "LAPTM5"    "MAP7D1"    "ZC3H12A"  </span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Ryan Morin, Kostiantyn Dreval, Vladimir Souza.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

      </footer></div>






  </body></html>

