<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Get SSM By Regions. — get_ssm_by_regions • GAMBLR.results</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Get SSM By Regions. — get_ssm_by_regions"><meta name="description" content="Efficiently retrieve all mutations across a range of genomic regions."><meta property="og:description" content="Efficiently retrieve all mutations across a range of genomic regions."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">GAMBLR.results</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.3</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/GAMBLR-results.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Developers</h6></li>
    <li><a class="dropdown-item" href="../articles/testing.html">Testing GAMBLR packages</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Get SSM By Regions.</h1>

      <div class="d-none name"><code>get_ssm_by_regions.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Efficiently retrieve all mutations across a range of genomic regions.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">get_ssm_by_regions</span><span class="op">(</span></span>
<span>  <span class="va">regions_list</span>,</span>
<span>  <span class="va">regions_bed</span>,</span>
<span>  these_sample_ids <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  these_samples_metadata <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  streamlined <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  maf_data <span class="op">=</span> <span class="va">maf_data</span>,</span>
<span>  use_name_column <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  from_indexed_flatfile <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  mode <span class="op">=</span> <span class="st">"slms-3"</span>,</span>
<span>  augmented <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  this_seq_type <span class="op">=</span> <span class="st">"genome"</span>,</span>
<span>  projection <span class="op">=</span> <span class="st">"grch37"</span>,</span>
<span>  min_read_support <span class="op">=</span> <span class="fl">4</span>,</span>
<span>  basic_columns <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-regions-list">regions_list<a class="anchor" aria-label="anchor" href="#arg-regions-list"></a></dt>
<dd><p>Either provide a vector of regions in the chr:start-end format OR.</p></dd>


<dt id="arg-regions-bed">regions_bed<a class="anchor" aria-label="anchor" href="#arg-regions-bed"></a></dt>
<dd><p>Better yet, provide a bed file with the coordinates you want to retrieve.</p></dd>


<dt id="arg-these-sample-ids">these_sample_ids<a class="anchor" aria-label="anchor" href="#arg-these-sample-ids"></a></dt>
<dd><p>Optional, a vector of multiple sample_id (or a single sample ID as a string) that you want results for.</p></dd>


<dt id="arg-these-samples-metadata">these_samples_metadata<a class="anchor" aria-label="anchor" href="#arg-these-samples-metadata"></a></dt>
<dd><p>Optional, a metadata table (with sample IDs in a column) to subset the return to.</p></dd>


<dt id="arg-streamlined">streamlined<a class="anchor" aria-label="anchor" href="#arg-streamlined"></a></dt>
<dd><p>If TRUE (default), only 3 columns will be kept in the maf (start, sample_id and region name).
To return more columns, set this parameter to FALSE, see <code>basic_column</code> for more info.
Note, if this parameter is TRUE, the function will disregard anything specified with <code>basic_columns</code>.</p></dd>


<dt id="arg-maf-data">maf_data<a class="anchor" aria-label="anchor" href="#arg-maf-data"></a></dt>
<dd><p>Use an already loaded MAF data frame.</p></dd>


<dt id="arg-use-name-column">use_name_column<a class="anchor" aria-label="anchor" href="#arg-use-name-column"></a></dt>
<dd><p>If your bed-format data frame has a name column (must be named "name") these can be used to name your regions.</p></dd>


<dt id="arg-from-indexed-flatfile">from_indexed_flatfile<a class="anchor" aria-label="anchor" href="#arg-from-indexed-flatfile"></a></dt>
<dd><p>Set to TRUE to avoid using the database and instead rely on flatfiles (only works for streamlined data, not full MAF details).</p></dd>


<dt id="arg-mode">mode<a class="anchor" aria-label="anchor" href="#arg-mode"></a></dt>
<dd><p>Only works with indexed flatfiles. Accepts 2 options of "slms-3" and "strelka2" to indicate which variant caller to use. Default is "slms-3".</p></dd>


<dt id="arg-augmented">augmented<a class="anchor" aria-label="anchor" href="#arg-augmented"></a></dt>
<dd><p>default: TRUE. Set to FALSE if you instead want the original MAF from each sample for multi-sample patients instead of the augmented MAF</p></dd>


<dt id="arg-this-seq-type">this_seq_type<a class="anchor" aria-label="anchor" href="#arg-this-seq-type"></a></dt>
<dd><p>The seq_type you want back, default is genome.</p></dd>


<dt id="arg-projection">projection<a class="anchor" aria-label="anchor" href="#arg-projection"></a></dt>
<dd><p>Obtain variants projected to this reference (one of grch37 or hg38).</p></dd>


<dt id="arg-min-read-support">min_read_support<a class="anchor" aria-label="anchor" href="#arg-min-read-support"></a></dt>
<dd><p>Only returns variants with at least this many reads in t_alt_count (for cleaning up augmented MAFs).</p></dd>


<dt id="arg-basic-columns">basic_columns<a class="anchor" aria-label="anchor" href="#arg-basic-columns"></a></dt>
<dd><p>Parameter to be used when streamlined is FALSE.
Set this parameter to TRUE for returning a maf with standard 45 columns, set to FALSE to keep all 116 maf columns in the returned object.
To return all 116 maf columns, set this parameter to FALSE.</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>Boolean parameter set to FALSE per default.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>Returns a data frame of variants in MAF-like format.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>This function internally calls <a href="get_ssm_by_region.html">get_ssm_by_region</a> to retrieve SSM calls for the specified regions.
See parameter descriptions for <a href="get_ssm_by_region.html">get_ssm_by_region</a> for more information on how the different parameters can be called.
Is this function not what you are looking for? Try one of the following, similar, functions; <a href="get_coding_ssm.html">get_coding_ssm</a>,
<a href="get_coding_ssm_status.html">get_coding_ssm_status</a>, <a href="get_ssm_by_sample.html">get_ssm_by_sample</a>,
<a href="get_ssm_by_samples.html">get_ssm_by_samples</a>, <a href="get_ssm_by_region.html">get_ssm_by_region</a></p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">regions_bed</span> <span class="op">=</span> <span class="fu">GAMBLR.utils</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/GAMBLR.utils/man/create_bed_data.html" class="external-link">create_bed_data</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>   <span class="fu">GAMBLR.data</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/GAMBLR.data/man/grch37_ashm_regions.html" class="external-link">grch37_ashm_regions</a></span>,</span></span>
<span class="r-in"><span>   fix_names <span class="op">=</span> <span class="st">"concat"</span>,</span></span>
<span class="r-in"><span>   concat_cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"gene"</span>,<span class="st">"region"</span><span class="op">)</span>,sep<span class="op">=</span><span class="st">"-"</span></span></span>
<span class="r-in"><span><span class="op">)</span> <span class="op"><a href="pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fl">20</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">DLBCL_meta</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="fu"><a href="get_gambl_metadata.html">get_gambl_metadata</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="pipe.html">%&gt;%</a></span> </span></span>
<span class="r-in"><span>                <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">pathology</span><span class="op">==</span><span class="st">"DLBCL"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">ashm_MAF</span> <span class="op">=</span> <span class="fu">get_ssm_by_regions</span><span class="op">(</span>regions_bed <span class="op">=</span> <span class="va">regions_bed</span>,</span></span>
<span class="r-in"><span>                             these_samples_metadata <span class="op">=</span> <span class="va">DLBCL_meta</span>,</span></span>
<span class="r-in"><span>                             streamlined<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span></span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>One or more parsing issues, call `problems()` on your data frame for details,</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> e.g.:</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   dat &lt;- vroom(...)</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   problems(dat)</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>One or more parsing issues, call `problems()` on your data frame for details,</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> e.g.:</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   dat &lt;- vroom(...)</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   problems(dat)</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>One or more parsing issues, call `problems()` on your data frame for details,</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> e.g.:</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   dat &lt;- vroom(...)</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   problems(dat)</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>One or more parsing issues, call `problems()` on your data frame for details,</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> e.g.:</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   dat &lt;- vroom(...)</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   problems(dat)</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>One or more parsing issues, call `problems()` on your data frame for details,</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> e.g.:</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   dat &lt;- vroom(...)</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   problems(dat)</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>One or more parsing issues, call `problems()` on your data frame for details,</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> e.g.:</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   dat &lt;- vroom(...)</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   problems(dat)</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>One or more parsing issues, call `problems()` on your data frame for details,</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> e.g.:</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   dat &lt;- vroom(...)</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   problems(dat)</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>One or more parsing issues, call `problems()` on your data frame for details,</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> e.g.:</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   dat &lt;- vroom(...)</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   problems(dat)</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>One or more parsing issues, call `problems()` on your data frame for details,</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> e.g.:</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   dat &lt;- vroom(...)</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   problems(dat)</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>One or more parsing issues, call `problems()` on your data frame for details,</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> e.g.:</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   dat &lt;- vroom(...)</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   problems(dat)</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>One or more parsing issues, call `problems()` on your data frame for details,</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> e.g.:</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   dat &lt;- vroom(...)</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   problems(dat)</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>One or more parsing issues, call `problems()` on your data frame for details,</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> e.g.:</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   dat &lt;- vroom(...)</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   problems(dat)</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>One or more parsing issues, call `problems()` on your data frame for details,</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> e.g.:</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   dat &lt;- vroom(...)</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   problems(dat)</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>One or more parsing issues, call `problems()` on your data frame for details,</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> e.g.:</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   dat &lt;- vroom(...)</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   problems(dat)</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>One or more parsing issues, call `problems()` on your data frame for details,</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> e.g.:</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   dat &lt;- vroom(...)</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   problems(dat)</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>One or more parsing issues, call `problems()` on your data frame for details,</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> e.g.:</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   dat &lt;- vroom(...)</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   problems(dat)</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>One or more parsing issues, call `problems()` on your data frame for details,</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> e.g.:</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   dat &lt;- vroom(...)</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   problems(dat)</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>One or more parsing issues, call `problems()` on your data frame for details,</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> e.g.:</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   dat &lt;- vroom(...)</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   problems(dat)</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>One or more parsing issues, call `problems()` on your data frame for details,</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> e.g.:</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   dat &lt;- vroom(...)</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   problems(dat)</span>
<span class="r-in"><span><span class="va">ashm_MAF</span> <span class="op"><a href="pipe.html">%&gt;%</a></span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="va">Start_Position</span>,<span class="va">Tumor_Sample_Barcode</span><span class="op">)</span> <span class="op"><a href="pipe.html">%&gt;%</a></span></span></span>
<span class="r-in"><span>              <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">Hugo_Symbol</span>,</span></span>
<span class="r-in"><span>                    <span class="va">Tumor_Sample_Barcode</span>,</span></span>
<span class="r-in"><span>                    <span class="va">Chromosome</span>,<span class="va">Start_Position</span>,</span></span>
<span class="r-in"><span>                    <span class="va">Reference_Allele</span>,</span></span>
<span class="r-in"><span>                    <span class="va">Tumor_Seq_Allele2</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> genomic_data Object</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Genome Build: grch37 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Showing first 10 rows:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    Hugo_Symbol      Tumor_Sample_Barcode Chromosome Start_Position</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1       KLHL21           13-26835_tumorA          1        6661537</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2       KLHL21           13-26835_tumorB          1        6661537</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3       KLHL21           13-26835_tumorD          1        6661537</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4       KLHL21                  SP193546          1        6661538</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5       KLHL21 HTMCP-01-06-00497-01A-01D          1        6661563</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6       KLHL21           17-40409_tumorA          1        6661575</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 7       KLHL21           17-40409_tumorB          1        6661575</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 8       KLHL21 HTMCP-01-06-00136-01A-01D          1        6661604</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 9       KLHL21                 15-26538T          1        6661607</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 10      KLHL21                 10-18191T          1        6661655</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    Reference_Allele Tumor_Seq_Allele2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1                 A                 T</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2                 A                 T</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3                 A                 T</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4                 C                 G</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5                 G                 C</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6                 C                 T</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 7                 C                 T</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 8                 G                 C</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 9                 G                 A</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 10                A                 G</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Ryan Morin, Kostiantyn Dreval, Vladimir Souza.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.9000.</p>
</div>

    </footer></div>





  </body></html>

