<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Get Gene Expression. — get_gene_expression • GAMBLR.results</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Get Gene Expression. — get_gene_expression"><meta name="description" content="Get the expression for one or more genes for all GAMBL samples."><meta property="og:description" content="Get the expression for one or more genes for all GAMBL samples."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">GAMBLR.results</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.3</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/GAMBLR-results.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Developers</h6></li>
    <li><a class="dropdown-item" href="../articles/testing.html">Testing GAMBLR packages</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Get Gene Expression.</h1>

      <div class="d-none name"><code>get_gene_expression.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Get the expression for one or more genes for all GAMBL samples.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">get_gene_expression</span><span class="op">(</span></span>
<span>  <span class="va">these_samples_metadata</span>,</span>
<span>  <span class="va">hugo_symbols</span>,</span>
<span>  <span class="va">ensembl_gene_ids</span>,</span>
<span>  all_genes <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  engine <span class="op">=</span> <span class="st">"grep"</span>,</span>
<span>  format <span class="op">=</span> <span class="st">"wide"</span>,</span>
<span>  lazy_join <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  arbitrarily_pick <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  HGNC <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-these-samples-metadata">these_samples_metadata<a class="anchor" aria-label="anchor" href="#arg-these-samples-metadata"></a></dt>
<dd><p>The data frame with sample metadata.
Usually output of the get_gambl_metadata().</p></dd>


<dt id="arg-hugo-symbols">hugo_symbols<a class="anchor" aria-label="anchor" href="#arg-hugo-symbols"></a></dt>
<dd><p>One or more gene symbols.
Cannot be used in conjunction with ensembl_gene_ids.</p></dd>


<dt id="arg-ensembl-gene-ids">ensembl_gene_ids<a class="anchor" aria-label="anchor" href="#arg-ensembl-gene-ids"></a></dt>
<dd><p>One or more ensembl gene IDs.
Cannot be used in conjunction with hugo_symbols.</p></dd>


<dt id="arg-all-genes">all_genes<a class="anchor" aria-label="anchor" href="#arg-all-genes"></a></dt>
<dd><p>Set to TRUE for the full expression data without
any subsetting (<em>see warnings below</em>).</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>Set to TRUE for a more chatty output</p></dd>


<dt id="arg-engine">engine<a class="anchor" aria-label="anchor" href="#arg-engine"></a></dt>
<dd><p>Either readr or grep. The grep engine usually will increase
the speed of loading but doesn't work if you want all genes or a very
long list.</p></dd>


<dt id="arg-format">format<a class="anchor" aria-label="anchor" href="#arg-format"></a></dt>
<dd><p>Either <code>wide</code> or <code>long</code>. Wide format returns one column
of expression values per gene. Long format returns one column of expression
values with the gene stored in a separate column.</p></dd>


<dt id="arg-lazy-join">lazy_join<a class="anchor" aria-label="anchor" href="#arg-lazy-join"></a></dt>
<dd><p>If TRUE, your data frame will also have capture_sample_id
and genome_sample_id columns provided. See <code>check_gene_expression</code> for more
information.</p></dd>


<dt id="arg-arbitrarily-pick">arbitrarily_pick<a class="anchor" aria-label="anchor" href="#arg-arbitrarily-pick"></a></dt>
<dd><p>A stop-gap for handling the rare scenario where
the same Hugo_Symbol has more than one ensembl_gene_id. Set to TRUE only
if you encounter an error that states "Values are not uniquely identified;
output will contain list-cols."</p></dd>


<dt id="arg-hgnc">HGNC<a class="anchor" aria-label="anchor" href="#arg-hgnc"></a></dt>
<dd><p>When you request the wide matrix and all genes, this forces
the columns to contain hgnc_id rather than ensembl_gene_id</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>Optional parameters to pass along to <code>get_gambl_metadata</code> (only used in conjunction with lazy_join)</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A data frame with the first 9 columns identical to the columns from check_gene_expression and the remaining columns containing the expression values for each gene requested.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>Efficiently retrieve variance-stabilized and batch effect
corrected gene expression values for one, multiple or all genes
for all GAMBL samples.
For more information, refer to the parameter descriptions and examples.</p>
<p>Warnings:</p><ol><li><p>The speed of loading data is heavily impacted by how many samples
you load. For the sake of efficiency, be sure not to specify
extraneous samples.</p></li>
<li><p>To reduce impact on memory (RAM), load only the data for the genes
you need.</p></li>
<li><p>Combining lazy_join with all_genes will result in a data table
with samples on rows and genes on columns. <em>Use with caution</em>.
This is practically guaranteed to use more RAM than you want.</p></li>
<li><p>Before you run this function, it's recommended that you run
<code>check_gene_expression</code> to determine which samples are available</p></li>
</ol></div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>

    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Ryan Morin, Kostiantyn Dreval, Vladimir Souza.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.9000.</p>
</div>

    </footer></div>





  </body></html>

