<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Get SSM By Sample. — get_ssm_by_sample • GAMBLR.results</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Get SSM By Sample. — get_ssm_by_sample"><meta property="og:description" content="Get the SSMs (i.e. load MAF) for a single sample."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">GAMBLR.results</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Get SSM By Sample.</h1>
    
    <div class="hidden name"><code>get_ssm_by_sample.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Get the SSMs (i.e. load MAF) for a single sample.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">get_ssm_by_sample</span><span class="op">(</span></span>
<span>  <span class="va">these_samples_metadata</span>,</span>
<span>  tool_name <span class="op">=</span> <span class="st">"slms-3"</span>,</span>
<span>  projection <span class="op">=</span> <span class="st">"grch37"</span>,</span>
<span>  augmented <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  flavour <span class="op">=</span> <span class="st">"clustered"</span>,</span>
<span>  min_read_support <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  basic_columns <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  maf_cols <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  <span class="va">this_sample_id</span>,</span>
<span>  <span class="va">this_seq_type</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>these_samples_metadata</dt>
<dd><p>Required if not specifying both this_sample_id and this_seq_type a single row or entire metadata table containing your sample_id.</p></dd>


<dt>tool_name</dt>
<dd><p>The name of the variant calling pipeline (currently only slms-3 is supported).</p></dd>


<dt>projection</dt>
<dd><p>The projection genome build. Supports hg38 and grch37.</p></dd>


<dt>augmented</dt>
<dd><p>default: TRUE. Set to FALSE if you instead want the original MAF from each sample for multi-sample patients instead of the augmented MAF.</p></dd>


<dt>flavour</dt>
<dd><p>Currently this function only supports one flavour option but this feature is meant for eventual compatibility with additional variant calling parameters/versions.</p></dd>


<dt>min_read_support</dt>
<dd><p>Only returns variants with at least this many reads in t_alt_count (for cleaning up augmented MAFs).</p></dd>


<dt>basic_columns</dt>
<dd><p>Return first 43 columns of MAF rather than full details. Default is TRUE.</p></dd>


<dt>maf_cols</dt>
<dd><p>if basic_columns is set to FALSE, the user can specify what columns to be returned within the MAF. This parameter can either be a vector of indexes (integer) or a vector of characters.</p></dd>


<dt>verbose</dt>
<dd><p>Enable for debugging/noisier output.</p></dd>


<dt>this_sample_id</dt>
<dd><p>Deprecated. Inferred from these_samples_metadata</p></dd>


<dt>this_seq_type</dt>
<dd><p>Deprecated. Inferred from these_samples_metadata</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>data frame in MAF format.</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>This was implemented to allow flexibility because there are some samples that we may want to use a different set of variants than those in the main GAMBL merge.
The current use case is to allow a force_unmatched output to be used to replace the SSMs from the merge for samples with known contamination in the normal.
This will also be useful to apply a blacklist to individual MAFs when coupled with <a href="annotate_ssm_blacklist.html">annotate_ssm_blacklist</a>.
Is this function not what you are looking for? Try one of the following, similar, functions; <a href="get_coding_ssm.html">get_coding_ssm</a>, <a href="get_coding_ssm_status.html">get_coding_ssm_status</a>,
get_ssm_by_patients, <a href="get_ssm_by_samples.html">get_ssm_by_samples</a>, <a href="get_ssm_by_region.html">get_ssm_by_region</a>, <a href="get_ssm_by_regions.html">get_ssm_by_regions</a></p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">maf_samp</span> <span class="op">=</span> <span class="fu">GAMBLR.results</span><span class="fu">:::</span><span class="fu">get_ssm_by_sample</span><span class="op">(</span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="get_gambl_metadata.html">get_gambl_metadata</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="pipe.html">%&gt;%</a></span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">sample_id</span><span class="op">==</span><span class="st">"13-27975_tumorA"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  augmented <span class="op">=</span> <span class="cn">FALSE</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 3258 capture samples are missing a value for protocol. Assuming Exome.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 123 biopsies are missing from the biopsy metadata. This should be fixed!</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> affected cohorts:  DLBCL_LSARP_Trios,SMZL_Strefford,cHL_Maura,MCL_Barcelona</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 110 biopsies with discrepancies in the pathology field. This should be fixed!</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 10 biopsies with discrepancies in the time_point field. This should be fixed!</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">maf_samp</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 4705</span>
<span class="r-in"><span><span class="va">maf_samp_aug</span> <span class="op">=</span> <span class="fu">GAMBLR.results</span><span class="fu">:::</span><span class="fu">get_ssm_by_sample</span><span class="op">(</span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="get_gambl_metadata.html">get_gambl_metadata</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="pipe.html">%&gt;%</a></span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">sample_id</span><span class="op">==</span><span class="st">"13-27975_tumorA"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  augmented <span class="op">=</span> <span class="cn">TRUE</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 3258 capture samples are missing a value for protocol. Assuming Exome.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 123 biopsies are missing from the biopsy metadata. This should be fixed!</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> affected cohorts:  DLBCL_LSARP_Trios,SMZL_Strefford,cHL_Maura,MCL_Barcelona</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 110 biopsies with discrepancies in the pathology field. This should be fixed!</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 10 biopsies with discrepancies in the time_point field. This should be fixed!</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">maf_samp_aug</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 6118</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span> <span class="va">some_maf</span> <span class="op">=</span> <span class="fu">GAMBLR.results</span><span class="fu">:::</span><span class="fu">get_ssm_by_sample</span><span class="op">(</span></span></span>
<span class="r-in"><span>                          these_samples_metadata <span class="op">=</span> <span class="fu"><a href="get_gambl_metadata.html">get_gambl_metadata</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="pipe.html">%&gt;%</a></span></span></span>
<span class="r-in"><span>                            <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">sample_id</span> <span class="op">==</span> <span class="st">"HTMCP-01-06-00485-01A-01D"</span>,</span></span>
<span class="r-in"><span>                                     <span class="va">seq_type</span> <span class="op">==</span> <span class="st">"genome"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                         projection <span class="op">=</span> <span class="st">"hg38"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 3258 capture samples are missing a value for protocol. Assuming Exome.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 123 biopsies are missing from the biopsy metadata. This should be fixed!</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> affected cohorts:  DLBCL_LSARP_Trios,SMZL_Strefford,cHL_Maura,MCL_Barcelona</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 110 biopsies with discrepancies in the pathology field. This should be fixed!</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 10 biopsies with discrepancies in the time_point field. This should be fixed!</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>The following named parsers don't match the column names: GENE_PHENO, FILTER, flanking_bps, vcf_id, vcf_qual, gnomAD_AF, gnomAD_AFR_AF, gnomAD_AMR_AF, gnomAD_SAS_AF, vcf_pos, gnomADg_AF, blacklist_count</span>
<span class="r-in"><span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">some_maf</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> genomic_data Object</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Genome Build: hg38 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Showing first 10 rows:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>        Hugo_Symbol Entrez_Gene_Id Center NCBI_Build Chromosome Start_Position</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1          Unknown              0      .     GRCh38       chr1        5446963</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2           CAMTA1              0      .     GRCh38       chr1        7244319</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3             PER3              0      .     GRCh38       chr1        7828459</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4          TNFRSF9              0      .     GRCh38       chr1        7911008</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5  ENSR00000000893              0      .     GRCh38       chr1        8181883</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6           PIK3CD              0      .     GRCh38       chr1        9726972</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 7            CASZ1              0      .     GRCh38       chr1       10720942</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 8           CELA2A              0      .     GRCh38       chr1       15454068</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 9          PLA2G2A              0      .     GRCh38       chr1       19975558</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 10         PLA2G2C              0      .     GRCh38       chr1       20172387</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    End_Position Strand Variant_Classification Variant_Type</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1       5446963      +                    IGR          SNP</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2       7244319      +                 Intron          SNP</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3       7828459      +                 Intron          SNP</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4       7911008      +                3'Flank          SNP</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5       8181883      +                    IGR          SNP</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6       9726972      +      Missense_Mutation          SNP</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 7      10720942      +                 Intron          SNP</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 8      15454068      +                5'Flank          SNP</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 9      19975559      +                  3'UTR          DEL</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 10     20172387      +                 Intron          SNP</span>
<span class="r-in"><span> </span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Ryan Morin, Kostiantyn Dreval, Vladimir Souza.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

