<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Get Raw Expression Counts — get_raw_expression_counts • GAMBLR.results</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Get Raw Expression Counts — get_raw_expression_counts"><meta property="og:description" content="Get the raw read counts from RNA-seq for one or more genes for all GAMBL samples."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">GAMBLR.results</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/Overview.html">What's new in GAMBLR</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Get Raw Expression Counts</h1>

    <div class="hidden name"><code>get_raw_expression_counts.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Get the raw read counts from RNA-seq for one or more genes for all GAMBL samples.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">get_raw_expression_counts</span><span class="op">(</span></span>
<span>  <span class="va">these_samples_metadata</span>,</span>
<span>  <span class="va">existing_sample_id_column</span>,</span>
<span>  <span class="va">new_sample_id_column</span>,</span>
<span>  all_samples <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  check <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  from_flatfile <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  map_to_symbol <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  TPM <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>


<dl><dt id="arg-these-samples-metadata">these_samples_metadata<a class="anchor" aria-label="anchor" href="#arg-these-samples-metadata"></a></dt>
<dd><p>The data frame with sample metadata. Usually output of the get_gambl_metadata().</p></dd>


<dt id="arg-existing-sample-id-column">existing_sample_id_column<a class="anchor" aria-label="anchor" href="#arg-existing-sample-id-column"></a></dt>
<dd><p>Specify which column contains the sample_id you want replaced with the contents of new_sample_id_column</p></dd>


<dt id="arg-new-sample-id-column">new_sample_id_column<a class="anchor" aria-label="anchor" href="#arg-new-sample-id-column"></a></dt>
<dd><p>Specify which column in your metadata contains the sample_id you want used instead of the existing_sample_id</p></dd>


<dt id="arg-all-samples">all_samples<a class="anchor" aria-label="anchor" href="#arg-all-samples"></a></dt>
<dd><p>Set to TRUE to force the function to return all available data (should rarely be necessary)</p></dd>


<dt id="arg-check">check<a class="anchor" aria-label="anchor" href="#arg-check"></a></dt>
<dd><p>For basic debugging. Set to TRUE to obtain basic information about the number of samples in your metadata with expression data available</p></dd>


<dt id="arg-from-flatfile">from_flatfile<a class="anchor" aria-label="anchor" href="#arg-from-flatfile"></a></dt>
<dd><p>Set to FALSE to use the database instead of reading from flatfiles</p></dd>


<dt id="arg-map-to-symbol">map_to_symbol<a class="anchor" aria-label="anchor" href="#arg-map-to-symbol"></a></dt>
<dd><p>Set to TRUE to obtain the mappings between the rows in the count matrix and HGNC gene symbol/alias</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>Set to TRUE for more print statements and such</p></dd>


<dt id="arg-tpm">TPM<a class="anchor" aria-label="anchor" href="#arg-tpm"></a></dt>
<dd><p>Set to TRUE to get TPM estimates instead of counts</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>A list containing a counts matrix and the associated metadata for DESeq2</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>Efficiently retrieve raw gene expression values (read counts) for one, multiple or all genes for all GAMBL samples.
For examples and more info, refer to the parameter descriptions as well as vignette examples.</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span> <span class="co"># \dontrun{</span></span></span>
<span class="r-in"><span><span class="va">schmitz_meta</span> <span class="op">=</span> <span class="fu"><a href="get_gambl_metadata.html">get_gambl_metadata</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="pipe.html">%&gt;%</a></span> </span></span>
<span class="r-in"><span>    <span class="fu"><a href="https://rdrr.io/r/stats/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">seq_type</span><span class="op">==</span><span class="st">"mrna"</span>,<span class="va">cohort</span><span class="op">==</span><span class="st">"dlbcl_schmitz"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">exp_out</span> <span class="op">=</span> <span class="fu">get_raw_expression_counts</span><span class="op">(</span>these_samples_metadata <span class="op">=</span> <span class="va">schmitz_meta</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Create DESeq data set directly from the two named objects in the output</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">dds</span> <span class="op">&lt;-</span> <span class="fu">DESeqDataSetFromMatrix</span><span class="op">(</span>countData <span class="op">=</span> <span class="va">exp_out</span><span class="op">$</span><span class="va">counts</span>,</span></span>
<span class="r-in"><span>    colData <span class="op">=</span> <span class="va">exp_out</span><span class="op">$</span><span class="va">metadata</span>,</span></span>
<span class="r-in"><span>    design <span class="op">=</span> <span class="op">~</span> <span class="va">COO_consensus</span> <span class="op">+</span> <span class="va">sex</span><span class="op">)</span></span></span>
<span class="r-in"><span>    </span></span>
<span class="r-in"><span><span class="co"># Run a basic DESeq analysis</span></span></span>
<span class="r-in"><span><span class="va">dds</span> <span class="op">&lt;-</span> <span class="fu">DESeq</span><span class="op">(</span><span class="va">dds</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu">results</span><span class="op">(</span><span class="va">dds</span>, </span></span>
<span class="r-in"><span>    name<span class="op">=</span><span class="st">"COO_consensus_GCB_vs_ABC"</span>,</span></span>
<span class="r-in"><span>    lfcThreshold<span class="op">=</span><span class="fl">2</span>,alpha<span class="op">=</span><span class="fl">0.1</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># Filter outputs using padj, logFC and baseMean (more highly expressed overall)     </span></span></span>
<span class="r-in"><span><span class="va">res_df</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span> <span class="op"><a href="pipe.html">%&gt;%</a></span> </span></span>
<span class="r-in"><span>    <span class="fu"><a href="https://rdrr.io/r/stats/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">padj</span><span class="op">&lt;</span><span class="fl">0.1</span>,<span class="va">baseMean</span><span class="op">&gt;</span><span class="fl">500</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">show_genes</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">res_df</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">vsd</span> <span class="op">&lt;-</span> <span class="fu">vst</span><span class="op">(</span><span class="va">dds</span>, blind<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">#Visualize the results with a heatmap</span></span></span>
<span class="r-in"><span><span class="va">column_ha</span> <span class="op">=</span> <span class="fu">HeatmapAnnotation</span><span class="op">(</span>df<span class="op">=</span><span class="fu">select</span><span class="op">(</span><span class="va">exp_out</span><span class="op">$</span><span class="va">metadata</span>,<span class="va">COO_consensus</span>,<span class="va">sex</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">Heatmap</span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">vsd</span><span class="op">)</span><span class="op">[</span><span class="va">show_genes</span>,<span class="op">]</span>,</span></span>
<span class="r-in"><span>    row_names_gp <span class="op">=</span> <span class="fu">gpar</span><span class="op">(</span>fontsize<span class="op">=</span><span class="fl">5</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    bottom_annotation <span class="op">=</span> <span class="va">column_ha</span>,</span></span>
<span class="r-in"><span>    show_column_names <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span> <span class="co"># }</span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Ryan Morin, Kostiantyn Dreval, Vladimir Souza.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

      </footer></div>






  </body></html>

