<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Get Coding SSM. — get_coding_ssm • GAMBLR.results</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Get Coding SSM. — get_coding_ssm"><meta name="description" content="Retrieve all coding SSMs from one seq_type in GAMBL
in MAF-like format."><meta property="og:description" content="Retrieve all coding SSMs from one seq_type in GAMBL
in MAF-like format."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">GAMBLR.results</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.3</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/GAMBLR-results.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Developers</h6></li>
    <li><a class="dropdown-item" href="../articles/testing.html">Testing GAMBLR packages</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Get Coding SSM.</h1>

      <div class="d-none name"><code>get_coding_ssm.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Retrieve all coding SSMs from one seq_type in GAMBL
in MAF-like format.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">get_coding_ssm</span><span class="op">(</span></span>
<span>  these_samples_metadata <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  <span class="va">force_unmatched_samples</span>,</span>
<span>  projection <span class="op">=</span> <span class="st">"grch37"</span>,</span>
<span>  this_seq_type <span class="op">=</span> <span class="st">"genome"</span>,</span>
<span>  basic_columns <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  maf_cols <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  augmented <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  min_read_support <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  groups <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"gambl"</span>, <span class="st">"icgc_dart"</span><span class="op">)</span>,</span>
<span>  include_silent <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  <span class="va">engine</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  <span class="va">limit_cohort</span>,</span>
<span>  <span class="va">exclude_cohort</span>,</span>
<span>  <span class="va">limit_pathology</span>,</span>
<span>  <span class="va">limit_samples</span>,</span>
<span>  <span class="va">from_flatfile</span>,</span>
<span>  <span class="va">these_sample_ids</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-these-samples-metadata">these_samples_metadata<a class="anchor" aria-label="anchor" href="#arg-these-samples-metadata"></a></dt>
<dd><p>Optional (but highly recommended) metadata
table to tell the function how to subset the data. Only sample_id in
this table with the matching seq_type will be in the output. Not all
samples may be in the output, though, if they are missing SSM results
or had no mutations detected.</p></dd>


<dt id="arg-force-unmatched-samples">force_unmatched_samples<a class="anchor" aria-label="anchor" href="#arg-force-unmatched-samples"></a></dt>
<dd><p>Optional argument for forcing unmatched
samples, using <a href="get_ssm_by_samples.html">get_ssm_by_samples</a>.</p></dd>


<dt id="arg-projection">projection<a class="anchor" aria-label="anchor" href="#arg-projection"></a></dt>
<dd><p>Reference genome build for the coordinates in the MAF file.
The default is grch37.</p></dd>


<dt id="arg-this-seq-type">this_seq_type<a class="anchor" aria-label="anchor" href="#arg-this-seq-type"></a></dt>
<dd><p>The seq_type you want SSMs from, default is genome.</p></dd>


<dt id="arg-basic-columns">basic_columns<a class="anchor" aria-label="anchor" href="#arg-basic-columns"></a></dt>
<dd><p>Basic columns refers to the first 45 standard MAF
columns. Set this to FALSE if you want all the available columns instead.</p></dd>


<dt id="arg-maf-cols">maf_cols<a class="anchor" aria-label="anchor" href="#arg-maf-cols"></a></dt>
<dd><p>if basic_columns is set to FALSE, the user can specify
which columns to be returned within the MAF. This parameter can either
be a vector of indexes (integer) or a vector of characters (matching
columns in MAF).</p></dd>


<dt id="arg-augmented">augmented<a class="anchor" aria-label="anchor" href="#arg-augmented"></a></dt>
<dd><p>Set to FALSE if you instead want the original MAF from
each sample for multi-sample patients instead of the augmented MAF</p></dd>


<dt id="arg-min-read-support">min_read_support<a class="anchor" aria-label="anchor" href="#arg-min-read-support"></a></dt>
<dd><p>Only returns variants with at least this many
reads in t_alt_count (for cleaning up augmented MAFs)</p></dd>


<dt id="arg-groups">groups<a class="anchor" aria-label="anchor" href="#arg-groups"></a></dt>
<dd><p>Deprecated. Use these_samples_metadata instead.</p></dd>


<dt id="arg-include-silent">include_silent<a class="anchor" aria-label="anchor" href="#arg-include-silent"></a></dt>
<dd><p>Logical indicating whether to include
silent mutations in coding regions (i.e. synonymous). Default is TRUE.</p></dd>


<dt id="arg-engine">engine<a class="anchor" aria-label="anchor" href="#arg-engine"></a></dt>
<dd><p>Deprecated. Ignored</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>Controls the "verboseness" of this function (and
internally called helpers).</p></dd>


<dt id="arg-limit-cohort">limit_cohort<a class="anchor" aria-label="anchor" href="#arg-limit-cohort"></a></dt>
<dd><p>Deprecated. Use these_samples_metadata instead.</p></dd>


<dt id="arg-exclude-cohort">exclude_cohort<a class="anchor" aria-label="anchor" href="#arg-exclude-cohort"></a></dt>
<dd><p>Deprecated. Use these_samples_metadata instead.</p></dd>


<dt id="arg-limit-pathology">limit_pathology<a class="anchor" aria-label="anchor" href="#arg-limit-pathology"></a></dt>
<dd><p>Deprecated. Use these_samples_metadata instead.</p></dd>


<dt id="arg-limit-samples">limit_samples<a class="anchor" aria-label="anchor" href="#arg-limit-samples"></a></dt>
<dd><p>Deprecated. Use these_samples_metadata instead.</p></dd>


<dt id="arg-from-flatfile">from_flatfile<a class="anchor" aria-label="anchor" href="#arg-from-flatfile"></a></dt>
<dd><p>Deprecated. Ignored</p></dd>


<dt id="arg-these-sample-ids">these_sample_ids<a class="anchor" aria-label="anchor" href="#arg-these-sample-ids"></a></dt>
<dd><p>Deprecated. Use these_samples_metadata instead.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A data frame containing all the MAF data columns
(one row per mutation).</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>Effectively retrieve simple somatic mutations (SSM) results for
either capture or genome seq_type (but not both at once).
The resulting data frame will be a maf_data object, which tracks the
genome build (projection) for the variants and will have a
maf_seq_type column that tracks the origin seq_type each variant.
In most cases, users should be using the related function that
is able to obtain SSMs across both genome and capture seq_type:
<a href="get_all_coding_ssm.html">get_all_coding_ssm</a>
Is this function not what you are looking for? Try one of:
<a href="get_coding_ssm_status.html">get_coding_ssm_status</a>,
get_ssm_by_patients,
<a href="get_ssm_by_sample.html">get_ssm_by_sample</a>,
<a href="get_ssm_by_samples.html">get_ssm_by_samples</a>,
<a href="get_ssm_by_region.html">get_ssm_by_region</a>,
<a href="get_ssm_by_regions.html">get_ssm_by_regions</a></p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span></span></span>
<span class="r-in"><span>  <span class="co">#basic usage (defaults to genome seq_type)</span></span></span>
<span class="r-in"><span>  <span class="va">maf_genome</span> <span class="op">=</span> <span class="fu">get_coding_ssm</span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>One or more parsing issues, call `problems()` on your data frame for details,</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> e.g.:</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   dat &lt;- vroom(...)</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   problems(dat)</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">maf_genome</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 248754</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">maf_genome</span>,<span class="fl">1</span>,<span class="fl">4</span>,<span class="fl">5</span>,<span class="fl">6</span>,<span class="fl">9</span>,<span class="va">maf_seq_type</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> genomic_data Object</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Genome Build: grch37 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Showing first 10 rows:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      Hugo_Symbol NCBI_Build Chromosome Start_Position Variant_Classification</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1     AL627309.1     GRCh37          1         138626                 Silent</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2     AL627309.1     GRCh37          1         138972        Frame_Shift_Ins</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3  RP11-206L10.9     GRCh37          1         730845          Splice_Region</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4         FAM87B     GRCh37          1         753589          Splice_Region</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5         SAMD11     GRCh37          1         871158                 Silent</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6         SAMD11     GRCh37          1         871192      Missense_Mutation</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 7         SAMD11     GRCh37          1         874416          Splice_Region</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 8         SAMD11     GRCh37          1         874467      Missense_Mutation</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 9         SAMD11     GRCh37          1         874648          Splice_Region</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 10        SAMD11     GRCh37          1         874763      Missense_Mutation</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    maf_seq_type</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1        genome</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2        genome</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3        genome</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4        genome</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5        genome</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6        genome</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 7        genome</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 8        genome</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 9        genome</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 10       genome</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span>  <span class="va">maf_exome_hg38</span> <span class="op">=</span> <span class="fu">get_coding_ssm</span><span class="op">(</span>this_seq_type <span class="op">=</span> <span class="st">"capture"</span>,</span></span>
<span class="r-in"><span>                                  projection<span class="op">=</span><span class="st">"hg38"</span><span class="op">)</span> </span></span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>The following named parsers don't match the column names: GENE_PHENO, FILTER, flanking_bps, vcf_id, vcf_qual, gnomAD_AF, gnomAD_AFR_AF, gnomAD_AMR_AF, gnomAD_SAS_AF, vcf_pos, gnomADg_AF, blacklist_count</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>One or more parsing issues, call `problems()` on your data frame for details,</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> e.g.:</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   dat &lt;- vroom(...)</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   problems(dat)</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">maf_exome_hg38</span>,<span class="fl">1</span>,<span class="fl">4</span>,<span class="fl">5</span>,<span class="fl">6</span>,<span class="fl">9</span>,<span class="va">maf_seq_type</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> genomic_data Object</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Genome Build: hg38 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Showing first 10 rows:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    Hugo_Symbol NCBI_Build Chromosome Start_Position Variant_Classification</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1        OR4F5     GRCh38       chr1          69634      Missense_Mutation</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2        OR4F5     GRCh38       chr1          69644      Missense_Mutation</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3   FO538757.2     GRCh38       chr1         183189      Missense_Mutation</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4   FO538757.2     GRCh38       chr1         183937      Missense_Mutation</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5   FO538757.1     GRCh38       chr1         186356      Nonsense_Mutation</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6   FO538757.1     GRCh38       chr1         186385      Missense_Mutation</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 7   FO538757.1     GRCh38       chr1         186404      Missense_Mutation</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 8   FO538757.1     GRCh38       chr1         186440      Missense_Mutation</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 9   FO538757.1     GRCh38       chr1         186440      Missense_Mutation</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 10  FO538757.1     GRCh38       chr1         186475          Splice_Region</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    maf_seq_type</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1       capture</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2       capture</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3       capture</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4       capture</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5       capture</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6       capture</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 7       capture</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 8       capture</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 9       capture</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 10      capture</span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Ryan Morin, Kostiantyn Dreval, Vladimir Souza.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.9000.</p>
</div>

    </footer></div>





  </body></html>

