<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Get CNV and coding SSM combined status — get_cnv_and_ssm_status • GAMBLR.results</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Get CNV and coding SSM combined status — get_cnv_and_ssm_status"><meta property="og:description" content="For each specified chromosome region (gene name), return status 1 if the copy number (CN)
state is non-neutral, i.e. different from 2, or if the region contains any coding simple somatic mutation (SSM)."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">GAMBLR.results</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Get CNV and coding SSM combined status</h1>
    
    <div class="hidden name"><code>get_cnv_and_ssm_status.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>For each specified chromosome region (gene name), return status 1 if the copy number (CN)
state is non-neutral, <em>i.e.</em> different from 2, or if the region contains any coding simple somatic mutation (SSM).</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">get_cnv_and_ssm_status</span><span class="op">(</span></span>
<span>  <span class="va">genes_and_cn_threshs</span>,</span>
<span>  <span class="va">these_samples_metadata</span>,</span>
<span>  <span class="va">maf_df</span>,</span>
<span>  <span class="va">seg_data</span>,</span>
<span>  only_cnv <span class="op">=</span> <span class="st">"none"</span>,</span>
<span>  genome_build <span class="op">=</span> <span class="st">"grch37"</span>,</span>
<span>  include_hotspots <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  review_hotspots <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  adjust_for_ploidy <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  include_silent <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  <span class="va">this_seq_type</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>genes_and_cn_threshs</dt>
<dd><p>A data frame with columns "gene_id" and "cn_thresh". The "gene_id" column stores
gene symbols (characters) which determine the regions to return CNV and/or coding SSM status. The "cn_thresh"
column stores integers that mean the maximum or minimum CN states to return status 1 (contains CNV) for
its respective gene. If this integer is below 2 (neutral CN state for diploids), it is taken as the maximum
(gene consider as tumor suppressor); if above 2, it is the minimum (oncogene); if equal to 2, do not consider
CNV to return status.</p></dd>


<dt>these_samples_metadata</dt>
<dd><p>The metadata for samples of interest to be included in the returned matrix.
Can be created with <code>get_gambl_metadata</code> function.</p></dd>


<dt>maf_df</dt>
<dd><p>Optional data frame containing the coding variants for your samples (i.e. output from <code>get_all_coding_ssm</code>)</p></dd>


<dt>seg_data</dt>
<dd><p>Optionally provide the function with a data frame of segments that will be used instead of the GAMBL flatfiles</p></dd>


<dt>only_cnv</dt>
<dd><p>A vector of gene names indicating the genes for which only CNV status should be considered,
ignoring SSM status. Set this argument to "all" or "none" (default) to apply this behavior to all or none
of the genes, respectively.</p></dd>


<dt>genome_build</dt>
<dd><p>Reference genome build. Possible values are "grch37" (default) or "hg38".</p></dd>


<dt>include_hotspots</dt>
<dd><p>Logical parameter indicating whether hotspots object should also be tabulated. Default is TRUE.</p></dd>


<dt>review_hotspots</dt>
<dd><p>Logical parameter indicating whether hotspots object should be reviewed to include
functionally relevant mutations or rare lymphoma-related genes. Default is TRUE.</p></dd>


<dt>adjust_for_ploidy</dt>
<dd><p>Set to FALSE to disable scaling of CN values by the genome-wide average per sample</p></dd>


<dt>include_silent</dt>
<dd><p>Set to TRUE if you want Synonymous mutations to also be considered</p></dd>


<dt>this_seq_type</dt>
<dd><p>Deprecated</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>A data frame with CNV and SSM combined status.</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>The user can choose from which regions are intended to return only copy number variation (CNV) status,
only coding SSM status, or at least the presence of one of them. This behavior is controlled by the arguments
<code>genes_and_cn_threshs</code> (column <code>cn_thresh</code>) and <code>only_cnv</code>.</p>
<p>This function internally calls the <code>get_cn_states</code>, <code>get_ssm_by_samples</code> and <code>get_coding_ssm_status</code>functions.
Therefore, many of its arguments are assigned to these functions. If needed, see the documentation of these
functions for more information.</p>
<p>In the case of returning NA values, this is due to the <code>get_cn_segments</code> function not being able to internally
return any copy number segments from the specified chromosome region.</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Get sample metadata including a mix of seq_type</span></span></span>
<span class="r-in"><span><span class="va">all_types_meta</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="fu"><a href="get_gambl_metadata.html">get_gambl_metadata</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="pipe.html">%&gt;%</a></span> </span></span>
<span class="r-in"><span>            <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">pathology</span> <span class="op">==</span> <span class="st">"BL"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">all_types_meta</span>, <span class="va">seq_type</span><span class="op">)</span> <span class="op"><a href="pipe.html">%&gt;%</a></span> </span></span>
<span class="r-in"><span>     <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarize</a></span><span class="op">(</span>n<span class="op">=</span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/context.html" class="external-link">n</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># A tibble: 3 × 2</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   seq_type     n</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">1</span> capture    174</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">2</span> genome     259</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">3</span> mrna       279</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># For MYC and SYNCRIP, return CNV and SSM combined status; for MIR17HG, </span></span></span>
<span class="r-in"><span><span class="co"># return only CNV status; for CCND3 return only SSM status</span></span></span>
<span class="r-in"><span><span class="va">genes_and_cn_threshs</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>  gene_id<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"MYC"</span>, <span class="st">"MIR17HG"</span>, <span class="st">"CCND3"</span>,<span class="st">"ID3"</span>,<span class="st">"DDX3X"</span>, <span class="st">"SYNCRIP"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  cn_thresh<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">3</span>, <span class="fl">3</span>, <span class="fl">2</span>, <span class="fl">2</span>, <span class="fl">2</span>, <span class="fl">1</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">genome_cnv_ssm_status</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="fu">get_cnv_and_ssm_status</span><span class="op">(</span></span></span>
<span class="r-in"><span>                           <span class="va">genes_and_cn_threshs</span>,</span></span>
<span class="r-in"><span>                           <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">all_types_meta</span>,<span class="va">seq_type</span><span class="op">==</span><span class="st">"genome"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                           only_cnv <span class="op">=</span> <span class="st">"MIR17HG"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">genome_cnv_ssm_status</span><span class="op">)</span><span class="op">)</span>    </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 259   6</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">genome_cnv_ssm_status</span><span class="op">)</span>   </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                           MYC MIR17HG CCND3 ID3 DDX3X SYNCRIP</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> BLGSP-71-06-00001-01A-11D   0       0     1   1     1       0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> BLGSP-71-06-00002-01C-01D   1       0     0   1     0       0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> BLGSP-71-06-00004-01A-11D   0       0     1   1     1       0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> BLGSP-71-06-00005-01A-21D   0       0     1   1     0       0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> BLGSP-71-06-00007-01A-11D   1       0     1   1     0       0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> BLGSP-71-06-00008-01A-11D   0       0     0   0     0       0</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">colSums</a></span><span class="op">(</span><span class="va">genome_cnv_ssm_status</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     MYC MIR17HG   CCND3     ID3   DDX3X SYNCRIP </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     187      47      76     120     122      14 </span>
<span class="r-in"><span>                    </span></span>
<span class="r-in"><span><span class="va">all_seq_type_status</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="fu">get_cnv_and_ssm_status</span><span class="op">(</span></span></span>
<span class="r-in"><span>                           <span class="va">genes_and_cn_threshs</span>,</span></span>
<span class="r-in"><span>                           <span class="va">all_types_meta</span>,</span></span>
<span class="r-in"><span>                           only_cnv <span class="op">=</span> <span class="st">"MIR17HG"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">all_seq_type_status</span><span class="op">)</span><span class="op">)</span>   </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 433   6</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">all_seq_type_status</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                           MYC MIR17HG CCND3 ID3 DDX3X SYNCRIP</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> BLGSP-71-06-00001-01A-11D   0       0     1   1     1       0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> BLGSP-71-06-00002-01C-01D   1       0     0   1     0       0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> BLGSP-71-06-00004-01A-11D   0       0     1   1     1       0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> BLGSP-71-06-00005-01A-21D   0       0     1   1     0       0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> BLGSP-71-06-00007-01A-11D   1       0     1   1     0       0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> BLGSP-71-06-00008-01A-11D   0       0     0   0     0       0</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">colSums</a></span><span class="op">(</span><span class="va">all_seq_type_status</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     MYC MIR17HG   CCND3     ID3   DDX3X SYNCRIP </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     288      68     120     197     192      26 </span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Ryan Morin, Kostiantyn Dreval, Vladimir Souza.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

