<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Get GAMBL metadata. — og_get_gambl_metadata • GAMBLR.results</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Get GAMBL metadata. — og_get_gambl_metadata"><meta name="description" content="Return metadata for a selection of samples."><meta property="og:description" content="Return metadata for a selection of samples."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">GAMBLR.results</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.3</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/GAMBLR-results.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Developers</h6></li>
    <li><a class="dropdown-item" href="../articles/testing.html">Testing GAMBLR packages</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Get GAMBL metadata.</h1>

      <div class="d-none name"><code>og_get_gambl_metadata.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Return metadata for a selection of samples.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">og_get_gambl_metadata</span><span class="op">(</span></span>
<span>  seq_type_filter <span class="op">=</span> <span class="st">"genome"</span>,</span>
<span>  tissue_status_filter <span class="op">=</span> <span class="st">"tumour"</span>,</span>
<span>  <span class="va">case_set</span>,</span>
<span>  remove_benchmarking <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  with_outcomes <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  <span class="va">sample_flatfile</span>,</span>
<span>  <span class="va">biopsy_flatfile</span>,</span>
<span>  only_available <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  seq_type_priority <span class="op">=</span> <span class="st">"genome"</span>,</span>
<span>  <span class="va">from_flatfile</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-seq-type-filter">seq_type_filter<a class="anchor" aria-label="anchor" href="#arg-seq-type-filter"></a></dt>
<dd><p>Filtering criteria (default: all genomes).</p></dd>


<dt id="arg-tissue-status-filter">tissue_status_filter<a class="anchor" aria-label="anchor" href="#arg-tissue-status-filter"></a></dt>
<dd><p>Filtering criteria for tissue status. Possible values are "tumour" (the default) or "normal".</p></dd>


<dt id="arg-case-set">case_set<a class="anchor" aria-label="anchor" href="#arg-case-set"></a></dt>
<dd><p>Optional short name for a pre-defined set of cases avoiding any embargoed cases (current options: 'BLGSP-study', 'FL-study', 'DLBCL-study', 'FL-DLBCL-study', 'FL-DLBCL-all', 'DLBCL-unembargoed', 'BL-DLBCL-manuscript', 'MCL','MCL-CLL').</p></dd>


<dt id="arg-remove-benchmarking">remove_benchmarking<a class="anchor" aria-label="anchor" href="#arg-remove-benchmarking"></a></dt>
<dd><p>By default the FFPE benchmarking duplicate samples will be dropped.</p></dd>


<dt id="arg-with-outcomes">with_outcomes<a class="anchor" aria-label="anchor" href="#arg-with-outcomes"></a></dt>
<dd><p>Optionally join to gambl outcome data.</p></dd>


<dt id="arg-sample-flatfile">sample_flatfile<a class="anchor" aria-label="anchor" href="#arg-sample-flatfile"></a></dt>
<dd><p>Optionally provide the full path to a sample table to use instead of the default.</p></dd>


<dt id="arg-biopsy-flatfile">biopsy_flatfile<a class="anchor" aria-label="anchor" href="#arg-biopsy-flatfile"></a></dt>
<dd><p>Optionally provide the full path to a biopsy table to use instead of the default.</p></dd>


<dt id="arg-only-available">only_available<a class="anchor" aria-label="anchor" href="#arg-only-available"></a></dt>
<dd><p>If TRUE, will remove samples with FALSE or NA in the bam_available column (default: TRUE).</p></dd>


<dt id="arg-seq-type-priority">seq_type_priority<a class="anchor" aria-label="anchor" href="#arg-seq-type-priority"></a></dt>
<dd><p>For duplicate sample_id with different seq_type available, the metadata will prioritize this seq_type and drop the others. Possible values are "genome" or "capture".</p></dd>


<dt id="arg-from-flatfile">from_flatfile<a class="anchor" aria-label="anchor" href="#arg-from-flatfile"></a></dt>
<dd><p>Deprecated (will be ignored)</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A data frame with metadata for each biopsy in GAMBL</p>
<dl><dt>compression</dt>
<dd><p>Format of the original data used as input for our analysis pipelines (cram, bam or fastq)</p></dd>

<dt>bam_available</dt>
<dd><p>Whether or not this file was available when last checked.</p></dd>

<dt>patient_id</dt>
<dd><p>The anonymized unique identifier for this patient. For BC samples, this will be Res ID.</p></dd>

<dt>sample_id</dt>
<dd><p>A unique identifier for the sample analyzed.</p></dd>

<dt>seq_type</dt>
<dd><p>The assay type used to produce this data (one of "genome","capture, "mrna", "promethION")</p></dd>

<dt>genome_build</dt>
<dd><p>The name of the genome reference the data were aligned to.</p></dd>

<dt>tissue_status</dt>
<dd><p>Whether the sample was atumour or normal.</p></dd>

<dt>cohort</dt>
<dd><p>Name for a group of samples that were added together (usually from a single study), often in the format pathology_cohort_descriptor.</p></dd>

<dt>library_id</dt>
<dd><p>The unique identifier for the sequencing library.</p></dd>

<dt>pathology</dt>
<dd><p>The diagnosis or pathology for the sample</p></dd>

<dt>time_point</dt>
<dd><p>Timing of biopsy in increasing alphabetical order (A = diagnosis, B = first relapse etc)</p></dd>

<dt>protocol</dt>
<dd><p>General protocol for library construction. e.g. "Ribodepletion", "PolyA", or "Genome"</p></dd>

<dt>ffpe_or_frozen</dt>
<dd><p>Whether the nucleic acids were extracted from a frozen or FFPE sample</p></dd>

<dt>read_length</dt>
<dd><p>The length of reads (required for RNA-seq libraries)</p></dd>

<dt>strandedness</dt>
<dd><p>Whether the RNA-seq librayr construction was strand-specific and, if so, which strand. Required for RNAseq; "positive", "negative", or "unstranded")</p></dd>

<dt>seq_source_type</dt>
<dd><p>Required for RNAseq. Usually the same value as ffpe_or_frozen but sometimes immunotube or sorted cells</p></dd>

<dt>data_path</dt>
<dd><p>Symbolic link to the bam or cram file (not usually relevant for GAMBLR)</p></dd>

<dt>link_name</dt>
<dd><p>Standardized naming for symbolic link (not usually relevant for GAMBLR)</p></dd>

<dt>data_path</dt>
<dd><p>Symbolic link to the fastq file (not usually relevant for GAMBLR)</p></dd>

<dt>fastq_link_name</dt>
<dd><p>Standardized naming for symbolic link for FASTQ file, if used (not usually relevant for GAMBLR)</p></dd>

<dt>unix_group</dt>
<dd><p>Whether a source is external and restricted by data access agreements (icgc_dart) or internal (gambl)</p></dd>

<dt>COO_consensus</dt>
<dd><p>TODO</p></dd>

<dt>DHITsig_consensus</dt>
<dd><p>TODO</p></dd>

<dt>COO_PRPS_class</dt>
<dd><p>TODO</p></dd>

<dt>DHITsig_PRPS_class</dt>
<dd><p>TODO</p></dd>

<dt>DLBCL90_dlbcl_call</dt>
<dd><p>TODO</p></dd>

<dt>DLBCL90_dhitsig_call</dt>
<dd><p>TODO</p></dd>

<dt>res_id</dt>
<dd><p>duplicate of sample_id for local samples and NA otherwise</p></dd>

<dt>DLBCL90_code_set</dt>
<dd><p>Code set used for DLBCL90 call. One of DLBCL90 DLBCL90v2 DLBCL90v3</p></dd>

<dt>DLBCL90_dlbcl_score</dt>
<dd><p>TODO</p></dd>

<dt>DLBCL90_pmbl_score</dt>
<dd><p>TODO</p></dd>

<dt>DLBCL90_pmbl_call</dt>
<dd><p>TODO</p></dd>

<dt>DLBCL90_dhitsig_score</dt>
<dd><p>TODO</p></dd>

<dt>myc_ba</dt>
<dd><p>Result from breakapart FISH for MYC locus</p></dd>

<dt>myc_cn</dt>
<dd><p>Result from copy number FISH for MYC locus</p></dd>

<dt>bcl2_ba</dt>
<dd><p>Result from breakapart FISH for BCL2 locus</p></dd>

<dt>bcl2_cn</dt>
<dd><p>Result from copy number FISH for BCL2 locus</p></dd>

<dt>bcl6_ba</dt>
<dd><p>Result from breakapart FISH for BCL6 locus</p></dd>

<dt>bcl6_cn</dt>
<dd><p>Result from copy number FISH for BCL6 locus</p></dd>

<dt>time_since_diagnosis_years</dt>
<dd><p>TODO</p></dd>

<dt>relapse_timing</dt>
<dd><p>TODO</p></dd>

<dt>dtbx</dt>
<dd><p>TODO. OR REMOVE?</p></dd>

<dt>dtdx</dt>
<dd><p>TODO. OR REMOVE?</p></dd>

<dt>lymphgen_no_cnv</dt>
<dd><p>TODO</p></dd>

<dt>lymphgen_with_cnv</dt>
<dd><p>TODO</p></dd>

<dt>lymphgen_cnv_noA53</dt>
<dd><p>TODO</p></dd>

<dt>lymphgen_wright</dt>
<dd><p>The LymphGen call for this sample from Wright et all (if applicable)</p></dd>

<dt>fl_grade</dt>
<dd><p>TODO</p></dd>

<dt>capture_frozen_sample_id</dt>
<dd><p>TODO</p></dd>

<dt>capture_FFPE_sample_id</dt>
<dd><p>TODO</p></dd>

<dt>capture_unknown_sample_id</dt>
<dd><p>TODO</p></dd>

<dt>genome_frozen_sample_id</dt>
<dd><p>TODO</p></dd>

<dt>genome_ctDNA_sample_id</dt>
<dd><p>TODO</p></dd>

<dt>genome_FFPE_sample_id</dt>
<dd><p>TODO</p></dd>

<dt>mrna_PolyA_frozen_sample_id</dt>
<dd><p>TODO</p></dd>

<dt>mrna_Ribodepletion_frozen_sample_id</dt>
<dd><p>TODO</p></dd>

<dt>mrna_Ribodepletion_frozen_sample_id</dt>
<dd><p>TODO</p></dd>

<dt>XXX_cohort</dt>
<dd><p>Cohort name for batch effect correction(?)</p></dd>

<dt>transformation</dt>
<dd><p>TODO</p></dd>

<dt>relapse</dt>
<dd><p>TODO</p></dd>

<dt>ighv_mutation_original</dt>
<dd><p>TODO</p></dd>

<dt>normal_sample_id</dt>
<dd><p>TODO</p></dd>

<dt>pairing_status</dt>
<dd><p>TODO</p></dd>

<dt>ICGC_ID</dt>
<dd><p>TODO</p></dd>

<dt>ICGC_XXX</dt>
<dd><p>metadata value for ICGC cohort inferred from external metadata</p></dd>

<dt>detailed_pathology</dt>
<dd><p>TODO</p></dd>

<dt>COO_final</dt>
<dd><p>TODO</p></dd>

<dt>consensus_pathology</dt>
<dd><p>TODO</p></dd>

<dt>lymphgen</dt>
<dd><p>TODO</p></dd>

<dt>Tumor_Sample_Barcode</dt>
<dd><p>Duplicate of sample_id for simplifying joins to MAF data frames</p></dd>

<dt>consensus_coo_dhitsig</dt>
<dd><p>TODO</p></dd>

<dt>pathology_rank</dt>
<dd><p>Numeric rank for consistent ordering of samples by pathology</p></dd>

<dt>lymphgen_rank</dt>
<dd><p>Numeric rank for consistent ordering of samples by LymphGen</p></dd>

<dt>hiv_status</dt>
<dd><p>TODO</p></dd>

<dt>CODE_XXX</dt>
<dd><p>Event-free status at last follow-up for overall survival (OS), progression-free survival (PFS) etc. 0 = no event/censored. 1 = event</p></dd>

<dt>XXX_YEARS</dt>
<dd><p>Time, in years, from diagnosis to last follow-up for overall survival (OS), progression-free survival (PFS)</p></dd>

<dt>alive</dt>
<dd><p>Theoretically redundant with CODE_OS</p></dd>

<dt>is_adult</dt>
<dd><p>Adult or pediatric at diagnosis. One of "Adult" for adults and "Pediatric" otherwise</p></dd>

<dt>age_group</dt>
<dd><p>Adult_BL or Pediatric_BL or Other, specific to the BLGSP study</p></dd>

<dt>age</dt>
<dd><p>patient age at diagnosis</p></dd>

<dt>sex</dt>
<dd><p>The biological sex of the patient, if available. Allowable options: M, F, NA</p></dd>

<dt>tx_primary</dt>
<dd><p>TODO</p></dd>

<dt>cause_of_death</dt>
<dd><p>TODO</p></dd>


</dl></div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>This function returns metadata for GAMBL samples. Options for subset and filter the returned data are available.
For more information on how to use this function with different filtering criteria, refer to the parameter descriptions,
examples and vignettes. Embargoed cases (current options: 'BLGSP-study', 'FL-study', 'DLBCL-study', 'FL-DLBCL-study', 'FL-DLBCL-all', 'DLBCL-unembargoed', 'BL-DLBCL-manuscript', 'MCL','MCL-CLL')</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span> <span class="co"># \dontrun{</span></span></span>
<span class="r-in"><span><span class="co">#basic usage</span></span></span>
<span class="r-in"><span><span class="va">my_metadata</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="fu"><a href="get_gambl_metadata.html">get_gambl_metadata</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">#use pre-defined custom sample sets</span></span></span>
<span class="r-in"><span><span class="va">only_blgsp_metadata</span> <span class="op">=</span> <span class="fu"><a href="get_gambl_metadata.html">get_gambl_metadata</a></span><span class="op">(</span>case_set <span class="op">=</span> <span class="st">"BLGSP-study"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">#override default filters and request metadata for samples other than tumour genomes,</span></span></span>
<span class="r-in"><span><span class="co">#e.g. also get the normals</span></span></span>
<span class="r-in"><span><span class="va">tumour_and_normal_metadata</span> <span class="op">=</span> <span class="fu"><a href="get_gambl_metadata.html">get_gambl_metadata</a></span><span class="op">(</span>tissue_status_filter <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'tumour'</span>,<span class="st">'normal'</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">non_redundant_genome_and_capture</span> <span class="op">=</span> <span class="fu"><a href="get_gambl_metadata.html">get_gambl_metadata</a></span><span class="op">(</span>seq_type_filter <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'genome'</span>, <span class="st">'capture'</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                                                       seq_type_priority <span class="op">=</span> <span class="st">"genome"</span><span class="op">)</span></span></span>
<span class="r-in"><span>                                                       </span></span>
<span class="r-in"><span><span class="va">absolutely_everything</span> <span class="op">=</span> <span class="fu"><a href="get_gambl_metadata.html">get_gambl_metadata</a></span><span class="op">(</span>seq_type_filter <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'genome'</span>, <span class="st">'capture'</span>,<span class="st">'mrna'</span><span class="op">)</span>, tissue_status_filter<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'tumour'</span>,<span class="st">'normal'</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span> <span class="co"># }</span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Ryan Morin, Kostiantyn Dreval, Vladimir Souza.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.9000.</p>
</div>

    </footer></div>





  </body></html>

